<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>14&nbsp; Adventures in Covariance – *Statistical rethinking* 2 with brms and the tidyverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./15.html" rel="next">
<link href="./13.html" rel="prev">
<link href="./mischa_favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-234273d1456647dabc34a594ac50e507.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-7d082c015fa132ca880f975003d7bf2f.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<meta property="og:title" content="14&nbsp; Adventures in Covariance – Statistical rethinking 2 with brms and the tidyverse">
<meta property="og:description" content="">
<meta property="og:image" content="https://solomon.quarto.pub/sr2/14_files/figure-html/unnamed-chunk-4-1.png">
<meta property="og:site_name" content="*Statistical rethinking* 2 with brms and the tidyverse">
<meta property="og:image:height" content="720">
<meta property="og:image:width" content="1536">
</head>

<body class="nav-sidebar floating slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./14.html"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adventures in Covariance</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./"><em>Statistical rethinking</em> 2 with brms and the tidyverse</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The Golem of Prague</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Small Worlds and Large Worlds</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sampling the Imaginary</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Geocentric Models</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">The Many Variables &amp; The Spurious Waffles</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">The Haunted DAG &amp; The Causal Terror</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Ulysses’ Compass</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Conditional Manatees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Markov Chain Monte Carlo</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Big Entropy and the Generalized Linear Model</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">God Spiked the Integers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Monsters and Mixtures</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Models With Memory</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adventures in Covariance</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Missing Data and Other Opportunities</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Generalized Linear Madness</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><del>Horoscopes</del> Insights</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#varying-slopes-by-construction" id="toc-varying-slopes-by-construction" class="nav-link active" data-scroll-target="#varying-slopes-by-construction"><span class="header-section-number">14.1</span> Varying slopes by construction</a>
  <ul>
  <li><a href="#rethinking-why-gaussian" id="toc-rethinking-why-gaussian" class="nav-link" data-scroll-target="#rethinking-why-gaussian"><span class="header-section-number">14.1.0.1</span> Rethinking: Why Gaussian?</a></li>
  <li><a href="#simulate-the-population" id="toc-simulate-the-population" class="nav-link" data-scroll-target="#simulate-the-population"><span class="header-section-number">14.1.1</span> Simulate the population</a></li>
  <li><a href="#simulate-observations" id="toc-simulate-observations" class="nav-link" data-scroll-target="#simulate-observations"><span class="header-section-number">14.1.2</span> Simulate observations</a>
  <ul>
  <li><a href="#rethinking-simulation-and-misspecification" id="toc-rethinking-simulation-and-misspecification" class="nav-link" data-scroll-target="#rethinking-simulation-and-misspecification"><span class="header-section-number">14.1.2.1</span> Rethinking: Simulation and misspecification</a></li>
  </ul></li>
  <li><a href="#the-varying-slopes-model" id="toc-the-varying-slopes-model" class="nav-link" data-scroll-target="#the-varying-slopes-model"><span class="header-section-number">14.1.3</span> The varying slopes model</a></li>
  </ul></li>
  <li><a href="#advanced-varying-slopes" id="toc-advanced-varying-slopes" class="nav-link" data-scroll-target="#advanced-varying-slopes"><span class="header-section-number">14.2</span> Advanced varying slopes</a></li>
  <li><a href="#instruments-and-causal-designs" id="toc-instruments-and-causal-designs" class="nav-link" data-scroll-target="#instruments-and-causal-designs"><span class="header-section-number">14.3</span> Instruments and causal designs</a>
  <ul>
  <li><a href="#instrumental-variables" id="toc-instrumental-variables" class="nav-link" data-scroll-target="#instrumental-variables"><span class="header-section-number">14.3.1</span> Instrumental variables</a>
  <ul>
  <li><a href="#rethinking-two-stage-worst-squares" id="toc-rethinking-two-stage-worst-squares" class="nav-link" data-scroll-target="#rethinking-two-stage-worst-squares"><span class="header-section-number">14.3.1.1</span> Rethinking: Two-stage worst squares</a></li>
  </ul></li>
  <li><a href="#other-designs" id="toc-other-designs" class="nav-link" data-scroll-target="#other-designs"><span class="header-section-number">14.3.2</span> Other designs</a></li>
  </ul></li>
  <li><a href="#sec-Social-relations-as" id="toc-sec-Social-relations-as" class="nav-link" data-scroll-target="#sec-Social-relations-as"><span class="header-section-number">14.4</span> Social relations as correlated varying effects</a></li>
  <li><a href="#continuous-categories-and-the-gaussian-process" id="toc-continuous-categories-and-the-gaussian-process" class="nav-link" data-scroll-target="#continuous-categories-and-the-gaussian-process"><span class="header-section-number">14.5</span> Continuous categories and the Gaussian process</a>
  <ul>
  <li><a href="#example-spatial-autocorrelation-in-oceanic-tools" id="toc-example-spatial-autocorrelation-in-oceanic-tools" class="nav-link" data-scroll-target="#example-spatial-autocorrelation-in-oceanic-tools"><span class="header-section-number">14.5.1</span> Example: Spatial autocorrelation in Oceanic tools</a>
  <ul>
  <li><a href="#rethinking-dispersion-by-other-names" id="toc-rethinking-dispersion-by-other-names" class="nav-link" data-scroll-target="#rethinking-dispersion-by-other-names"><span class="header-section-number">14.5.1.0.1</span> Rethinking: Dispersion by other names</a></li>
  </ul></li>
  <li><a href="#example-phylogenetic-distance" id="toc-example-phylogenetic-distance" class="nav-link" data-scroll-target="#example-phylogenetic-distance"><span class="header-section-number">14.5.2</span> Example: Phylogenetic distance</a></li>
  </ul></li>
  <li><a href="#summary-bonus-multilevel-growth-models-and-the-melsm" id="toc-summary-bonus-multilevel-growth-models-and-the-melsm" class="nav-link" data-scroll-target="#summary-bonus-multilevel-growth-models-and-the-melsm"><span class="header-section-number">14.6</span> <del>Summary</del> Bonus: Multilevel growth models and the MELSM</a>
  <ul>
  <li><a href="#borrow-some-data" id="toc-borrow-some-data" class="nav-link" data-scroll-target="#borrow-some-data"><span class="header-section-number">14.6.1</span> Borrow some data</a></li>
  <li><a href="#conventional-multilevel-growth-model" id="toc-conventional-multilevel-growth-model" class="nav-link" data-scroll-target="#conventional-multilevel-growth-model"><span class="header-section-number">14.6.2</span> Conventional multilevel growth model</a></li>
  <li><a href="#learn-more-about-your-data-with-the-melsm" id="toc-learn-more-about-your-data-with-the-melsm" class="nav-link" data-scroll-target="#learn-more-about-your-data-with-the-melsm"><span class="header-section-number">14.6.3</span> Learn more about your data with the MELSM</a></li>
  <li><a href="#time-to-go-multivariate" id="toc-time-to-go-multivariate" class="nav-link" data-scroll-target="#time-to-go-multivariate"><span class="header-section-number">14.6.4</span> Time to go multivariate</a>
  <ul>
  <li><a href="#plot-with-uncertainty" id="toc-plot-with-uncertainty" class="nav-link" data-scroll-target="#plot-with-uncertainty"><span class="header-section-number">14.6.4.1</span> Plot with uncertainty</a></li>
  </ul></li>
  <li><a href="#growth-modelmelsm-wrap-up" id="toc-growth-modelmelsm-wrap-up" class="nav-link" data-scroll-target="#growth-modelmelsm-wrap-up"><span class="header-section-number">14.6.5</span> Growth model/MELSM wrap-up</a></li>
  </ul></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  <li><a href="#comments" id="toc-comments" class="nav-link" data-scroll-target="#comments">Comments</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-Adventures-in-Covariance" class="quarto-section-identifier"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adventures in Covariance</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<blockquote class="blockquote">
<p>In this chapter, you’ll see how to… specify <strong>varying slopes</strong> in combination with the varying intercepts of the previous chapter. This will enable pooling that will improve estimates of how different units respond to or are influenced by predictor variables. It will also improve estimates of intercepts, by borrowing information across parameter types. Essentially, varying slopes models are massive interaction machines. They allow every unit in the data to have its own response to any treatment or exposure or event, while also improving estimates via pooling. When the variation in slopes is large, the average slope is of less interest. Sometimes, the pattern of variation in slopes provides hints about omitted variables that explain why some units respond more or less. We’ll see an example in this chapter.</p>
<p>The machinery that makes such complex varying effects possible will be used later in the chapter to extend the varying effects strategy to more subtle model types, including the use of continuous categories, using <strong>Gaussian process</strong>. Ordinary varying effects work only with discrete, unordered categories, such as individuals, countries, or ponds. In these cases, each category is equally different from all of the others. But it is possible to use pooling with categories such as age or location. In these cases, some ages and some locations are more similar than others. You’ll see how to model covariation among continuous categories of this kind, as well as how to generalize the strategy to seemingly unrelated types of models such as phylogenetic and network regressions. Finally, we’ll circle back to causal inference and use our new powers over covariance to go beyond the tools of Chapter 6, introducing <strong>instrumental variables</strong>. Instruments are ways of inferring cause without closing backdoor paths. However they are very tricky both in design and estimation. <span class="citation" data-cites="mcelreathStatisticalRethinkingBayesian2020">(<a href="references.html#ref-mcelreathStatisticalRethinkingBayesian2020" role="doc-biblioref">McElreath, 2020, pp. 436–437</a>, <strong>emphasis</strong> in the original)</span></p>
</blockquote>
<section id="varying-slopes-by-construction" class="level2" data-number="14.1">
<h2 data-number="14.1" class="anchored" data-anchor-id="varying-slopes-by-construction"><span class="header-section-number">14.1</span> Varying slopes by construction</h2>
<blockquote class="blockquote">
<p>How should the robot pool information across intercepts and slopes? By modeling the joint population of intercepts and slopes, which means by modeling their covariance. In conventional multilevel models, the device that makes this possible is a joint multivariate Gaussian distribution for all of the varying effects, both intercepts and slopes. So instead of having two independent Gaussian distributions of intercepts and of slopes, the robot can do better by assigning a two-dimensional Gaussian distribution to both the intercepts (first dimension) and the slopes (second dimension). (p.&nbsp;437)</p>
</blockquote>
<section id="rethinking-why-gaussian" class="level4" data-number="14.1.0.1">
<h4 data-number="14.1.0.1" class="anchored" data-anchor-id="rethinking-why-gaussian"><span class="header-section-number">14.1.0.1</span> Rethinking: Why Gaussian?</h4>
<blockquote class="blockquote">
<p>There is no reason the multivariate distribution of intercepts and slopes must be Gaussian. But there are both practical and epistemological justifications. On the practical side, there aren’t many multivariate distributions that are easy to work with. The only common ones are multivariate Gaussian and multivariate Student-t distributions. On the epistemological side, if all we want to say about these intercepts and slopes is their means, variances, and covariances, then the maximum entropy distribution is multivariate Gaussian. (p.&nbsp;437)</p>
</blockquote>
<p>As it turns out, <strong>brms</strong> does currently allow users to use the multivariate Student-<span class="math inline">\(t\)</span> distribution in this way. For details, check out <a href="https://github.com/paul-buerkner/brms/issues/231">this discussion from the <strong>brms</strong> GitHub repository</a>. Bürkner’s exemplar syntax from his comment on May 13, 2018, was <code>y ~ x + (x | gr(g, dist = "student"))</code>. I haven’t experimented with this, but if you do, do consider <a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed/issues">sharing how it went</a>.</p>
</section>
<section id="simulate-the-population" class="level3" data-number="14.1.1">
<h3 data-number="14.1.1" class="anchored" data-anchor-id="simulate-the-population"><span class="header-section-number">14.1.1</span> Simulate the population</h3>
<p>If you follow this section closely, it’s a great template for simulating multilevel code for any of your future projects. You might think of this as an alternative to a frequentist power analysis. Vourre has done <a href="https://gitlab.com/vuorre/bayesplan">some nice work along these lines</a>, I have a <a href="https://solomonkurz.netlify.app/blog/bayesian-power-analysis-part-i/">blog series</a> on Bayesian power analysis, and Kruschke covered the topic in Chapter 13 of his <span class="citation" data-cites="kruschkeDoingBayesianData2015">(<a href="references.html#ref-kruschkeDoingBayesianData2015" role="doc-biblioref">2015</a>)</span> <a href="https://sites.google.com/site/doingbayesiandataanalysis/">text</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>a       <span class="ot">&lt;-</span>  <span class="fl">3.5</span>  <span class="co"># Average morning wait time</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>b       <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span>    <span class="co"># Average difference afternoon wait time</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>sigma_a <span class="ot">&lt;-</span>  <span class="dv">1</span>    <span class="co"># SD in intercepts</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>sigma_b <span class="ot">&lt;-</span>  <span class="fl">0.5</span>  <span class="co"># SD in slopes</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>rho     <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fl">0.7</span>  <span class="co"># Correlation between intercepts and slopes</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># The next three lines of code simply combine the terms, above</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>mu     <span class="ot">&lt;-</span> <span class="fu">c</span>(a, b)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>cov_ab <span class="ot">&lt;-</span> sigma_a <span class="sc">*</span> sigma_b <span class="sc">*</span> rho</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>sigma  <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(sigma_a<span class="sc">^</span><span class="dv">2</span>, cov_ab, </span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>                   cov_ab, sigma_b<span class="sc">^</span><span class="dv">2</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It’s common to refer to a covariance matrix as <span class="math inline">\(\mathbf \Sigma\)</span>. The mathematical notation for those last couple lines of code is</p>
<p><span class="math display">\[
\mathbf \Sigma = \begin{bmatrix} \sigma_\alpha^2 &amp; \sigma_\alpha \sigma_\beta \rho \\ \sigma_\alpha \sigma_\beta \rho &amp; \sigma_\beta^2 \end{bmatrix}.
\]</span></p>
<p>Anyway, if you haven’t used the <code>matrix()</code> function before, you might get a sense of the elements like so.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    1    3
[2,]    2    4</code></pre>
</div>
</div>
<p>This next block of code will finally yield our café data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>sigmas <span class="ot">&lt;-</span> <span class="fu">c</span>(sigma_a, sigma_b)          <span class="co"># SDs</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>rho    <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, rho,             <span class="co"># Correlation matrix</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                   rho, <span class="dv">1</span>), <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Now matrix multiply to get covariance matrix</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>sigma <span class="ot">&lt;-</span> <span class="fu">diag</span>(sigmas) <span class="sc">%*%</span> rho <span class="sc">%*%</span> <span class="fu">diag</span>(sigmas)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co"># How many cafes would you like?</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>n_cafes <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5</span>)  <span class="co"># Used to replicate example</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>vary_effects <span class="ot">&lt;-</span> MASS<span class="sc">::</span><span class="fu">mvrnorm</span>(n_cafes, mu, sigma) <span class="sc">|&gt;</span> </span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_names</span>(<span class="st">"a_cafe"</span>, <span class="st">"b_cafe"</span>)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(vary_effects)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    a_cafe     b_cafe
1 4.223962 -1.6093565
2 2.010498 -0.7517704
3 4.565811 -1.9482646
4 3.343635 -1.1926539
5 1.700971 -0.5855618
6 4.134373 -1.1444539</code></pre>
</div>
</div>
<p>Let’s make sure we’re keeping this all straight. <code>a_cafe</code> = our café-specific intercepts; <code>b_cafe</code> = our café-specific slopes. These aren’t the actual data, yet. But at this stage, it might make sense to ask <em>What’s the distribution of <code>a_cafe</code> and <code>b_cafe</code>?</em> Our variant of Figure 14.2 contains the answer.</p>
<p>For our plots in this chapter, we’ll make our own custom <strong>ggplot2</strong> theme. The color palette will come from the “pearl_earring” palette of the <a href="https://github.com/EdwinTh/dutchmasters"><strong>dutchmasters</strong> package</a> <span class="citation" data-cites="R-dutchmasters">(<a href="references.html#ref-R-dutchmasters" role="doc-biblioref">Thoen, 2022</a>)</span>. You can learn more about the original painting, Vermeer’s <span class="citation" data-cites="vermeerGirlPearlEarring1665">(<a href="references.html#ref-vermeerGirlPearlEarring1665" role="doc-biblioref">1665</a>)</span> <em>Girl with a Pearl Earring</em>, <a href="https://en.wikipedia.org/wiki/Girl_with_a_Pearl_Earring">here</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dutchmasters)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>dutchmasters<span class="sc">$</span>pearl_earring</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        red(lips)              skin      blue(scarf1)      blue(scarf2)      white(colar)       gold(dress)      gold(dress2) 
        "#A65141"         "#E7CDC2"         "#80A0C7"         "#394165"         "#FCF9F0"         "#B1934A"         "#DCA258" 
black(background)      grey(scarf3)    yellow(scarf4)                   
        "#100F14"         "#8B9DAF"         "#EEDA9D"         "#E8DCCF" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>scales<span class="sc">::</span><span class="fu">show_col</span>(dutchmasters<span class="sc">$</span>pearl_earring)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>We’ll name our custom theme <code>theme_pearl_earring()</code>. I cobbled together this approach to defining a custom <strong>ggplot2</strong> theme with help from</p>
<ul>
<li><a href="https://ggplot2-book.org/programming.html">Chapter 19</a> of Wichkam’s <span class="citation" data-cites="wickhamGgplot2ElegantGraphics2016">(<a href="references.html#ref-wickhamGgplot2ElegantGraphics2016" role="doc-biblioref">2016</a>)</span> <em>ggplot2: Elegant graphics for data analysis</em>;</li>
<li><a href="https://bookdown.org/rdpeng/RProgDA/building-a-new-theme.html">Section 4.6</a> of Peng, Kross, and Anderson’s <span class="citation" data-cites="pengMasteringSoftwareDevelopment2017">(<a href="references.html#ref-pengMasteringSoftwareDevelopment2017" role="doc-biblioref">2017</a>)</span> <a href="https://bookdown.org/rdpeng/RProgDA/building-a-new-theme.html"><em>Mastering Software Development in R</em></a>;</li>
<li>Lea Waniek’s blog post, <a href="https://www.statworx.com/de/blog/custom-themes-in-ggplot2/"><em>Custom themes in ggplot2</em></a>, and</li>
<li>Joey Stanley’s blog post of the same name, <a href="https://joeystanley.com/blog/custom-themes-in-ggplot2"><em>Custom themes in ggplot2</em></a>.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>theme_pearl_earring <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">light_color =</span> <span class="st">"#E8DCCF"</span>, </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">dark_color =</span> <span class="st">"#100F14"</span>, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">my_family =</span> <span class="st">"Courier"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                                ...) {</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> light_color),</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> light_color, <span class="at">family =</span> my_family),</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.line =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> light_color),</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_line</span>(<span class="at">color =</span> light_color),</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> dark_color, <span class="at">color =</span> <span class="st">"transparent"</span>),</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.key =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> dark_color, <span class="at">color =</span> <span class="st">"transparent"</span>),</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> dark_color, <span class="at">color =</span> light_color),</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> dark_color, <span class="at">color =</span> dark_color),</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> dark_color, <span class="at">color =</span> <span class="st">"transparent"</span>),</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> light_color, <span class="at">family =</span> my_family),</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>        ...)</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Now set `theme_pearl_earring()` as the default theme</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_pearl_earring</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note how our custom <code>theme_pearl_earing()</code> function has a few adjustable parameters. Feel free to play around with alternative settings to see how they work. If we just use the defaults as we have defined them, here is our Figure 14.2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>vary_effects <span class="sc">|&gt;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> a_cafe, <span class="at">y =</span> b_cafe)) <span class="sc">+</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"#80A0C7"</span>) <span class="sc">+</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="312"></p>
</figure>
</div>
</div>
</div>
<p>Again, these are not “data.” Figure 14.2 shows a distribution of <em>parameters</em>. Here’s their Pearson’s correlation coefficient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(vary_effects<span class="sc">$</span>a_cafe, vary_effects<span class="sc">$</span>b_cafe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -0.5721537</code></pre>
</div>
</div>
<p>Since there are only 20 rows in our <code>vary_effects</code> simulation, it shouldn’t be a surprise that the Pearson’s correlation is a bit off from the population value of <span class="math inline">\(\rho = -.7\)</span>. If you rerun the simulation with <code>n_cafes &lt;- 200</code>, the Pearson’s correlation is much closer to the data generating value.</p>
</section>
<section id="simulate-observations" class="level3" data-number="14.1.2">
<h3 data-number="14.1.2" class="anchored" data-anchor-id="simulate-observations"><span class="header-section-number">14.1.2</span> Simulate observations</h3>
<p>Here we put those simulated parameters to use and simulate actual data from them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>n_visits <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sigma    <span class="ot">&lt;-</span>  <span class="fl">0.5</span>  <span class="co"># SD within cafes</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">22</span>)  <span class="co"># Used to replicate example</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> vary_effects <span class="sc">|&gt;</span> </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cafe =</span> <span class="dv">1</span><span class="sc">:</span>n_cafes) <span class="sc">|&gt;</span> </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(<span class="at">visit =</span> <span class="dv">1</span><span class="sc">:</span>n_visits) <span class="sc">|&gt;</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">afternoon =</span> <span class="fu">rep</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">1</span>, <span class="at">times =</span> <span class="fu">n</span>() <span class="sc">/</span> <span class="dv">2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mu =</span> a_cafe <span class="sc">+</span> b_cafe <span class="sc">*</span> afternoon) <span class="sc">|&gt;</span> </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">wait =</span> <span class="fu">rnorm</span>(<span class="at">n =</span> <span class="fu">n</span>(), <span class="at">mean =</span> mu, <span class="at">sd =</span> sigma)) <span class="sc">|&gt;</span> </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cafe, <span class="fu">everything</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We might peek at the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>d <span class="sc">|&gt;</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 200
Columns: 7
$ cafe      &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4,…
$ a_cafe    &lt;dbl&gt; 4.223962, 4.223962, 4.223962, 4.223962, 4.223962, 4.223962, 4.223962, 4.223962, 4.223962, 4.223962, 2.010498, …
$ b_cafe    &lt;dbl&gt; -1.6093565, -1.6093565, -1.6093565, -1.6093565, -1.6093565, -1.6093565, -1.6093565, -1.6093565, -1.6093565, -1…
$ visit     &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2, 3, 4, 5, 6,…
$ afternoon &lt;int&gt; 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,…
$ mu        &lt;dbl&gt; 4.223962, 2.614606, 4.223962, 2.614606, 4.223962, 2.614606, 4.223962, 2.614606, 4.223962, 2.614606, 2.010498, …
$ wait      &lt;dbl&gt; 3.9678929, 3.8571978, 4.7278755, 2.7610133, 4.1194827, 3.5436522, 4.1909492, 2.5332235, 4.1240321, 2.7648868, …</code></pre>
</div>
</div>
<p>Now we’ve finally simulated our data, we are ready to make our version of Figure 14.1, from way back on page 436.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>d <span class="sc">|&gt;</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">afternoon =</span> <span class="fu">ifelse</span>(afternoon <span class="sc">==</span> <span class="dv">0</span>, <span class="st">"M"</span>, <span class="st">"A"</span>),</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">day       =</span> <span class="fu">rep</span>(<span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="at">each =</span> <span class="dv">2</span>), <span class="at">times =</span> n_cafes)) <span class="sc">|&gt;</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cafe <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">5</span>)) <span class="sc">|&gt;</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cafe =</span> <span class="fu">str_c</span>(<span class="st">"café #"</span>, cafe)) <span class="sc">|&gt;</span> </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> visit, <span class="at">y =</span> wait, <span class="at">group =</span> day)) <span class="sc">+</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> afternoon), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">labels =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"A"</span>), <span class="at">times =</span> <span class="dv">5</span>)) <span class="sc">+</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="st">"wait time in minutes"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#EEDA9D"</span>)) <span class="sc">+</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> cafe, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_earring</span>(<span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>                      <span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="336"></p>
</figure>
</div>
</div>
</div>
<section id="rethinking-simulation-and-misspecification" class="level4" data-number="14.1.2.1">
<h4 data-number="14.1.2.1" class="anchored" data-anchor-id="rethinking-simulation-and-misspecification"><span class="header-section-number">14.1.2.1</span> Rethinking: Simulation and misspecification</h4>
<blockquote class="blockquote">
<p>In this exercise, we are simulating data from a generative process and then analyzing that data with a model that reflects exactly the correct structure of that process. But in the real world, we’re never so lucky. Instead we are always forced to analyze data with a model that is misspecified: The true data-generating process is different than the model. Simulation can be used however to explore misspecification. Just simulate data from a process and then see how a number of models, none of which match exactly the data-generating process, perform. And always remember that Bayesian inference does not depend upon data-generating assumptions, such as the likelihood, being true. (p.&nbsp;441)</p>
</blockquote>
</section>
</section>
<section id="the-varying-slopes-model" class="level3" data-number="14.1.3">
<h3 data-number="14.1.3" class="anchored" data-anchor-id="the-varying-slopes-model"><span class="header-section-number">14.1.3</span> The varying slopes model</h3>
<p>The statistical formula for our varying intercepts and slopes café model follows the form</p>
<p><span class="math display">\[\begin{align*}
\text{wait}_i &amp; \sim \operatorname{Normal}(\mu_i, \sigma) \\
\mu_i         &amp; = \alpha_{\text{café}[i]} + \beta_{\text{café}[i]} \text{afternoon}_i \\
\begin{bmatrix} \alpha_\text{café} \\ \beta_\text{café} \end{bmatrix} &amp; \sim \operatorname{MVNormal} \begin{pmatrix} \begin{bmatrix} \alpha \\ \beta \end{bmatrix}, \mathbf \Sigma \end{pmatrix} \\
\mathbf \Sigma     &amp; = \begin{bmatrix} \sigma_\alpha &amp; 0 \\ 0 &amp; \sigma_\beta \end{bmatrix} \mathbf R \begin{bmatrix} \sigma_\alpha &amp; 0 \\ 0 &amp; \sigma_\beta \end{bmatrix} \\
\alpha        &amp; \sim \operatorname{Normal}(5, 2) \\
\beta         &amp; \sim \operatorname{Normal}(-1, 0.5) \\
\sigma        &amp; \sim \operatorname{Exponential}(1) \\
\sigma_\alpha &amp; \sim \operatorname{Exponential}(1) \\
\sigma_\beta  &amp; \sim \operatorname{Exponential}(1) \\
\mathbf R     &amp; \sim \operatorname{LKJcorr}(2),
\end{align*}\]</span></p>
<p>where <span class="math inline">\(\mathbf \Sigma\)</span> is the covariance matrix and <span class="math inline">\(\mathbf R\)</span> is the corresponding correlation matrix, which we might more fully express as</p>
<p><span class="math display">\[\mathbf R = \begin{bmatrix} 1 &amp; \rho \\ \rho &amp; 1 \end{bmatrix}.\]</span></p>
<p>And according to our prior, <span class="math inline">\(\mathbf R\)</span> is distributed as <span class="math inline">\(\operatorname{LKJcorr}(2)\)</span>. We’ll use <code>rethinking::rlkjcorr()</code> to get a better sense of what that even is.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rethinking)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>n_sim <span class="ot">&lt;-</span> <span class="fl">1e4</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>r_1 <span class="ot">&lt;-</span> <span class="fu">rlkjcorr</span>(n_sim, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">eta =</span> <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>r_2 <span class="ot">&lt;-</span> <span class="fu">rlkjcorr</span>(n_sim, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">eta =</span> <span class="dv">2</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>r_4 <span class="ot">&lt;-</span> <span class="fu">rlkjcorr</span>(n_sim, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">eta =</span> <span class="dv">4</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here are the <span class="math inline">\(\text{LKJcorr}\)</span> distributions of Figure 14.3.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For annotation</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>text <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x     =</span> <span class="fu">c</span>(<span class="fl">0.83</span>, <span class="fl">0.625</span>, <span class="fl">0.45</span>),</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y     =</span> <span class="fu">c</span>(<span class="fl">0.56</span>, <span class="fl">0.750</span>, <span class="fl">1.07</span>),</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="fu">c</span>(<span class="st">"eta = 1"</span>, <span class="st">"eta = 2"</span>, <span class="st">"eta = 4"</span>))</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>r_1 <span class="sc">|&gt;</span> </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> X2)) <span class="sc">+</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">adjust =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">alpha =</span> <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"transparent"</span>, <span class="at">fill =</span> <span class="st">"#394165"</span>) <span class="sc">+</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">data =</span> r_2,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>               <span class="at">adjust =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">alpha =</span> <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"transparent"</span>, <span class="at">fill =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">data =</span> r_4,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>               <span class="at">adjust =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">alpha =</span> <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"transparent"</span>, <span class="at">fill =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> text,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">label =</span> label),</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#A65141"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"correlation"</span>,</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="fu">expression</span>(<span class="fu">LKJcorr</span>(eta)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<p>As it turns out, the shape of the LKJ is sensitive to both <span class="math inline">\(\eta\)</span> and the <span class="math inline">\(K\)</span> dimensions of the correlation matrix. Our simulations only considered the shapes for when <span class="math inline">\(K = 2\)</span>. We can use a combination of the <code>parse_dist()</code> and <code>stat_dist_halfeye()</code> functions from the <strong>tidybayes</strong> package to derive analytic solutions for different combinations of <span class="math inline">\(\eta\)</span> and <span class="math inline">\(K\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidybayes)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">crossing</span>(<span class="at">k   =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">eta =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) <span class="sc">|&gt;</span> </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prior =</span> <span class="fu">str_c</span>(<span class="st">"lkjcorr_marginal("</span>, k, <span class="st">", "</span>, eta, <span class="st">")"</span>),</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">strip =</span> <span class="fu">str_c</span>(<span class="st">"K=="</span>, k)) <span class="sc">|&gt;</span> </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">parse_dist</span>(prior) <span class="sc">|&gt;</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> eta, <span class="at">dist =</span> .dist, <span class="at">args =</span> .args)) <span class="sc">+</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_dist_halfeye</span>(<span class="at">.width =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.95</span>),</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">fill =</span> <span class="st">"#A65141"</span>) <span class="sc">+</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="fu">expression</span>(rho), <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>),</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>                     <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"-1"</span>, <span class="st">"-.5"</span>, <span class="st">"0"</span>, <span class="st">".5"</span>, <span class="st">"1"</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="fu">expression</span>(eta), <span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">expression</span>(<span class="st">"Marginal correlation for the LKJ prior relative to K and "</span><span class="sc">*</span>eta)) <span class="sc">+</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> strip, <span class="at">labeller =</span> label_parsed, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>To learn more about this plotting method, check out Kay’s <span class="citation" data-cites="kayMarginalDistributionSingle2020">(<a href="references.html#ref-kayMarginalDistributionSingle2020" role="doc-biblioref">2020</a>)</span> <a href="https://mjskay.github.io/ggdist/reference/lkjcorr_marginal.html"><em>Marginal distribution of a single correlation from an LKJ distribution</em></a>. To get a better intuition what that plot means, check out the illuminating blog post by <a href="http://srmart.in/">Stephen Martin</a>, <a href="http://srmart.in/is-the-lkj1-prior-uniform-yes/"><em>Is the LKJ(1) prior uniform? “Yes”</em></a>.</p>
<p>Okay, let’s get ready to model and switch out <strong>rethinking</strong> for <strong>brms</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">detach</span>(package<span class="sc">:</span>rethinking, <span class="at">unload =</span> T)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(brms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As defined above, our first model has both varying intercepts and <code>afternoon</code> slopes. I should point out that the <code>(1 + afternoon | cafe)</code> syntax specifies that we’d like <code>brm()</code> to fit the random effects for <code>1</code> (i.e., the intercept) and the <code>afternoon</code> slope as correlated. Had we wanted to fit a model in which they were orthogonal, we’d have coded <code>(1 + afternoon || cafe)</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d, </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  wait <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> afternoon <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> afternoon <span class="sc">|</span> cafe),</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">5</span>, <span class="dv">2</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b),</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd),</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma),</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor)),</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">867530</span>,</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.01"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With Figure 14.4, we assess how the posterior for the correlation of the random effects compares to its prior.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>post <span class="ot">&lt;-</span> <span class="fu">as_draws_df</span>(b14<span class="fl">.1</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>post <span class="sc">|&gt;</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">data =</span> r_2, <span class="fu">aes</span>(<span class="at">x =</span> X2),</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">color =</span> <span class="st">"transparent"</span>, <span class="at">fill =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> cor_cafe__Intercept__afternoon),</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="dv">9</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">color =</span> <span class="st">"transparent"</span>, <span class="at">fill =</span> <span class="st">"#A65141"</span>) <span class="sc">+</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">"text"</span>, </span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.15</span>, <span class="dv">0</span>), <span class="at">y =</span> <span class="fu">c</span>(<span class="fl">2.21</span>, <span class="fl">0.85</span>), </span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">label =</span> <span class="fu">c</span>(<span class="st">"posterior"</span>, <span class="st">"prior"</span>), </span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="fu">c</span>(<span class="st">"#A65141"</span>, <span class="st">"#EEDA9D"</span>), <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"correlation"</span>,</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Correlation between intercepts</span><span class="sc">\n</span><span class="st">and slopes, prior and posterior"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<p>McElreath then depicted multidimensional shrinkage by plotting the posterior mean of the varying effects compared to their raw, unpooled estimated. With <strong>brms</strong>, we can get the <code>cafe</code>-specific intercepts and <code>afternoon</code> slopes with <code>coef()</code>, which returns a three-dimensional list.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># coef(b14.1) |&gt; glimpse()</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">coef</span>(b14<span class="fl">.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$cafe
, , Intercept

   Estimate Est.Error     Q2.5    Q97.5
1  4.213297 0.1978596 3.829642 4.590972
2  2.157338 0.2013847 1.769423 2.549774
3  4.378303 0.2049147 3.978473 4.786580
4  3.243736 0.1985901 2.843128 3.621809
5  1.878760 0.2058461 1.473869 2.280121
6  4.263000 0.1978386 3.858712 4.639112
7  3.615683 0.1994118 3.222665 4.007438
8  3.947060 0.1967307 3.561294 4.339305
9  3.979611 0.1976517 3.597588 4.364261
10 3.558939 0.2000770 3.172513 3.955428
11 1.932402 0.2053137 1.528221 2.331512
12 3.836842 0.2022665 3.429331 4.222623
13 3.891585 0.2014402 3.506540 4.280477
14 3.176418 0.1942546 2.791193 3.557988
15 4.450167 0.2060186 4.037785 4.851168
16 3.390435 0.2063019 2.981886 3.792157
17 4.216905 0.2013102 3.827670 4.611066
18 5.745353 0.2054073 5.337968 6.155167
19 3.244080 0.2066967 2.842267 3.636957
20 3.739302 0.1970768 3.360855 4.137549

, , afternoon

     Estimate Est.Error       Q2.5      Q97.5
1  -1.1555017 0.2568967 -1.6603627 -0.6388167
2  -0.9014608 0.2655887 -1.4199688 -0.3626409
3  -1.9416692 0.2732649 -2.4709132 -1.4135964
4  -1.2340735 0.2579623 -1.7488369 -0.7370762
5  -0.1342429 0.2704266 -0.6536181  0.4017659
6  -1.3025528 0.2560035 -1.8104277 -0.8086929
7  -1.0223715 0.2641168 -1.5501771 -0.5117426
8  -1.6305479 0.2662107 -2.1559578 -1.1222900
9  -1.2991796 0.2632013 -1.8216574 -0.7910278
10 -0.9515703 0.2630617 -1.4607206 -0.4360276
11 -0.4279579 0.2753179 -0.9847493  0.1043620
12 -1.1761338 0.2669048 -1.6873338 -0.6666824
13 -1.8149738 0.2710414 -2.3536671 -1.2954402
14 -0.9432858 0.2579568 -1.4528844 -0.4381805
15 -2.1901761 0.2855867 -2.7617111 -1.6327983
16 -1.0467058 0.2693629 -1.5738346 -0.5120496
17 -1.2246217 0.2678882 -1.7545433 -0.6888909
18 -1.0276014 0.2841628 -1.5786265 -0.4645198
19 -0.2607186 0.2790455 -0.8029477  0.2769846
20 -1.0669509 0.2613099 -1.5783545 -0.5341076</code></pre>
</div>
</div>
<p>Here’s the code to extract the relevant elements from the <code>coef()</code> list, convert them to a tibble, and add the <code>cafe</code> index.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># With this line we select each of the 20 cafe's posterior mean (i.e., Estimate)</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># for both `Intercept` and `afternoon`</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>partially_pooled_params <span class="ot">&lt;-</span> <span class="fu">coef</span>(b14<span class="fl">.1</span>)<span class="sc">$</span>cafe[ , <span class="dv">1</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span>                 <span class="co"># Convert the two vectors to a data frame</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">Slope =</span> afternoon) <span class="sc">|&gt;</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cafe =</span> <span class="fu">row_number</span>()) <span class="sc">|&gt;</span>  <span class="co"># Add the `cafe` index</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cafe, <span class="fu">everything</span>())      <span class="co"># Simply moving `cafe` to the leftmost position</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Like McElreath, we’ll compute the unpooled estimates directly from the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute unpooled estimates directly from data</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>un_pooled_params <span class="ot">&lt;-</span> d <span class="sc">|&gt;</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># With these two lines, we compute the mean value for each cafe's </span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># wait time in the morning and then the afternoon</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(afternoon, cafe) <span class="sc">|&gt;</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean =</span> <span class="fu">mean</span>(wait)) <span class="sc">|&gt;</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">|&gt;</span>  <span class="co"># Ungrouping lets us to alter the grouping variable `afternoon`</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">afternoon =</span> <span class="fu">ifelse</span>(afternoon <span class="sc">==</span> <span class="dv">0</span>, <span class="st">"Intercept"</span>, <span class="st">"Slope"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">spread</span>(<span class="at">key =</span> afternoon, <span class="at">value =</span> mean) <span class="sc">|&gt;</span>  <span class="co"># Use `spread()` just like above</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Slope =</span> Slope <span class="sc">-</span> Intercept)         <span class="co"># Finally, here's our slope!</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we combine the partially-pooled and unpooled means into a single </span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a><span class="co"># data object, which will make plotting easier.</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="co"># `bind_rows()` will stack the second tibble below the first</span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>params <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(partially_pooled_params, un_pooled_params) <span class="sc">|&gt;</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Index whether the estimates are pooled</span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">pooled =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"partially"</span>, <span class="st">"not"</span>), <span class="at">each =</span> <span class="fu">n</span>() <span class="sc">/</span> <span class="dv">2</span>)) </span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Here's a glimpse at what we've been working for</span></span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>params <span class="sc">|&gt;</span></span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">36</span><span class="sc">:</span><span class="dv">40</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      cafe Intercept       Slope    pooled
1        1  4.213297 -1.15550174 partially
2        2  2.157338 -0.90146084 partially
3        3  4.378303 -1.94166924 partially
4        4  3.243736 -1.23407346 partially
5        5  1.878760 -0.13424290 partially
...6    16  3.373496 -1.02563866       not
...7    17  4.236192 -1.22236910       not
...8    18  5.755987 -0.87660383       not
...9    19  3.121060  0.01441784       not
...10   20  3.728481 -1.03811567       not</code></pre>
</div>
</div>
<p>Finally, here’s our code for Figure 14.5.a, showing shrinkage in two dimensions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> params, <span class="fu">aes</span>(<span class="at">x =</span> Intercept, <span class="at">y =</span> Slope)) <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">2</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">4</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">5</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">6</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">7</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">8</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="dv">9</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>, <span class="at">level =</span> <span class="fl">0.99</span>, <span class="at">linewidth =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">group =</span> cafe, <span class="at">color =</span> pooled)) <span class="sc">+</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> cafe), <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="st">"Pooled?"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">range</span>(params<span class="sc">$</span>Intercept),</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">range</span>(params<span class="sc">$</span>Slope))</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>p1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="432"></p>
</figure>
</div>
</div>
</div>
<p>Learn more about <code>stat_ellipse()</code>, <a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html">here</a>. Let’s prep for Figure 14.5.b.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Retrieve the partially-pooled estimates with `coef()`</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>partially_pooled_estimates <span class="ot">&lt;-</span> <span class="fu">coef</span>(b14<span class="fl">.1</span>)<span class="sc">$</span>cafe[ , <span class="dv">1</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Convert the two vectors to a data frame</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The Intercept is the wait time for morning (i.e., `afternoon == 0`)</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">morning =</span> Intercept) <span class="sc">|&gt;</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># `afternoon` wait time is the `morning` wait time plus the afternoon slope</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">afternoon =</span> morning <span class="sc">+</span> afternoon,</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">cafe      =</span> <span class="fu">row_number</span>()) <span class="sc">|&gt;</span>  <span class="co"># Add the `cafe` index</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cafe, <span class="fu">everything</span>()) </span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute unpooled estimates directly from data</span></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>un_pooled_estimates <span class="ot">&lt;-</span> d <span class="sc">|&gt;</span></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># As above, these two lines compute each cafe's mean wait value by time of day</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(afternoon, cafe) <span class="sc">|&gt;</span> </span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean =</span> <span class="fu">mean</span>(wait)) <span class="sc">|&gt;</span></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ungrouping lets us alter the grouping variable, `afternoon`</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">|&gt;</span> </span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">afternoon =</span> <span class="fu">ifelse</span>(afternoon <span class="sc">==</span> <span class="dv">0</span>, <span class="st">"morning"</span>, <span class="st">"afternoon"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This separates out the values into `morning` and `afternoon` columns</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">spread</span>(<span class="at">key =</span> afternoon, <span class="at">value =</span> mean)</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>estimates <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(partially_pooled_estimates, un_pooled_estimates) <span class="sc">|&gt;</span></span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">pooled =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"partially"</span>, <span class="st">"not"</span>), <span class="at">each =</span> <span class="fu">n</span>() <span class="sc">/</span> <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The code for Figure 14.5.b.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> estimates, <span class="fu">aes</span>(<span class="at">x =</span> morning, <span class="at">y =</span> afternoon)) <span class="sc">+</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Nesting `stat_ellipse()` within `mapply()` is a less redundant way to produce the </span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ten-layered semitransparent ellipses we did with ten lines of `stat_ellipse()` </span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># functions in the previous plot</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mapply</span>(<span class="cf">function</span>(level) {</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_ellipse</span>(<span class="at">geom  =</span> <span class="st">"polygon"</span>, <span class="at">type =</span> <span class="st">"norm"</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">level =</span> level, <span class="at">linewidth =</span> <span class="dv">0</span>)</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    }, </span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Enter the levels here</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">level =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span> <span class="sc">/</span> <span class="dv">10</span>, <span class="fl">0.99</span>)) <span class="sc">+</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">group =</span> cafe, <span class="at">color =</span> pooled)) <span class="sc">+</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> cafe), <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="st">"Pooled?"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"morning wait (mins)"</span>,</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"afternoon wait (mins)"</span>) <span class="sc">+</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">range</span>(estimates<span class="sc">$</span>morning),</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">range</span>(estimates<span class="sc">$</span>afternoon))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we combine the two subplots together with <strong>patchwork</strong> syntax.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>(p1 <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)) <span class="sc">+</span> </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  p2 <span class="sc">+</span> </span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Shrinkage in two dimensions"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>What I want you to appreciate in this plot is that shrinkage on the parameter scale naturally produces shrinkage where we actually care about it: on the outcome scale. And it also implies a population of wait times, shown by the [semitransparent ellipses]. That population is now positively correlated–cafés with longer morning waits also tend to have longer afternoon waits. They are popular, after all. But the population lies mostly below the dashed line where the waits are equal. You’ll wait less in the afternoon, on average. (p.&nbsp;446)</p>
</blockquote>
</section>
</section>
<section id="advanced-varying-slopes" class="level2" data-number="14.2">
<h2 data-number="14.2" class="anchored" data-anchor-id="advanced-varying-slopes"><span class="header-section-number">14.2</span> Advanced varying slopes</h2>
<p>In <a href="13.html#sec-More-than-one-type-of-cluster" class="quarto-xref"><span>Section 13.3</span></a> we saw that data can be considered <strong>cross-classified</strong> if they have multiple grouping factors. We used the <code>chipanzees</code> data in that section and we only considered cross-classification by single intercepts. Turns out cross-classified models can be extended further. Let’s load and wrangle those data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(chimpanzees, <span class="at">package =</span> <span class="st">"rethinking"</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> chimpanzees</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(chimpanzees)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Wrangle</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> d <span class="sc">|&gt;</span> </span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">actor     =</span> <span class="fu">factor</span>(actor),</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">block     =</span> <span class="fu">factor</span>(block),</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">treatment =</span> <span class="fu">factor</span>(<span class="dv">1</span> <span class="sc">+</span> prosoc_left <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> condition),</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>         <span class="co"># This will come in handy, later</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">labels    =</span> <span class="fu">factor</span>(treatment,</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>                            <span class="at">levels =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>                            <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"r/n"</span>, <span class="st">"l/n"</span>, <span class="st">"r/p"</span>, <span class="st">"l/p"</span>)))</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 504
Columns: 10
$ actor        &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ recipient    &lt;int&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ condition    &lt;int&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
$ block        &lt;fct&gt; 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6,…
$ trial        &lt;int&gt; 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56,…
$ prosoc_left  &lt;int&gt; 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0,…
$ chose_prosoc &lt;int&gt; 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1,…
$ pulled_left  &lt;int&gt; 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0,…
$ treatment    &lt;fct&gt; 1, 1, 2, 1, 2, 2, 2, 2, 1, 1, 1, 2, 1, 2, 1, 2, 2, 1, 2, 1, 1, 1, 2, 2, 1, 1, 2, 2, 1, 2, 1, 2, 2, 2, 1, 1,…
$ labels       &lt;fct&gt; r/n, r/n, l/n, r/n, l/n, l/n, l/n, l/n, r/n, r/n, r/n, l/n, r/n, l/n, r/n, l/n, l/n, r/n, l/n, r/n, r/n, r/…</code></pre>
</div>
</div>
<p>If I’m following along correctly with the text, McElreath’s <code>m14.2</code> uses the centered parameterization. Recall from the last chapter that <strong>brms</strong> only supports the non-centered parameterization. Happily, McElreath’s <code>m14.3</code> appears to use the non-centered parameterization. Thus, we’ll skip making a <code>b14.2</code> and jump directly into making a <code>b14.3</code>. I believe one could describe the statistical model as</p>
<p><span class="math display">\[\begin{align*}
\text{left\_pull}_i &amp; \sim \operatorname{Binomial}(n_i = 1, p_i) \\
\operatorname{logit} (p_i) &amp; = \gamma_{\text{treatment}[i]} + \alpha_{\text{actor}[i], \text{treatment}[i]} + \beta_{\text{block}[i], \text{treatment}[i]} \\
\gamma_j &amp; \sim \operatorname{Normal}(0, 1), \;\;\; \text{for } j = 1, \dots, 4 \\
\begin{bmatrix} \alpha_{j, 1} \\ \alpha_{j, 2} \\ \alpha_{j, 3} \\ \alpha_{j, 4} \end{bmatrix} &amp; \sim \operatorname{MVNormal} \begin{pmatrix} \begin{bmatrix} 0 \\ 0 \\ 0 \\ 0 \end{bmatrix}, \mathbf \Sigma_\text{actor} \end{pmatrix} \\
\begin{bmatrix} \beta_{j, 1} \\ \beta_{j, 2} \\ \beta_{j, 3} \\ \beta_{j, 4} \end{bmatrix} &amp; \sim \operatorname{MVNormal} \begin{pmatrix} \begin{bmatrix} 0 \\ 0 \\ 0 \\ 0 \end{bmatrix}, \mathbf \Sigma_\text{block} \end{pmatrix} \\
\mathbf \Sigma_\text{actor} &amp; = \mathbf{S_\alpha R_\alpha S_\alpha} \\
\mathbf \Sigma_\text{block} &amp; = \mathbf{S_\beta R_\beta S_\beta} \\
\sigma_{\alpha, [1]}, \dots, \sigma_{\alpha, [4]} &amp; \sim \operatorname{Exponential}(1) \\
\sigma_{\beta, [1]}, \dots, \sigma_{\beta, [4]}   &amp; \sim \operatorname{Exponential}(1) \\
\mathbf R_\alpha &amp; \sim \operatorname{LKJ}(2) \\
\mathbf R_\beta  &amp; \sim \operatorname{LKJ}(2).
\end{align*}\]</span></p>
<p>In this model, we have four population-level intercepts, <span class="math inline">\(\gamma_1, \dots, \gamma_4\)</span>, one for each of the four levels of <code>treatment</code>. There are two higher-level grouping variables, <code>actor</code> and <code>block</code>, making this a cross-classified model.</p>
<p>The term <span class="math inline">\(\alpha_{\text{actor}[i], \text{treatment}[i]}\)</span> is meant to convey that each of the <code>treatment</code> effects can vary by <code>actor</code>. The first line containing the <span class="math inline">\(\operatorname{MVNormal}(\cdot)\)</span> operator indicates the <code>actor</code>-level deviations from the population-level estimates for <span class="math inline">\(\gamma_j\)</span> follow the multivariate normal distribution where the four means are set to zero (i.e., they are deviations) and their spread around those zeros are controlled by <span class="math inline">\(\Sigma_\text{actor}\)</span>. In the first line below the last line containing <span class="math inline">\(\operatorname{MVNormal}(\cdot)\)</span>, we learn that <span class="math inline">\(\Sigma_\text{actor}\)</span> can be decomposed into two terms, <span class="math inline">\(\mathbf S_\alpha\)</span> and <span class="math inline">\(\mathbf R_\alpha\)</span>. It may not yet be clear by the notation, but <span class="math inline">\(\mathbf S_\alpha\)</span> is a <span class="math inline">\(4 \times 4\)</span> matrix,</p>
<p><span class="math display">\[
\mathbf S_\alpha = \begin{bmatrix} \sigma_{\alpha, [1]} &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; \sigma_{\alpha, [2]} &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; \sigma_{\alpha, [3]} &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; \sigma_{\alpha, [4]} \end{bmatrix}.
\]</span></p>
<p>In a similar way, <span class="math inline">\(\mathbf R_\alpha\)</span> is a <span class="math inline">\(4 \times 4\)</span> matrix,</p>
<p><span class="math display">\[
\mathbf R_\alpha = \begin{bmatrix} 1 &amp; \rho_{\alpha, [1, 2]} &amp; \rho_{\alpha, [1, 3]} &amp; \rho_{\alpha, [1, 4]} \\ \rho_{\alpha, [2, 1]} &amp; 1 &amp; \rho_{\alpha, [2, 3]} &amp; \rho_{\alpha, [2, 4]} \\ \rho_{\alpha, [3, 1]} &amp; \rho_{\alpha, [3, 2]} &amp; 1 &amp; \rho_{\alpha, [3, 4]} \\ \rho_{\alpha, [4, 1]} &amp; \rho_{\alpha, [4, 2]} &amp; \rho_{\alpha, [4, 3]} &amp; 1 \end{bmatrix}.
\]</span></p>
<p>The same overall pattern holds true for <span class="math inline">\(\beta_{\text{block}[i], \text{treatment}[i]}\)</span> and the associated <span class="math inline">\(\beta\)</span> parameters connected to the <code>block</code> grouping variable. All the population parameters <span class="math inline">\(\sigma_{\alpha, [1]}, \dots, \sigma_{\alpha, [4]}\)</span> and <span class="math inline">\(\sigma_{\beta, [1]}, \dots, \sigma_{\beta, [4]}\)</span> have individual <span class="math inline">\(\operatorname{Exponential}(1)\)</span> priors. The two <span class="math inline">\(\mathbf R_{&lt; \cdot &gt;}\)</span> matrices have the priors <span class="math inline">\(\operatorname{LKJ}(2)\)</span>.</p>
<p>I know; this is a lot. This all takes time to grapple with. Here’s how to fit such a model with <strong>brms</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.3</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d, </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> binomial,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  pulled_left <span class="sc">|</span> <span class="fu">trials</span>(<span class="dv">1</span>) <span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> treatment <span class="sc">+</span> (<span class="dv">0</span> <span class="sc">+</span> treatment <span class="sc">|</span> actor) <span class="sc">+</span> (<span class="dv">0</span> <span class="sc">+</span> treatment <span class="sc">|</span> block),</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b),</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">group =</span> actor),</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">group =</span> block),</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor, <span class="at">group =</span> actor),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor, <span class="at">group =</span> block)),</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,  </span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">4387510</span>,</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.03"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Happily, we got no warnings about divergent transitions. Since it’s been a while, we’ll use <code>bayesplot::mcmc_rank_overlay()</code> to examine the primary parameters with a trank plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bayesplot)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Give the parameters fancy names</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_c</span>(<span class="st">"treatment["</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="st">"]"</span>), </span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_c</span>(<span class="st">"sigma['actor["</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="st">"]']"</span>), </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_c</span>(<span class="st">"sigma['block["</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="st">"]']"</span>), </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_c</span>(<span class="st">"rho['actor:treatment["</span>, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>), <span class="st">","</span>, <span class="fu">rep</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">times =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>), <span class="st">"]']"</span>), </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_c</span>(<span class="st">"rho['block:treatment["</span>, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>), <span class="st">","</span>, <span class="fu">rep</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">times =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>), <span class="st">"]']"</span>), </span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chain"</span>)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Wrangle</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="fu">as_draws_df</span>(b14<span class="fl">.3</span>) <span class="sc">|&gt;</span> </span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(b_treatment1<span class="sc">:</span><span class="st">`</span><span class="at">cor_block__treatment3__treatment4</span><span class="st">`</span>, .chain) <span class="sc">|&gt;</span> </span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_names</span>(names) <span class="sc">|&gt;</span> </span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot</span></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mcmc_rank_overlay</span>() <span class="sc">+</span></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">4</span> <span class="sc">*</span> <span class="fl">1e3</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">str_c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="st">"K"</span>))) <span class="sc">+</span></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#B1934A"</span>, <span class="st">"#A65141"</span>, <span class="st">"#EEDA9D"</span>)) <span class="sc">+</span></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"McElreath would love this trank plot."</span>) <span class="sc">+</span></span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> parameter, <span class="at">labeller =</span> label_parsed, <span class="at">ncol =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>Because we only fit a non-centered version of the model, we aren’t able to make a faithful version of McElreath’s Figure 14.6. However, we can still use <code>posterior::summarise_draws()</code> to help make histograms of the two kinds of effective sample sizes for our <code>b14.3</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(posterior)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as_draws_df</span>(b14<span class="fl">.3</span>) <span class="sc">|&gt;</span> </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise_draws</span>() <span class="sc">|&gt;</span> </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">starts_with</span>(<span class="st">"ess"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> value)) <span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">250</span>, <span class="at">color =</span> <span class="st">"#DCA258"</span>, <span class="at">fill =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>Here is a summary of the model parameters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: binomial 
  Links: mu = logit 
Formula: pulled_left | trials(1) ~ 0 + treatment + (0 + treatment | actor) + (0 + treatment | block) 
   Data: d (Number of observations: 504) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Multilevel Hyperparameters:
~actor (Number of levels: 7) 
                           Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sd(treatment1)                 1.39      0.50     0.69     2.58 1.00     2163     2656
sd(treatment2)                 0.91      0.40     0.34     1.91 1.00     2580     2609
sd(treatment3)                 1.83      0.57     0.99     3.25 1.00     2820     2923
sd(treatment4)                 1.58      0.61     0.75     3.07 1.00     3035     2829
cor(treatment1,treatment2)     0.42      0.29    -0.20     0.87 1.00     2540     2791
cor(treatment1,treatment3)     0.53      0.25    -0.06     0.90 1.00     2558     3083
cor(treatment2,treatment3)     0.48      0.27    -0.11     0.89 1.00     2732     2900
cor(treatment1,treatment4)     0.44      0.27    -0.16     0.87 1.00     2757     2733
cor(treatment2,treatment4)     0.44      0.28    -0.18     0.88 1.00     3169     3223
cor(treatment3,treatment4)     0.57      0.24     0.00     0.92 1.00     3078     3302

~block (Number of levels: 6) 
                           Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sd(treatment1)                 0.40      0.33     0.02     1.24 1.00     2141     2318
sd(treatment2)                 0.46      0.35     0.02     1.34 1.00     1724     2046
sd(treatment3)                 0.30      0.26     0.01     0.97 1.00     3124     2322
sd(treatment4)                 0.47      0.37     0.02     1.39 1.00     2078     2544
cor(treatment1,treatment2)    -0.08      0.37    -0.74     0.62 1.00     3633     3014
cor(treatment1,treatment3)    -0.01      0.38    -0.71     0.70 1.00     6210     3184
cor(treatment2,treatment3)    -0.03      0.38    -0.71     0.70 1.00     4125     2916
cor(treatment1,treatment4)     0.06      0.37    -0.65     0.74 1.00     4038     2875
cor(treatment2,treatment4)     0.05      0.37    -0.65     0.72 1.00     3795     3075
cor(treatment3,treatment4)     0.02      0.38    -0.70     0.71 1.00     3686     3623

Regression Coefficients:
           Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
treatment1     0.22      0.50    -0.75     1.23 1.00     1848     2448
treatment2     0.65      0.42    -0.21     1.48 1.00     1985     2617
treatment3    -0.02      0.58    -1.16     1.14 1.00     2720     2621
treatment4     0.68      0.53    -0.38     1.75 1.00     2733     2936

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Like McElreath explained on page 450, our <code>b14.3</code> has 76 parameters:</p>
<ul>
<li>4 average <code>treatment</code> effects, as listed in the ‘Population-Level Effects’ section;</li>
<li>7 <span class="math inline">\(\times\)</span> 4 = 28 varying effects on <code>actor</code>, as indicated in the ‘~actor:treatment (Number of levels: 7)’ header multiplied by the four levels of <code>treatment</code>;</li>
<li>6 <span class="math inline">\(\times\)</span> 4 = 24 varying effects on <code>block</code>, as indicated in the ‘~block:treatment (Number of levels: 6)’ header multiplied by the four levels of <code>treatment</code>;</li>
<li>8 standard deviations listed in the eight rows beginning with <code>sd(</code>; and</li>
<li>12 free correlation parameters listed in the eight rows beginning with <code>cor(</code>.</li>
</ul>
<p>Compute the WAIC estimate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.3</span> <span class="ot">&lt;-</span> <span class="fu">add_criterion</span>(b14<span class="fl">.3</span>, <span class="at">criterion =</span> <span class="st">"waic"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">waic</span>(b14<span class="fl">.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Computed from 4000 by 504 log-likelihood matrix.

          Estimate   SE
elpd_waic   -272.8  9.9
p_waic        27.2  1.5
waic         545.5 19.8

1 (0.2%) p_waic estimates greater than 0.4. We recommend trying loo instead. </code></pre>
</div>
</div>
<p>Like the <span class="math inline">\(p_\text{WAIC}\)</span>, our <strong>brms</strong> version of the model has about 27 effective parameters. Now we’ll get a better sense of the model with a posterior predictive check in the form of our version of Figure 14.7. McElreath described his <strong>R</strong> code 14.22 as “a big chunk of code” (p.&nbsp;451). I’ll leave up to the reader to decide whether our big code chunk is any better.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For annotation</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>text <span class="ot">&lt;-</span> <span class="fu">distinct</span>(d, labels) <span class="sc">|&gt;</span> </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">actor =</span> <span class="dv">1</span>,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">prop  =</span> <span class="fu">c</span>(<span class="fl">0.07</span>, <span class="fl">0.8</span>, <span class="fl">0.08</span>, <span class="fl">0.795</span>))</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>nd <span class="ot">&lt;-</span> d <span class="sc">|&gt;</span> </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(actor, condition, labels, prosoc_left, treatment) <span class="sc">|&gt;</span> </span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">block =</span> <span class="dv">5</span>)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute and wrangle the posterior predictions</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="fu">fitted</span>(b14<span class="fl">.3</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(nd) <span class="sc">|&gt;</span> </span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add the empirical proportions</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>    d <span class="sc">|&gt;</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">group_by</span>(actor, treatment) <span class="sc">|&gt;</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">proportion =</span> <span class="fu">mean</span>(pulled_left)) <span class="sc">|&gt;</span> </span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">distinct</span>(actor, treatment, proportion),</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"actor"</span>, <span class="st">"treatment"</span>)</span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">condition =</span> <span class="fu">factor</span>(condition)) <span class="sc">|&gt;</span> </span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot!</span></span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> labels)) <span class="sc">+</span></span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#E8DCCF"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Empirical proportions</span></span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> proportion, <span class="at">group =</span> prosoc_left),</span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#394165"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> proportion, <span class="at">shape =</span> condition),</span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#394165"</span>, <span class="at">fill =</span> <span class="st">"#100F14"</span>, <span class="at">size =</span> <span class="fl">2.5</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span> </span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Posterior predictions</span></span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> Estimate, <span class="at">group =</span> prosoc_left),</span>
<span id="cb41-34"><a href="#cb41-34" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">linewidth =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb41-35"><a href="#cb41-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">y =</span> Estimate, <span class="at">ymin =</span> Q2<span class="fl">.5</span>, <span class="at">ymax =</span> Q97<span class="fl">.5</span>, <span class="at">shape =</span> condition),</span>
<span id="cb41-36"><a href="#cb41-36" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">fill =</span> <span class="st">"#100F14"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span> </span>
<span id="cb41-37"><a href="#cb41-37" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Annotation for the conditions</span></span>
<span id="cb41-38"><a href="#cb41-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> text,</span>
<span id="cb41-39"><a href="#cb41-39" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">y =</span> prop, <span class="at">label =</span> labels), </span>
<span id="cb41-40"><a href="#cb41-40" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#DCA258"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb41-41"><a href="#cb41-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb41-42"><a href="#cb41-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="st">"proportion left lever"</span>, <span class="at">breaks =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">2</span> <span class="sc">/</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"0"</span>, <span class="st">".5"</span>, <span class="st">"1"</span>)) <span class="sc">+</span></span>
<span id="cb41-43"><a href="#cb41-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">19</span>)) <span class="sc">+</span></span>
<span id="cb41-44"><a href="#cb41-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">subtitle =</span> <span class="st">"Posterior predictions, in light blue, against the raw data, in dark blue, for</span><span class="sc">\n</span><span class="st">model b14.3, the cross-classified varying effects model."</span>) <span class="sc">+</span></span>
<span id="cb41-45"><a href="#cb41-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> actor, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">labeller =</span> label_both)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>These chimpanzees simply did not behave in any consistently different way in the partner treatments. The model we’ve used here does have some advantages, though. Since it allows for some individuals to differ in how they respond to the treatments, it could reveal a situation in which a treatment has no effect on average, even though some of the individuals respond strongly. That wasn’t the case here. But often we are more interested in the distribution of responses than in the average response, so a model that estimates the distribution of treatment effects is very useful. (p.&nbsp;452)</p>
</blockquote>
<p>For more practice with models of this kind, check out my blog post, <a href="https://solomonkurz.netlify.app/blog/2020-12-09-multilevel-models-and-the-index-variable-approach/"><em>Multilevel models and the index-variable approach</em></a>.</p>
</section>
<section id="instruments-and-causal-designs" class="level2 page-columns page-full" data-number="14.3">
<h2 data-number="14.3" class="anchored" data-anchor-id="instruments-and-causal-designs"><span class="header-section-number">14.3</span> Instruments and causal designs</h2>
<blockquote class="blockquote">
<p>Of course sometimes it won’t be possible to close all of the non-causal paths or rule of unobserved confounds. What can be done in that case? More than nothing. If you are lucky, there are ways to exploit a combination of natural experiments and clever modeling that allow causal inference even when non-causal paths cannot be closed. (p.&nbsp;455)</p>
</blockquote>
<section id="instrumental-variables" class="level3" data-number="14.3.1">
<h3 data-number="14.3.1" class="anchored" data-anchor-id="instrumental-variables"><span class="header-section-number">14.3.1</span> Instrumental variables</h3>
<p>Say were are interested in the causal impact of education <span class="math inline">\(E\)</span> on wages <span class="math inline">\(W\)</span>, <span class="math inline">\(E \rightarrow W\)</span>. Further imagine there is some unmeasured variable <span class="math inline">\(U\)</span> that has causal relations with both, <span class="math inline">\(E \leftarrow U \rightarrow W\)</span>, creating a backdoor path. We might use good old <strong>ggdag</strong> to plot the DAG.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdag)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>dag_coords <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"E"</span>, <span class="st">"U"</span>, <span class="st">"W"</span>),</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">x    =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">y    =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before we make the plot, we’ll make a custom theme, <code>theme_pearl_dag()</code>, to streamline our DAG plots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>theme_pearl_dag <span class="ot">&lt;-</span> <span class="cf">function</span>(...) {</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_earring</span>() <span class="sc">+</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_dag</span>() <span class="sc">+</span> </span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"#100F14"</span>),</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>          ...)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dagify</span>(E <span class="sc">~</span> U,</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>       W <span class="sc">~</span> E <span class="sc">+</span> U,</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">coords =</span> dag_coords) <span class="sc">|&gt;</span></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">xend =</span> xend, <span class="at">yend =</span> yend)) <span class="sc">+</span></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> name <span class="sc">==</span> <span class="st">"U"</span>),</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">"#FCF9F0"</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">6</span>, <span class="at">stroke =</span> <span class="dv">2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_text</span>(<span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_edges</span>(<span class="at">edge_colour =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#EEDA9D"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_dag</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<p>Instrumental variables will solve some of the difficulties we have in not being able to condition on <span class="math inline">\(U\)</span>. Here we’ll call our instrumental variable <span class="math inline">\(Q\)</span>. In the terms of the present example, the <strong>instrumental variable</strong> has the qualities that</p>
<ul>
<li><span class="math inline">\(Q\)</span> is independent of <span class="math inline">\(U\)</span>,</li>
<li><span class="math inline">\(Q\)</span> is not independent of <span class="math inline">\(E\)</span>, and</li>
<li><span class="math inline">\(Q\)</span> can only influence <span class="math inline">\(W\)</span> through <span class="math inline">\(E\)</span> (i.e., the effect of <span class="math inline">\(Q\)</span> on <span class="math inline">\(W\)</span> is fully mediated by <span class="math inline">\(E\)</span>).</li>
</ul>
<p>There is what this looks like in a DAG.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>dag_coords <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"Q"</span>, <span class="st">"E"</span>, <span class="st">"U"</span>, <span class="st">"W"</span>),</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x    =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y    =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dagify</span>(E <span class="sc">~</span> Q <span class="sc">+</span> U,</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>       W <span class="sc">~</span> E <span class="sc">+</span> U,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">coords =</span> dag_coords) <span class="sc">|&gt;</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">xend =</span> xend, <span class="at">yend =</span> yend)) <span class="sc">+</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> name <span class="sc">==</span> <span class="st">"U"</span>),</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">"#FCF9F0"</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">6</span>, <span class="at">stroke =</span> <span class="dv">2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_text</span>(<span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_edges</span>(<span class="at">edge_colour =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#EEDA9D"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_dag</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="336"></p>
</figure>
</div>
</div>
</div>
<p>Sadly, our condition that <span class="math inline">\(Q\)</span> can only influence <span class="math inline">\(W\)</span> through <span class="math inline">\(E\)</span>–often called the <strong>exclusion restriction</strong>–generally cannot be tested. Given <span class="math inline">\(U\)</span> is unmeasured, by definition, we also cannot test that <span class="math inline">\(Q\)</span> is independent of <span class="math inline">\(U\)</span>. These are model assumptions.</p>
<p>Let’s simulate data based on <span class="citation" data-cites="angristDoesCompulsorySchool1991">Angrist &amp; Keueger (<a href="references.html#ref-angristDoesCompulsorySchool1991" role="doc-biblioref">1991</a>)</span> to get a sense of how this works.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a standardizing function</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>standardize <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  (x <span class="sc">-</span> <span class="fu">mean</span>(x)) <span class="sc">/</span> <span class="fu">sd</span>(x)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">73</span>) </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>dat_sim <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">u_sim =</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>),</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>                  <span class="at">q_sim =</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">size =</span> n, <span class="at">replace =</span> T)) <span class="sc">|&gt;</span> </span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">e_sim =</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> u_sim <span class="sc">+</span> q_sim, <span class="at">sd =</span> <span class="dv">1</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">w_sim =</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> u_sim <span class="sc">+</span> <span class="dv">0</span> <span class="sc">*</span> e_sim, <span class="at">sd =</span> <span class="dv">1</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">w =</span> <span class="fu">standardize</span>(w_sim),</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">e =</span> <span class="fu">standardize</span>(e_sim),</span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">q =</span> <span class="fu">standardize</span>(q_sim))</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>dat_sim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 500 × 7
     u_sim q_sim e_sim   w_sim       w       e      q
     &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;
 1 -0.145      1 1.51   0.216   0.173  -0.575  -1.36 
 2  0.291      1 0.664  0.846   0.584  -1.09   -1.36 
 3  0.0938     3 2.44  -0.664  -0.402  -0.0185  0.428
 4 -0.127      3 4.09  -0.725  -0.442   0.978   0.428
 5 -0.847      4 2.62  -1.24   -0.780   0.0939  1.32 
 6  0.141      4 3.54  -0.0700 -0.0146  0.651   1.32 
 7  1.54       2 3.65   1.88    1.26    0.714  -0.464
 8  2.74       3 4.91   2.52    1.67    1.48    0.428
 9  1.55       3 4.18   0.624   0.439   1.04    0.428
10  0.462      1 0.360  0.390   0.286  -1.27   -1.36 
# ℹ 490 more rows</code></pre>
</div>
</div>
<p><span class="math inline">\(Q\)</span> in this context is like quarter in the school year, but inversely scaled such that larger numbers indicate more quarters. In this simulation, we have set the true effect of education on wages–<span class="math inline">\(E \rightarrow W\)</span>–to be zero. Any univariate association is through the confounding variable <span class="math inline">\(U\)</span>. Also, <span class="math inline">\(Q\)</span> has no direct effect on <span class="math inline">\(W\)</span> or <span class="math inline">\(U\)</span>, but it does have a causal relation with <span class="math inline">\(E\)</span>, which is <span class="math inline">\(Q \rightarrow E \leftarrow U\)</span>. First we fit the univariable model corresponding to <span class="math inline">\(E \rightarrow W\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.4</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat_sim,</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  w <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> e,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b),</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma)),</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,  </span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.04"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity 
Formula: w ~ 1 + e 
   Data: dat_sim (Number of observations: 500) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Regression Coefficients:
          Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept    -0.00      0.04    -0.08     0.08 1.00     4366     3289
e             0.40      0.04     0.32     0.48 1.00     3866     2712

Further Distributional Parameters:
      Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma     0.92      0.03     0.87     0.98 1.00     4561     2935

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Because we have not conditioned on <span class="math inline">\(U\)</span>, then model suggests a moderately large spurious causal relation for <span class="math inline">\(E \rightarrow W\)</span>. Now see what happens when we also condition directly on <span class="math inline">\(Q\)</span>, as in <span class="math inline">\(Q \rightarrow W \leftarrow E\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.5</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat_sim,</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  w <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> e <span class="sc">+</span> q,</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b),</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma)),</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,  </span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.05"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity 
Formula: w ~ 1 + e + q 
   Data: dat_sim (Number of observations: 500) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Regression Coefficients:
          Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept    -0.00      0.04    -0.07     0.07 1.00     3652     3159
e             0.64      0.05     0.54     0.73 1.00     3623     2952
q            -0.41      0.05    -0.50    -0.31 1.00     3444     2824

Further Distributional Parameters:
      Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma     0.86      0.03     0.81     0.91 1.00     4070     3007

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Holy smokes that’s a mess. This model suggests both <span class="math inline">\(E\)</span> and <span class="math inline">\(Q\)</span> have moderate to strong causal effects on <span class="math inline">\(W\)</span>, even though we know neither do based on the true data-generating model. Like McElreath said, “bad stuff happens” when we condition on an instrumental variable this way.</p>
<blockquote class="blockquote">
<p>There is no backdoor path through <span class="math inline">\(Q\)</span>, as you can see. But there is a non-causal path from <span class="math inline">\(Q\)</span> to <span class="math inline">\(W\)</span> through <span class="math inline">\(U\)</span>: <span class="math inline">\(Q \rightarrow E \leftarrow U \rightarrow W\)</span>. This is a non-causal path, because changing <span class="math inline">\(Q\)</span> doesn’t result in any change in <span class="math inline">\(W\)</span> through this path. But since we are conditioning on <span class="math inline">\(E\)</span> in the same model, and <span class="math inline">\(E\)</span> is a collider of <span class="math inline">\(Q\)</span> and <span class="math inline">\(U\)</span>, the non-causal path is open. This confounds the coefficient on <span class="math inline">\(Q\)</span>. It won’t be zero, because it’ll pick up the association between <span class="math inline">\(U\)</span> and <span class="math inline">\(W\)</span>. And then, as a result, the coefficient on <span class="math inline">\(E\)</span> can get even more confounded. Used this way, an instrument like <span class="math inline">\(Q\)</span> might be called a <strong>bias amplifier</strong>. (p.&nbsp;456, <strong>emphasis</strong> in the original)</p>
</blockquote>
<p>The statistical solution to this mess is to express the data-generating DAG as a multivariate statistical model following the form</p>
<p><span class="math display">\[\begin{align*}
\begin{bmatrix} W_i \\ E_i \end{bmatrix} &amp; \sim \operatorname{MVNormal} \begin{pmatrix} \begin{bmatrix} \mu_{\text W,i} \\ \mu_{\text E,i} \end{bmatrix}, \color{#A65141}{\mathbf \Sigma} \end{pmatrix} \\
\mu_{\text W,i} &amp; = \alpha_\text W + \beta_\text{EW} E_i \\
\mu_{\text E,i} &amp; = \alpha_\text E + \beta_\text{QE} Q_i \\
\color{#A65141}{\mathbf\Sigma} &amp; \color{#A65141}= \color{#A65141}{\begin{bmatrix} \sigma_\text W &amp; 0 \\ 0  &amp; \sigma_\text E \end{bmatrix} \mathbf R \begin{bmatrix} \sigma_\text W &amp; 0 \\ 0  &amp; \sigma_\text E \end{bmatrix}} \\
\color{#A65141}{\mathbf R} &amp; \color{#A65141}= \color{#A65141}{\begin{bmatrix} 1 &amp; \rho \\ \rho &amp; 1 \end{bmatrix}} \\
\alpha_\text W \text{ and } \alpha_\text E &amp; \sim \operatorname{Normal}(0, 0.2) \\
\beta_\text{EW} \text{ and } \beta_\text{QE} &amp; \sim \operatorname{Normal}(0, 0.5) \\
\sigma_\text W \text{ and } \sigma_\text E &amp; \sim \operatorname{Exponential}(1) \\
\rho &amp; \sim \operatorname{LKJ}(2).
\end{align*}\]</span></p>
<p>You might not remember, but we’ve actually fit a model like this before. It was <code>b5.3_A</code> from way back in <a href="05.html#sec-Counterfactual-plots" class="quarto-xref"><span>Section 5.1.5.3</span></a>. The big difference between that earlier model and this one is whereas the former did not include a residual correlation, <span class="math inline">\(\rho\)</span>, this one will. Thus, this time we will make sure to set <code>set_rescor(TRUE)</code> in the <code>formula</code>. Within <strong>brms</strong> parlance, priors for residual correlations are of <code>class = rescor</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>e_model <span class="ot">&lt;-</span> <span class="fu">bf</span>(e <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> q)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>w_model <span class="ot">&lt;-</span> <span class="fu">bf</span>(w <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> e)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.6</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat_sim, </span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  e_model <span class="sc">+</span> w_model <span class="sc">+</span> <span class="fu">set_rescor</span>(<span class="cn">TRUE</span>),</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># E model</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept, <span class="at">resp =</span> e),</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b, <span class="at">resp =</span> e),</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma, <span class="at">resp =</span> e),</span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># W model</span></span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept, <span class="at">resp =</span> w),</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b, <span class="at">resp =</span> w),</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma, <span class="at">resp =</span> w),</span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb53-19"><a href="#cb53-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rho</span></span>
<span id="cb53-20"><a href="#cb53-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> rescor)),</span>
<span id="cb53-21"><a href="#cb53-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb53-22"><a href="#cb53-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb53-23"><a href="#cb53-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.06"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: MV(gaussian, gaussian) 
  Links: mu = identity
         mu = identity 
Formula: e ~ 1 + q 
         w ~ 1 + e 
   Data: dat_sim (Number of observations: 500) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Regression Coefficients:
            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
e_Intercept     0.00      0.03    -0.07     0.07 1.00     2899     3024
w_Intercept     0.00      0.04    -0.09     0.09 1.00     2969     2621
e_q             0.59      0.04     0.52     0.66 1.00     2560     2739
w_e            -0.05      0.08    -0.20     0.09 1.00     1823     2100

Further Distributional Parameters:
        Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma_e     0.81      0.03     0.76     0.86 1.00     3579     2982
sigma_w     1.02      0.05     0.94     1.12 1.00     1949     2084

Residual Correlations: 
            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
rescor(e,w)     0.54      0.05     0.43     0.64 1.00     1822     2135

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Now the parameter for <span class="math inline">\(E \rightarrow W\)</span>, <code>w_e</code>, is just where it should be–near zero. The residual correlation between <span class="math inline">\(E\)</span> and <span class="math inline">\(Q\)</span>, <code>rescor(e,w)</code>, is positive and large in magnitude, indicating their common influence from the unmeasured variable <span class="math inline">\(U\)</span>. Next we’ll take McElreath’s direction to “adjust the simulation and try other scenarios” (p.&nbsp;459) by adjusting the causal relations, as in his <strong>R</strong> code 14.28.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">73</span>) </span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>dat_sim <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">u_sim =</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>),</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">q_sim =</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">size =</span> n, <span class="at">replace =</span> T)) <span class="sc">|&gt;</span> </span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">e_sim =</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> u_sim <span class="sc">+</span> q_sim, <span class="at">sd =</span> <span class="dv">1</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">w_sim =</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> <span class="sc">-</span>u_sim <span class="sc">+</span> <span class="fl">0.2</span> <span class="sc">*</span> e_sim, <span class="at">sd =</span> <span class="dv">1</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">w =</span> <span class="fu">standardize</span>(w_sim),</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">e =</span> <span class="fu">standardize</span>(e_sim),</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">q =</span> <span class="fu">standardize</span>(q_sim))</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>dat_sim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 500 × 7
     u_sim q_sim e_sim  w_sim       w       e      q
     &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;
 1 -0.145      1 1.51   0.809  0.248  -0.575  -1.36 
 2  0.291      1 0.664  0.396 -0.0563 -1.09   -1.36 
 3  0.0938     3 2.44  -0.364 -0.615  -0.0185  0.428
 4 -0.127      3 4.09   0.347 -0.0922  0.978   0.428
 5 -0.847      4 2.62   0.976  0.370   0.0939  1.32 
 6  0.141      4 3.54   0.357 -0.0852  0.651   1.32 
 7  1.54       2 3.65  -0.466 -0.690   0.714  -0.464
 8  2.74       3 4.91  -1.98  -1.80    1.48    0.428
 9  1.55       3 4.18  -1.64  -1.55    1.04    0.428
10  0.462      1 0.360 -0.461 -0.686  -1.27   -1.36 
# ℹ 490 more rows</code></pre>
</div>
</div>
<p>We’ll use <code>update()</code> to avoid re-compiling the models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.4</span>x <span class="ot">&lt;-</span> <span class="fu">update</span>(</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  b14<span class="fl">.4</span>,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">newdata =</span> dat_sim,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.04x"</span>)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.6</span>x <span class="ot">&lt;-</span> <span class="fu">update</span>(</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>  b14<span class="fl">.6</span>,</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">newdata =</span> dat_sim,</span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>, </span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.06x"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Just for kicks, let’s examine the results with a coefficient plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>text <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Estimate =</span> <span class="fu">c</span>(<span class="fu">fixef</span>(b14<span class="fl">.4</span>x)[<span class="dv">2</span>, <span class="dv">3</span>], <span class="fu">fixef</span>(b14<span class="fl">.6</span>x)[<span class="dv">4</span>, <span class="dv">4</span>]),</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y        =</span> <span class="fu">c</span>(<span class="fl">4.35</span>, <span class="fl">3.65</span>),</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">hjust    =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">fit      =</span> <span class="fu">c</span>(<span class="st">"b14.4x"</span>, <span class="st">"b14.6x"</span>))</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a><span class="fu">bind_rows</span>(</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># b_14.4x</span></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior_summary</span>(b14<span class="fl">.4</span>x)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ] <span class="sc">|&gt;</span> </span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">param =</span> <span class="fu">c</span>(<span class="st">"alpha[W]"</span>, <span class="st">"beta[EW]"</span>, <span class="st">"sigma[W]"</span>),</span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">fit =</span> <span class="st">"b14.4x"</span>),</span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># b_14.6x</span></span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior_summary</span>(b14<span class="fl">.6</span>x)[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, ] <span class="sc">|&gt;</span>  </span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">param =</span> <span class="fu">c</span>(<span class="st">"alpha[E]"</span>, <span class="st">"alpha[W]"</span>, <span class="st">"beta[QE]"</span>, <span class="st">"beta[EW]"</span>, <span class="st">"sigma[E]"</span>, <span class="st">"sigma[W]"</span>, <span class="st">"rho"</span>),</span>
<span id="cb59-17"><a href="#cb59-17" aria-hidden="true" tabindex="-1"></a>           <span class="at">fit =</span> <span class="st">"b14.6x"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb59-18"><a href="#cb59-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">param =</span> <span class="fu">factor</span>(param,</span>
<span id="cb59-19"><a href="#cb59-19" aria-hidden="true" tabindex="-1"></a>                        <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"rho"</span>, <span class="st">"sigma[W]"</span>, <span class="st">"sigma[E]"</span>, <span class="st">"beta[EW]"</span>, <span class="st">"beta[QE]"</span>, <span class="st">"alpha[W]"</span>, <span class="st">"alpha[E]"</span>))) <span class="sc">|&gt;</span></span>
<span id="cb59-20"><a href="#cb59-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb59-21"><a href="#cb59-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> param, <span class="at">y =</span> Estimate, <span class="at">color =</span> fit)) <span class="sc">+</span></span>
<span id="cb59-22"><a href="#cb59-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">color =</span> <span class="st">"#E8DCCF"</span>) <span class="sc">+</span></span>
<span id="cb59-23"><a href="#cb59-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Q2<span class="fl">.5</span>, <span class="at">ymax =</span> Q97<span class="fl">.5</span>),</span>
<span id="cb59-24"><a href="#cb59-24" aria-hidden="true" tabindex="-1"></a>                  <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.5</span>), <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb59-25"><a href="#cb59-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> text,</span>
<span id="cb59-26"><a href="#cb59-26" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> y, <span class="at">label =</span> fit, <span class="at">hjust =</span> hjust)) <span class="sc">+</span></span>
<span id="cb59-27"><a href="#cb59-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe) <span class="sc">+</span></span>
<span id="cb59-28"><a href="#cb59-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="cn">NULL</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#E7CDC2"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb59-29"><a href="#cb59-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"marginal posterior"</span>) <span class="sc">+</span></span>
<span id="cb59-30"><a href="#cb59-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb59-31"><a href="#cb59-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>),</span>
<span id="cb59-32"><a href="#cb59-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb59-33"><a href="#cb59-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>With the help from <code>b14.6x</code>, we found “that <span class="math inline">\(E\)</span> and <span class="math inline">\(W\)</span> have a negative correlation in their residual variance, because the confound positively influences one and negatively influences the other” (p.&nbsp;459).</p>
<p>One can use the <code>dagitty()</code> and <code>instrumentalVariables()</code> functions from the <strong>dagitty</strong> package to first define a DAG and then query whether there are instrumental variables for a given exposure and outcome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dagitty)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>dagIV <span class="ot">&lt;-</span> <span class="fu">dagitty</span>(<span class="st">"dag{Q -&gt; E &lt;- U -&gt; W &lt;- E}"</span>)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="fu">instrumentalVariables</span>(dagIV, <span class="at">exposure =</span> <span class="st">"E"</span>, <span class="at">outcome =</span> <span class="st">"W"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Q</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>The hardest thing about instrumental variables is believing in any particular instrument. If you believe in your DAG, they are easy to believe. But should you believe in your DAG?…</p>
<p>In general, it is not possible to statistically prove whether a variable is a good instrument. As always, we need scientific knowledge outside of the data to make sense of the data. (p.&nbsp;460)</p>
</blockquote>
<section id="rethinking-two-stage-worst-squares" class="level4" data-number="14.3.1.1">
<h4 data-number="14.3.1.1" class="anchored" data-anchor-id="rethinking-two-stage-worst-squares"><span class="header-section-number">14.3.1.1</span> Rethinking: Two-stage worst squares</h4>
<p>“The instrumental variable model is often discussed with an estimation procedure known as <strong>two-stage least squares</strong> (2SLS)” (p.&nbsp;460, <strong>emphasis</strong> in the original). For a nice introduction to instrumental variables via 2SLS, see this <a href="https://evalf20.classes.andrewheiss.com/example/iv/">practical introduction</a>, and also the <a href="https://evalf20.classes.andrewheiss.com/content/11-content/">slides and video-lecture files</a>, from the great <a href="https://www.andrewheiss.com/">Andrew Heiss</a>. For the clinical researchers in the room, <a href="https://rpsychologist.com/about">Kristoffer Magnusson</a> has a nice <strong>brms</strong>-based blog post on the topic called <a href="https://rpsychologist.com/adherence-analysis-IV-brms"><em>Confounded dose-response effects of treatment adherence: fitting Bayesian instrumental variable models using brms</em></a>.</p>
</section>
</section>
<section id="other-designs" class="level3 page-columns page-full" data-number="14.3.2">
<h3 data-number="14.3.2" class="anchored" data-anchor-id="other-designs"><span class="header-section-number">14.3.2</span> Other designs</h3>
<blockquote class="blockquote">
<p>There are potentially many ways to find natural experiments. Not all of them are strictly instrumental variables. But they can provide theoretically correct designs for causal inference, if you can believe the assumptions. Let’s consider two more.</p>
<p>In addition to the backdoor criterion you met in <a href="06.html" class="quarto-xref"><span>Chapter 6</span></a>, there is something called the <strong>front-door criterion</strong>. (p.&nbsp;460, <strong>emphasis</strong> in the original)</p>
</blockquote>
<p>To get a sense of the front-door criterion, consider the following DAG with observed variables <span class="math inline">\(X\)</span>, <span class="math inline">\(Y\)</span>, and <span class="math inline">\(Z\)</span> and an unobserved variable, <span class="math inline">\(U\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>dag_coords <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"X"</span>, <span class="st">"Z"</span>, <span class="st">"U"</span>, <span class="st">"Y"</span>),</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x    =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y    =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dagify</span>(X <span class="sc">~</span> U,</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>       Z <span class="sc">~</span> X,</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>       Y <span class="sc">~</span> U <span class="sc">+</span> Z,</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">coords =</span> dag_coords) <span class="sc">|&gt;</span></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">xend =</span> xend, <span class="at">yend =</span> yend)) <span class="sc">+</span></span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> name <span class="sc">==</span> <span class="st">"U"</span>),</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">"#FCF9F0"</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">6</span>, <span class="at">stroke =</span> <span class="dv">2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_text</span>(<span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_edges</span>(<span class="at">edge_colour =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#EEDA9D"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_dag</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>We are interested, as usual, in the causal influence of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span>. But there is an unobserved confound <span class="math inline">\(U\)</span>, again as usual. It turns out that, if we can find a perfect mediator <span class="math inline">\(Z\)</span>, then we can possibly estimate the causal effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span>. It isn’t crazy to think that causes are mediated by other causes. Everything has a mechanism. <span class="math inline">\(Z\)</span> in the DAG above is such a mechanism. If you have a believable <span class="math inline">\(Z\)</span> variable, then the causal effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span> is estimated by expressing the generative model as a statistical model, similar to the instrumental variable example before. (p.&nbsp;461)</p>
</blockquote>
<p>McElreath’s second example is the <strong>regression discontinuity</strong> approach. If you have a time series where the variable of interest is measured before and after some relevant intervention variable, you can estimate intercepts and slopes before and after the intervention, the cutoff. However,</p>
<blockquote class="blockquote">
<p>in practice, one trend is fit for individuals above the cutoff and another to those below the cutoff. Then an estimate of the causal effect is the average difference between individuals just above and just below the cutoff. While the difference near the cuttoff is of interest, the entire function influences this difference. So some care is needed in choosing functions for the overall relationship between the exposure and the outcome. (p.&nbsp;461)</p>
</blockquote>
<p>McElreath’s not kidding about the need for care when fitting regression discontinuity models. Gleman’s blog is littered with awful examples (e.g., <a href="https://statmodeling.stat.columbia.edu/2019/06/25/another-regression-discontinuity-disaster-and-what-can-we-learn-from-it/">here</a>, <a href="https://statmodeling.stat.columbia.edu/2018/08/02/38160/">here</a>, <a href="https://statmodeling.stat.columbia.edu/2020/07/02/no-i-dont-believe-that-claim-based-on-regression-discontinuity-analysis-that/">here</a>, <a href="https://statmodeling.stat.columbia.edu/2020/01/13/how-to-get-out-of-the-credulity-rut-regression-discontinuity-edition-getting-beyond-whack-a-mole/">here</a>, <a href="https://statmodeling.stat.columbia.edu/2020/07/18/please-socially-distance-me-from-this-regression-model/">here</a>). See also Gelman and Imbens’ <span class="citation" data-cites="gelmanWhyHighorderPolynomials2019">(<a href="references.html#ref-gelmanWhyHighorderPolynomials2019" role="doc-biblioref">2019</a>)</span> paper, <a href="https://stat.columbia.edu/~gelman/research/published/2018_gelman_jbes.pdf"><em>Why high-order polynomials should not be used in regression discontinuity designs</em></a>, or <a href="https://twitter.com/nickchk">Nick HK</a>’s informative <a href="https://twitter.com/nickchk/status/1329338429360914433">tweet</a> on how this applies to autocorrelated data.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<div class="no-row-height column-margin column-container"><div id="fn1"><p><sup>1</sup>&nbsp;We’ll briefly cover autoregressive models in <a href="16.html#sec-Population-dynamics" class="quarto-xref"><span>Section 16.4</span></a>.</p></div></div></section>
</section>
<section id="sec-Social-relations-as" class="level2" data-number="14.4">
<h2 data-number="14.4" class="anchored" data-anchor-id="sec-Social-relations-as"><span class="header-section-number">14.4</span> Social relations as correlated varying effects</h2>
<p>It looks like <strong>brms</strong> is not set up to fit a model like this, at this time. See the <a href="https://discourse.mc-stan.org/t/social-relations-model-srm/17121">Social relations model (SRM) thread</a> on the Stan Forums and <a href="https://github.com/paul-buerkner/brms/issues/502">issue #502</a> on the <strong>brms</strong> GitHub repo for details. In short, the difficulty is <strong>brms</strong> is not set up to allow covariances among distinct random effects with the same levels and it looks like this will not change any time soon. So, in this section we will fit the model with <strong>rethinking</strong>, but still use <strong>ggplot2</strong> and friends in the post processing.</p>
<p>Let’s load the <code>kl_dyads</code> data <span class="citation" data-cites="kosterFoodSharingNetworks2014">(<a href="references.html#ref-kosterFoodSharingNetworks2014" role="doc-biblioref">Koster &amp; Leckie, 2014</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rethinking)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(KosterLeckie)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Take a look at the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>kl_dyads <span class="sc">|&gt;</span> <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 300
Columns: 13
$ hidA    &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2…
$ hidB    &lt;int&gt; 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 3, 4, 5, 6, 7, 8, 9, 10,…
$ did     &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3…
$ giftsAB &lt;int&gt; 0, 6, 2, 4, 8, 2, 1, 0, 10, 1, 0, 0, 1, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 1, 2, 0, 3, 0, 43, 5, 0, 1, 0, 1, 1, 2…
$ giftsBA &lt;int&gt; 4, 31, 5, 2, 2, 1, 2, 1, 110, 0, 0, 6, 11, 0, 1, 4, 0, 2, 0, 7, 0, 13, 0, 2, 3, 1, 1, 0, 1, 10, 4, 1, 0, 0, 2, 0…
$ offset  &lt;dbl&gt; 0.000, -0.003, -0.019, 0.000, -0.003, 0.000, 0.000, 0.000, -0.186, 0.000, -0.471, -0.019, -0.011, -0.014, -0.019…
$ drel1   &lt;int&gt; 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0…
$ drel2   &lt;int&gt; 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…
$ drel3   &lt;int&gt; 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0…
$ drel4   &lt;int&gt; 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1…
$ dlndist &lt;dbl&gt; -2.790, -2.817, -1.886, -1.892, -3.499, -1.853, -1.475, -1.644, -1.897, -2.379, -2.200, -2.117, -2.830, -2.492, …
$ dass    &lt;dbl&gt; 0.000, 0.044, 0.025, 0.011, 0.022, 0.071, 0.046, 0.003, 0.552, 0.018, 0.004, 0.004, 0.036, 0.006, 0.014, 0.000, …
$ d0125   &lt;int&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># kl_households |&gt; glimpse()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>“The variables <code>hidA</code> and <code>hidB</code> tell us the household IDs in each dyad, and <code>did</code> is a unique dyad ID number” (p.&nbsp;462). To get a sense of the interrelation among those three ID variables, we’ll make a tile plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>kl_dyads <span class="sc">|&gt;</span> </span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> hidA, <span class="at">y =</span> hidB, <span class="at">label =</span> did)) <span class="sc">+</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> did),</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">size =</span> <span class="fl">2.25</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">24</span> <span class="sc">+</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"#394165"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span> <span class="sc">+</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"#394165"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"#DCA258"</span>, <span class="at">high =</span> <span class="st">"#EEDA9D"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">24</span>) <span class="sc">+</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">25</span>) <span class="sc">+</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>),</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>The orange/yellow gradient fill is a little silly, but I couldn’t stop myself. Here is our version of Figure 14.8, the bivariate distribution of dyadic gifts, collapsing across dyads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>kl_dyads <span class="sc">|&gt;</span> </span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> giftsAB, <span class="at">y =</span> giftsBA)) <span class="sc">+</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hex</span>(<span class="at">bins =</span> <span class="dv">70</span>) <span class="sc">+</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">color =</span> <span class="st">"#DCA258"</span>, <span class="at">linetype =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">"gifts household A to household B"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">113</span>)) <span class="sc">+</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="st">"gifts from B to A"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">113</span>)) <span class="sc">+</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"#E7CDC2"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Distribution of dyadic gifts"</span>) <span class="sc">+</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here’s the overall Pearson’s correlation coefficient, collapsing across grouping levels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>kl_dyads <span class="sc">|&gt;</span> </span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">rho =</span> <span class="fu">cor</span>(giftsAB, giftsBA))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        rho
1 0.2391549</code></pre>
</div>
</div>
<p>However, it would be a mistake to take this correlation seriously. It is a disentangled mixture of various kinds of associations, none of which are guaranteed to be even close to <span class="math inline">\(r = 0.24\)</span>. Remember this as we move along with the analyses and let the consequences burn a methodological mark into your soul.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>kl_data <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">N            =</span> <span class="fu">nrow</span>(kl_dyads),</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">N_households =</span> <span class="fu">max</span>(kl_dyads<span class="sc">$</span>hidB), </span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">did          =</span> kl_dyads<span class="sc">$</span>did,</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">hidA         =</span> kl_dyads<span class="sc">$</span>hidA,</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">hidB         =</span> kl_dyads<span class="sc">$</span>hidB,</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">giftsAB      =</span> kl_dyads<span class="sc">$</span>giftsAB, </span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">giftsBA      =</span> kl_dyads<span class="sc">$</span>giftsBA)</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>m14<span class="fl">.7</span> <span class="ot">&lt;-</span> <span class="fu">ulam</span>( </span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">alist</span>(</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>    giftsAB <span class="sc">~</span> <span class="fu">poisson</span>(lambdaAB),</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>    giftsBA <span class="sc">~</span> <span class="fu">poisson</span>(lambdaBA),</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">log</span>(lambdaAB) <span class="ot">&lt;-</span> a <span class="sc">+</span> gr[hidA, <span class="dv">1</span>] <span class="sc">+</span> gr[hidB, <span class="dv">2</span>] <span class="sc">+</span> d[did, <span class="dv">1</span>] , </span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">log</span>(lambdaBA) <span class="ot">&lt;-</span> a <span class="sc">+</span> gr[hidB, <span class="dv">1</span>] <span class="sc">+</span> gr[hidA, <span class="dv">2</span>] <span class="sc">+</span> d[did, <span class="dv">2</span>] , </span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>    a <span class="sc">~</span> <span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># `gr` matrix of varying effects</span></span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>    vector[<span class="dv">2</span>]<span class="sc">:</span>gr[N_households] <span class="sc">~</span> <span class="fu">multi_normal</span>(<span class="dv">0</span>, Rho_gr, sigma_gr), </span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a>    Rho_gr <span class="sc">~</span> <span class="fu">lkj_corr</span>(<span class="dv">4</span>),</span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a>    sigma_gr <span class="sc">~</span> <span class="fu">exponential</span>(<span class="dv">1</span>),</span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-23"><a href="#cb71-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Dyad effects</span></span>
<span id="cb71-24"><a href="#cb71-24" aria-hidden="true" tabindex="-1"></a>    transpars<span class="sc">&gt;</span> matrix[N,<span class="dv">2</span>]<span class="sc">:</span>d <span class="ot">&lt;-</span></span>
<span id="cb71-25"><a href="#cb71-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">compose_noncentered</span>(<span class="fu">rep_vector</span>(sigma_d, <span class="dv">2</span>), L_Rho_d, z), </span>
<span id="cb71-26"><a href="#cb71-26" aria-hidden="true" tabindex="-1"></a>    matrix[<span class="dv">2</span>,N]<span class="sc">:</span>z <span class="sc">~</span> <span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb71-27"><a href="#cb71-27" aria-hidden="true" tabindex="-1"></a>    cholesky_factor_corr[<span class="dv">2</span>]<span class="sc">:</span>L_Rho_d <span class="sc">~</span> <span class="fu">lkj_corr_cholesky</span>(<span class="dv">8</span>), </span>
<span id="cb71-28"><a href="#cb71-28" aria-hidden="true" tabindex="-1"></a>    sigma_d <span class="sc">~</span> <span class="fu">exponential</span>(<span class="dv">1</span>),</span>
<span id="cb71-29"><a href="#cb71-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-30"><a href="#cb71-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Compute correlation matrix for dyads</span></span>
<span id="cb71-31"><a href="#cb71-31" aria-hidden="true" tabindex="-1"></a>    gq<span class="sc">&gt;</span> matrix[<span class="dv">2</span>, <span class="dv">2</span>]<span class="sc">:</span>Rho_d <span class="ot">&lt;&lt;-</span> <span class="fu">Chol_to_Corr</span>(L_Rho_d)</span>
<span id="cb71-32"><a href="#cb71-32" aria-hidden="true" tabindex="-1"></a>  ), </span>
<span id="cb71-33"><a href="#cb71-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> kl_data, </span>
<span id="cb71-34"><a href="#cb71-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>, <span class="at">iter =</span> <span class="dv">2000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We don’t get a lot of information from the default <code>precis()</code> output for this model, but at least we’ll get a summary for <span class="math inline">\(\alpha\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">precis</span>(m14<span class="fl">.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>        mean   sd 5.5% 94.5% rhat ess_bulk
a       0.54 0.16 0.27   0.8    1   680.38
sigma_d 1.10 0.06 1.01   1.2    1  1217.32</code></pre>
</div>
</div>
<p>One of the interesting things about this model is we only have one <span class="math inline">\(\alpha\)</span> parameter for two criterion variables. This makes <span class="math inline">\(\alpha\)</span> like the grand mean of counts. Here is a focused look at the <code>precis()</code> output when you set <code>depth = 3</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">precis</span>(m14<span class="fl">.7</span>, <span class="at">depth =</span> <span class="dv">3</span>, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"Rho_gr"</span>, <span class="st">"sigma_gr"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>             mean   sd  5.5% 94.5% rhat ess_bulk
Rho_gr[1,1]  1.00 0.00  1.00  1.00   NA       NA
Rho_gr[2,1] -0.41 0.20 -0.70 -0.08 1.00  1326.90
Rho_gr[1,2] -0.41 0.20 -0.70 -0.08 1.00  1326.90
Rho_gr[2,2]  1.00 0.00  1.00  1.00   NA       NA
sigma_gr[1]  0.83 0.14  0.64  1.06 1.00  2267.45
sigma_gr[2]  0.42 0.09  0.30  0.58 1.01   942.52</code></pre>
</div>
</div>
<p>These are the posterior summaries for the part of the model McElreath defined in the middle of page 463,</p>
<p><span class="math display">\[
\begin{bmatrix} g_i \\ r_i \end{bmatrix} \sim \operatorname{MVNormal}\begin{pmatrix} \begin{bmatrix} 0 \\ 0 \end{bmatrix}, \begin{bmatrix} \sigma_g^2 &amp; \sigma_g \sigma_r \rho_{gr} \\ \sigma_g \sigma_r \rho_{rg} &amp; \sigma_r^2 \end{bmatrix} \end{pmatrix},
\]</span></p>
<p>the population of household effects. But as per usual with Stan, the variance parameters are expressed in a <span class="math inline">\(\sigma\)</span> metric. Also, <code>rethinking::precis()</code> returned the summary for <span class="math inline">\(\rho_{gr}\)</span> in matrix form,</p>
<p><span class="math display">\[
\begin{bmatrix} 1 &amp; \rho_{gr} \\ \rho_{rg} &amp; 1 \end{bmatrix},
\]</span></p>
<p>where <span class="math inline">\(\rho_{gr} = \rho_{rg}\)</span>. The correlation is negative. We might view <span class="math inline">\(\sigma_g\)</span>, <span class="math inline">\(\sigma_r\)</span>, and their correlation in a plot. First we extract the posterior draws with <code>extract.samples()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>post <span class="ot">&lt;-</span> <span class="fu">extract.samples</span>(m14<span class="fl">.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="st">`</span><span class="at">sigma[italic(g)]</span><span class="st">`</span>          <span class="ot">=</span> post<span class="sc">$</span>sigma_gr[, <span class="dv">1</span>],</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>       <span class="st">`</span><span class="at">sigma[italic(r)]</span><span class="st">`</span>          <span class="ot">=</span> post<span class="sc">$</span>sigma_gr[, <span class="dv">2</span>],</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>       <span class="st">`</span><span class="at">rho[italic(g)][italic(r)]</span><span class="st">`</span> <span class="ot">=</span> post<span class="sc">$</span>Rho_gr[, <span class="dv">2</span>, <span class="dv">1</span>]) <span class="sc">|&gt;</span> </span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">everything</span>()) <span class="sc">|&gt;</span> </span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">y =</span> name, <span class="at">fill =</span> name)) <span class="sc">+</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_halfeye</span>(<span class="at">.width =</span> <span class="fl">0.89</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">height =</span> <span class="fl">1.5</span>) <span class="sc">+</span> </span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe) <span class="sc">+</span></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#EEDA9D"</span>, <span class="st">"#DCA258"</span>)) <span class="sc">+</span></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"marginal posterior"</span>) <span class="sc">+</span></span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">3.9</span>)) <span class="sc">+</span></span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="432"></p>
</figure>
</div>
</div>
</div>
<p>“This implies that individuals who give more across all dyads tend to receive less[, and ] clear evidence that rates of giving are more variable than rates of receiving” (p.&nbsp;465). McElreath suggested we “try <code>plot(exp(g[,1]),exp(r[,1]))</code> for example to show the posterior distribution of giving/receiving for household number 1” (p.&nbsp;465). Here’s a <strong>tidyverse</strong> version of that plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="cf">function</span>(i) post<span class="sc">$</span>a <span class="sc">+</span> post<span class="sc">$</span>gr[, i, <span class="dv">1</span>]) </span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="cf">function</span>(i) post<span class="sc">$</span>a <span class="sc">+</span> post<span class="sc">$</span>gr[, i, <span class="dv">2</span>]) </span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">g =</span> <span class="fu">exp</span>(g[, <span class="dv">1</span>]),</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">r =</span> <span class="fu">exp</span>(r[, <span class="dv">1</span>])) <span class="sc">|&gt;</span> </span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> g, <span class="at">y =</span> r)) <span class="sc">+</span></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="co"># white "#FCF9F0" # gold "#B1934A"</span></span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#B1934A"</span>, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">level =</span> <span class="fl">0.5</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb78-10"><a href="#cb78-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">level =</span> <span class="fl">0.9</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb78-11"><a href="#cb78-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">expression</span>(giving[<span class="fu">italic</span>(i)<span class="sc">==</span><span class="dv">1</span>]),</span>
<span id="cb78-12"><a href="#cb78-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">expression</span>(receiving[<span class="fu">italic</span>(i)<span class="sc">==</span><span class="dv">1</span>])) <span class="sc">+</span></span>
<span id="cb78-13"><a href="#cb78-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">5</span>),</span>
<span id="cb78-14"><a href="#cb78-14" aria-hidden="true" tabindex="-1"></a>              <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The gold dots are the bivariate posterior draws and the two blue ellipses mark off the 50% and 90% intervals, presuming a bivariate Gaussian distribution. Here’s a programmatic way to make our version of Figure 14.9.a.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>d_fig <span class="ot">&lt;-</span> <span class="fu">rbind</span>(<span class="fu">exp</span>(g), <span class="fu">exp</span>(r)) <span class="sc">|&gt;</span> </span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_names</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>) <span class="sc">|&gt;</span> </span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">parameter =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"g"</span>, <span class="st">"r"</span>), <span class="at">each =</span> <span class="fu">n</span>() <span class="sc">/</span> <span class="dv">2</span>),</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">iter      =</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4000</span>, <span class="at">times =</span> <span class="dv">2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">c</span>(parameter, iter), <span class="at">names_to =</span> <span class="st">"household"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> parameter, <span class="at">values_from =</span> value) <span class="sc">|&gt;</span> </span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(household) <span class="sc">|&gt;</span> </span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mu_g =</span> <span class="fu">mean</span>(g),</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">mu_r =</span> <span class="fu">mean</span>(r)) <span class="sc">|&gt;</span> </span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nest</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="st">"g"</span>, <span class="st">"r"</span>, <span class="st">"iter"</span>)) </span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-13"><a href="#cb79-13" aria-hidden="true" tabindex="-1"></a>d_fig <span class="sc">|&gt;</span> </span>
<span id="cb79-14"><a href="#cb79-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">group =</span> household)) <span class="sc">+</span></span>
<span id="cb79-15"><a href="#cb79-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb79-16"><a href="#cb79-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">data =</span> d_fig <span class="sc">|&gt;</span> </span>
<span id="cb79-17"><a href="#cb79-17" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">unnest</span>(data),</span>
<span id="cb79-18"><a href="#cb79-18" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> g, <span class="at">y =</span> r),</span>
<span id="cb79-19"><a href="#cb79-19" aria-hidden="true" tabindex="-1"></a>               <span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">level =</span> <span class="fl">0.5</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb79-20"><a href="#cb79-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> mu_g, <span class="at">y =</span> mu_r),</span>
<span id="cb79-21"><a href="#cb79-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb79-22"><a href="#cb79-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"generalized giving"</span>,</span>
<span id="cb79-23"><a href="#cb79-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"generalized receiving"</span>) <span class="sc">+</span></span>
<span id="cb79-24"><a href="#cb79-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">8.5</span>),</span>
<span id="cb79-25"><a href="#cb79-25" aria-hidden="true" tabindex="-1"></a>              <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">8.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here is a look at the covariance matrix for the dyadic effects, the posterior summaries for the part of the model McElreath defined in the middle of page 463 as,</p>
<p><span class="math display">\[
\begin{bmatrix} d_{ij} \\ d_{ji} \end{bmatrix} \sim \operatorname{MVNormal}\begin{pmatrix} \begin{bmatrix} 0 \\ 0 \end{bmatrix}, \begin{bmatrix} \sigma_d^2 &amp; \sigma_d^2 \rho_d \\ \sigma_d^2 \rho_d &amp; \sigma_d^2 \end{bmatrix} \end{pmatrix},
\]</span></p>
<p>where there is only one standard deviation parameter <span class="math inline">\(\sigma_d\)</span> because the labels for each dyad are arbitrary. Here they are in the <code>precis()</code> output.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">precis</span>(m14<span class="fl">.7</span>, <span class="at">depth =</span> <span class="dv">3</span>, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"Rho_d"</span>, <span class="st">"sigma_d"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                mean         sd      5.5%     94.5%     rhat ess_bulk
Rho_d[1,1] 1.0000000 0.00000000 1.0000000 1.0000000       NA       NA
Rho_d[2,1] 0.8820644 0.03275936 0.8254698 0.9296409 1.002855 1110.151
Rho_d[1,2] 0.8820644 0.03275936 0.8254698 0.9296409 1.002855 1110.151
Rho_d[2,2] 1.0000000 0.00000000 1.0000000 1.0000000       NA       NA
sigma_d    1.1015718 0.05885262 1.0123212 1.1988938 1.004155 1217.318</code></pre>
</div>
</div>
<p>Here they are in a plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="st">`</span><span class="at">sigma[italic(d)]</span><span class="st">`</span> <span class="ot">=</span> post<span class="sc">$</span>sigma_d,</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>       <span class="st">`</span><span class="at">rho[italic(d)]</span><span class="st">`</span>   <span class="ot">=</span> post<span class="sc">$</span>Rho_d[, <span class="dv">2</span>, <span class="dv">1</span>]) <span class="sc">|&gt;</span> </span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">everything</span>()) <span class="sc">|&gt;</span> </span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">y =</span> name, <span class="at">fill =</span> name)) <span class="sc">+</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_halfeye</span>(<span class="at">.width =</span> <span class="fl">0.89</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span> </span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe) <span class="sc">+</span></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#A65141"</span>, <span class="st">"#B1934A"</span>)) <span class="sc">+</span></span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"marginal posterior"</span>) <span class="sc">+</span></span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="432"></p>
</figure>
</div>
</div>
</div>
<p>“The correlation here is positive and strong. And there is more variation among dyads than there is among household in giving rates” (p.&nbsp;467). Now make the right hand panel of Figure 14.9.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">dy1 =</span> <span class="fu">apply</span>(post<span class="sc">$</span>d[, , <span class="dv">1</span>], <span class="dv">2</span>, mean),</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">dy2 =</span> <span class="fu">apply</span>(post<span class="sc">$</span>d[, , <span class="dv">2</span>], <span class="dv">2</span>, mean)) <span class="sc">|&gt;</span> </span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> dy1, <span class="at">y =</span> dy2)) <span class="sc">+</span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">x =</span> <span class="fu">mean</span>(post<span class="sc">$</span>d[, <span class="dv">1</span>, <span class="dv">1</span>]),</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">y =</span> <span class="fu">mean</span>(post<span class="sc">$</span>d[, <span class="dv">1</span>, <span class="dv">2</span>]),</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="st">"1"</span>,</span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#EEDA9D"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"household A in dyad"</span>,</span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"household B in dyad"</span>) <span class="sc">+</span></span>
<span id="cb83-15"><a href="#cb83-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="fl">3.5</span>),</span>
<span id="cb83-16"><a href="#cb83-16" aria-hidden="true" tabindex="-1"></a>              <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="fl">3.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As McElreath pointed out, each dot is the posterior mean for one of the 300 levels of <code>did</code>. Do you see the yellow #1 toward the middle? It’s marking off the posterior mean for <code>did == 1</code>. To give a better sense of the uncertainty in each of the levels of <code>did</code>, here is the full bivariate distribution for <code>did == 1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">dy1 =</span> post<span class="sc">$</span>d[, <span class="dv">1</span>, <span class="dv">1</span>],</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">dy2 =</span> post<span class="sc">$</span>d[, <span class="dv">1</span>, <span class="dv">2</span>]) <span class="sc">|&gt;</span> </span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> dy1, <span class="at">y =</span> dy2)) <span class="sc">+</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">level =</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"#EEDA9D"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">type =</span> <span class="st">"norm"</span>, <span class="at">level =</span> <span class="fl">0.9</span>, <span class="at">color =</span> <span class="st">"#EEDA9D"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"household A in dyad"</span>[<span class="fu">italic</span>(i)<span class="sc">==</span><span class="dv">1</span>]),</span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">expression</span>(<span class="st">"household B in dyad"</span>[<span class="fu">italic</span>(i)<span class="sc">==</span><span class="dv">1</span>])) <span class="sc">+</span></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="fl">3.5</span>),</span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>              <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="fl">3.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The two yellow ellipses mark off the 50% and 90% intervals, again presuming a bivariate Gaussian distribution for the posterior.</p>
</section>
<section id="continuous-categories-and-the-gaussian-process" class="level2" data-number="14.5">
<h2 data-number="14.5" class="anchored" data-anchor-id="continuous-categories-and-the-gaussian-process"><span class="header-section-number">14.5</span> Continuous categories and the Gaussian process</h2>
<blockquote class="blockquote">
<p>There is a way to apply the varying effects approach to continuous categories… The general approach is known as <strong>Gaussian process regression</strong>. This name is unfortunately wholly uninformative about what it is for and how it works.</p>
<p>We’ll proceed to work through a basic example that demonstrates both what it is for and how it works. The general purpose is to define some dimension along which cases differ. This might be individual differences in age. Or it could be differences in location. Then we measure the distance between each pair of cases. What the model then does is estimate a function for the covariance between pairs of cases at different distances. This covariance function provides one continuous category generalization of the varying effects approach. (p.&nbsp;468, <strong>emphasis</strong> in the original)</p>
</blockquote>
<section id="example-spatial-autocorrelation-in-oceanic-tools" class="level3" data-number="14.5.1">
<h3 data-number="14.5.1" class="anchored" data-anchor-id="example-spatial-autocorrelation-in-oceanic-tools"><span class="header-section-number">14.5.1</span> Example: Spatial autocorrelation in Oceanic tools</h3>
<p>We start by loading the matrix of geographic distances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the distance matrix</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rethinking)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(islandsDistMatrix)</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Display (measured in thousands of km)</span></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>d_mat <span class="ot">&lt;-</span> islandsDistMatrix</span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(d_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ml"</span>, <span class="st">"Ti"</span>, <span class="st">"SC"</span>, <span class="st">"Ya"</span>, <span class="st">"Fi"</span>, <span class="st">"Tr"</span>, <span class="st">"Ch"</span>, <span class="st">"Mn"</span>, <span class="st">"To"</span>, <span class="st">"Ha"</span>)</span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(d_mat, <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Ml  Ti  SC  Ya  Fi  Tr  Ch  Mn  To  Ha
Malekula   0.0 0.5 0.6 4.4 1.2 2.0 3.2 2.8 1.9 5.7
Tikopia    0.5 0.0 0.3 4.2 1.2 2.0 2.9 2.7 2.0 5.3
Santa Cruz 0.6 0.3 0.0 3.9 1.6 1.7 2.6 2.4 2.3 5.4
Yap        4.4 4.2 3.9 0.0 5.4 2.5 1.6 1.6 6.1 7.2
Lau Fiji   1.2 1.2 1.6 5.4 0.0 3.2 4.0 3.9 0.8 4.9
Trobriand  2.0 2.0 1.7 2.5 3.2 0.0 1.8 0.8 3.9 6.7
Chuuk      3.2 2.9 2.6 1.6 4.0 1.8 0.0 1.2 4.8 5.8
Manus      2.8 2.7 2.4 1.6 3.9 0.8 1.2 0.0 4.6 6.7
Tonga      1.9 2.0 2.3 6.1 0.8 3.9 4.8 4.6 0.0 5.0
Hawaii     5.7 5.3 5.4 7.2 4.9 6.7 5.8 6.7 5.0 0.0</code></pre>
</div>
</div>
<p>If you wanted to use color to more effectively visualize the values in the matrix, you might do something like this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>d_mat <span class="sc">|&gt;</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"row"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gather</span>(column, distance, <span class="sc">-</span>row) <span class="sc">|&gt;</span> </span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">column =</span> <span class="fu">factor</span>(column, <span class="at">levels =</span> <span class="fu">colnames</span>(d_mat)),</span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">row    =</span> <span class="fu">factor</span>(row,    <span class="at">levels =</span> <span class="fu">rownames</span>(d_mat)) <span class="sc">|&gt;</span> <span class="fu">fct_rev</span>(),</span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">label  =</span> <span class="fu">formatC</span>(distance, <span class="at">format =</span> <span class="st">'f'</span>, <span class="at">digits =</span> <span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb87-8"><a href="#cb87-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb87-9"><a href="#cb87-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> column, <span class="at">y =</span> row)) <span class="sc">+</span> </span>
<span id="cb87-10"><a href="#cb87-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill =</span> distance)) <span class="sc">+</span> </span>
<span id="cb87-11"><a href="#cb87-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> label),</span>
<span id="cb87-12"><a href="#cb87-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb87-13"><a href="#cb87-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">position =</span> <span class="st">"top"</span>) <span class="sc">+</span></span>
<span id="cb87-14"><a href="#cb87-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb87-15"><a href="#cb87-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"#FCF9F0"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>) <span class="sc">+</span></span>
<span id="cb87-16"><a href="#cb87-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_earring</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb87-17"><a href="#cb87-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.ticks =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="528"></p>
</figure>
</div>
</div>
</div>
<p>Figure 14.10 shows the “shape of the function relating distance to the covariance <span class="math inline">\(\mathbf K_{ij}\)</span>.”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">x       =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">4</span>, <span class="at">by =</span> <span class="fl">0.01</span>),</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">linear  =</span> <span class="fu">exp</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> x),</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">squared =</span> <span class="fu">exp</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> x<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x)) <span class="sc">+</span></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> linear),</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#B1934A"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> squared),</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">"distance"</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="st">"correlation"</span>, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="312"></p>
</figure>
</div>
</div>
</div>
<p>Now load the primary data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Kline2)  <span class="co"># Load the ordinary data, now with coordinates</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> Kline2 <span class="sc">|&gt;</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">society =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(Kline2)</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>d <span class="sc">|&gt;</span> <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 10
Columns: 10
$ culture     &lt;fct&gt; Malekula, Tikopia, Santa Cruz, Yap, Lau Fiji, Trobriand, Chuuk, Manus, Tonga, Hawaii
$ population  &lt;int&gt; 1100, 1500, 3600, 4791, 7400, 8000, 9200, 13000, 17500, 275000
$ contact     &lt;fct&gt; low, low, low, high, high, high, high, low, high, low
$ total_tools &lt;int&gt; 13, 22, 24, 43, 33, 19, 40, 28, 55, 71
$ mean_TU     &lt;dbl&gt; 3.2, 4.7, 4.0, 5.0, 5.0, 4.0, 3.8, 6.6, 5.4, 6.6
$ lat         &lt;dbl&gt; -16.3, -12.3, -10.7, 9.5, -17.7, -8.7, 7.4, -2.1, -21.2, 19.9
$ lon         &lt;dbl&gt; 167.5, 168.8, 166.0, 138.1, 178.1, 150.9, 151.6, 146.9, -175.2, -155.6
$ lon2        &lt;dbl&gt; -12.5, -11.2, -14.0, -41.9, -1.9, -29.1, -28.4, -33.1, 4.8, 24.4
$ logpop      &lt;dbl&gt; 7.003065, 7.313220, 8.188689, 8.474494, 8.909235, 8.987197, 9.126959, 9.472705, 9.769956, 12.524526
$ society     &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10</code></pre>
</div>
</div>
<p>👋 <strong>Heads up</strong>: The <strong>brms</strong> package is capable of handling a variety of Gaussian process models using the <code>gp()</code> function. As we will see throughout this section, this method will depart in important ways from how McElreath fits Gaussian process models with <strong>rethinking</strong>. Due in large part to these differences, this section and its analogue in the first edition of <em>Statistical rethinking</em> <span class="citation" data-cites="mcelreathStatisticalRethinkingBayesian2015">(<a href="references.html#ref-mcelreathStatisticalRethinkingBayesian2015" role="doc-biblioref">McElreath, 2015</a>)</span> baffled me, at first. Happily, fellow enthusiasts <a href="https://twitter.com/LBliard">Louis Bliard</a> and <a href="https://torkar.github.io/">Richard Torkar</a> reached out and helped me hammer this section out behind the scenes. The method to follow is due in large part to their efforts. 🤝</p>
<p>The <code>brms::gp()</code> function takes a handful of arguments. The first and most important argument, <code>...</code>, accepts the names of one or more predictors from the data. When fitting a spatial Gaussian process of this kind, we’ll enter in the latitude and longitude data for each of levels of <code>culture</code>. This will be an important departure from the text. For his <code>m14.8</code>, McElreath directly entered in the <code>Dmat</code> distance matrix data into <code>ulam()</code>. In so doing, he defined <span class="math inline">\(D_{ij}\)</span>, the matrix of distances between each of the societies. When using <strong>brms</strong>, we instead <em>estimate</em> the distance matrix from the latitude and longitude variables.</p>
<p>Before we practice fitting a Gaussian process with the <code>brms::gp()</code> function, we’ll first need to think a little bit about our data. McElreath’s <code>Dmat</code> measured the distances in thousands of km. However, the <code>lat</code> and <code>lon2</code> variables in the data above are in decimal degrees, which means they need to be transformed to keep our model in the same metric as McElreath’s. Turns out that <a href="https://en.wikipedia.org/wiki/Decimal_degrees#:~:text=A%20value%20in%20decimal%20degrees,1.1132%20m%20at%20the%20equator.">one decimal degree is 111.32km (at the equator)</a>. Thus, we can turn both <code>lat</code> and <code>lon2</code> into 1,000 km units by multiplying each by 0.11132. Here’s the conversion.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> d <span class="sc">|&gt;</span> </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">lat_adj  =</span> lat  <span class="sc">*</span> <span class="fl">0.11132</span>,</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">lon2_adj =</span> lon2 <span class="sc">*</span> <span class="fl">0.11132</span>)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>d <span class="sc">|&gt;</span> </span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(culture, lat, lon2, lat_adj<span class="sc">:</span>lon2_adj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      culture   lat  lon2   lat_adj  lon2_adj
1    Malekula -16.3 -12.5 -1.814516 -1.391500
2     Tikopia -12.3 -11.2 -1.369236 -1.246784
3  Santa Cruz -10.7 -14.0 -1.191124 -1.558480
4         Yap   9.5 -41.9  1.057540 -4.664308
5    Lau Fiji -17.7  -1.9 -1.970364 -0.211508
6   Trobriand  -8.7 -29.1 -0.968484 -3.239412
7       Chuuk   7.4 -28.4  0.823768 -3.161488
8       Manus  -2.1 -33.1 -0.233772 -3.684692
9       Tonga -21.2   4.8 -2.359984  0.534336
10     Hawaii  19.9  24.4  2.215268  2.716208</code></pre>
</div>
</div>
<p>Note that because this conversion is valid <strong>at the equator</strong>, it is only an <em>approximation</em> for latitude and longitude coordinates for our island societies.</p>
<p>Now we’ve scaled our two spatial variables, the basic way to use them in a <strong>brms</strong> Gaussian process is including <code>gp(lat_adj, lon2_adj)</code> into the <code>formula</code> argument within the <code>brm()</code> function. Note however that one of the default <code>gp()</code> settings is <code>scale = TRUE</code>, which scales predictors so that the maximum distance between two points is 1. We don’t want this for our example, so we will set <code>scale = FALSE</code> instead.</p>
<p>Our Gaussian process model is an extension of the non-linear model from <a href="11.html#sec-Modeling-tool-innovation" class="quarto-xref"><span>Section 11.2.1.1</span></a>, <code>b11.11</code>. Thus our model here will also use the non-linear syntax. Here’s how we might use <strong>brms</strong> to fit our amended non-centered version of McElreath’s <code>m14.8</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.8</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d, </span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">poisson</span>(<span class="at">link =</span> <span class="st">"identity"</span>),</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bf</span>(total_tools <span class="sc">~</span> <span class="fu">exp</span>(a) <span class="sc">*</span> population<span class="sc">^</span>b <span class="sc">/</span> g,</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>     a <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fu">gp</span>(lat_adj, lon2_adj, <span class="at">scale =</span> <span class="cn">FALSE</span>),</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>     b <span class="sc">+</span> g <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">nl =</span> <span class="cn">TRUE</span>),</span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">nlpar =</span> a),</span>
<span id="cb93-9"><a href="#cb93-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">nlpar =</span> b, <span class="at">lb =</span> <span class="dv">0</span>),</span>
<span id="cb93-10"><a href="#cb93-10" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">nlpar =</span> g, <span class="at">lb =</span> <span class="dv">0</span>),</span>
<span id="cb93-11"><a href="#cb93-11" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">inv_gamma</span>(<span class="fl">2.874624</span>, <span class="fl">2.941204</span>), <span class="at">class =</span> lscale, <span class="at">coef =</span> gplat_adjlon2_adj, <span class="at">nlpar =</span> a),</span>
<span id="cb93-12"><a href="#cb93-12" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sdgp, <span class="at">coef =</span> gplat_adjlon2_adj, <span class="at">nlpar =</span> a)),</span>
<span id="cb93-13"><a href="#cb93-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb93-14"><a href="#cb93-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb93-15"><a href="#cb93-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_prior =</span> T,</span>
<span id="cb93-16"><a href="#cb93-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.08"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: poisson 
  Links: mu = identity 
Formula: total_tools ~ exp(a) * population^b/g 
         a ~ 1 + gp(lat_adj, lon2_adj, scale = FALSE)
         b ~ 1
         g ~ 1
   Data: d (Number of observations: 10) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Gaussian Process Hyperparameters:
                            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sdgp(a_gplat_adjlon2_adj)       0.48      0.31     0.15     1.33 1.00     1184     1975
lscale(a_gplat_adjlon2_adj)     1.66      0.94     0.51     4.11 1.00     1158     1785

Regression Coefficients:
            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
a_Intercept     0.31      0.87    -1.44     1.95 1.00     2731     2632
b_Intercept     0.26      0.08     0.10     0.43 1.00     1656     1551
g_Intercept     0.65      0.63     0.06     2.44 1.00     2013     2258

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>The <code>posterior_summary()</code> function will return a summary that looks more like the one in the text.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior_summary</span>(b14<span class="fl">.8</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, ] <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            Estimate Est.Error  Q2.5 Q97.5
b_a_Intercept                   0.31      0.87 -1.44  1.95
b_b_Intercept                   0.26      0.08  0.10  0.43
b_g_Intercept                   0.65      0.63  0.06  2.44
sdgp_a_gplat_adjlon2_adj        0.48      0.31  0.15  1.33
lscale_a_gplat_adjlon2_adj      1.66      0.94  0.51  4.11
zgp_a_gplat_adjlon2_adj[1]     -0.48      0.76 -2.02  0.93
zgp_a_gplat_adjlon2_adj[2]      0.44      0.87 -1.31  2.14
zgp_a_gplat_adjlon2_adj[3]     -0.63      0.71 -2.00  0.86
zgp_a_gplat_adjlon2_adj[4]      0.96      0.68 -0.30  2.39
zgp_a_gplat_adjlon2_adj[5]      0.20      0.76 -1.32  1.66
zgp_a_gplat_adjlon2_adj[6]     -1.03      0.77 -2.47  0.52
zgp_a_gplat_adjlon2_adj[7]      0.16      0.72 -1.35  1.51
zgp_a_gplat_adjlon2_adj[8]     -0.21      0.85 -1.84  1.52
zgp_a_gplat_adjlon2_adj[9]      0.43      0.93 -1.51  2.17
zgp_a_gplat_adjlon2_adj[10]    -0.41      0.79 -1.99  1.15</code></pre>
</div>
</div>
<p>Let’s focus on our three non-linear parameters, first. Happily, both our <code>b_b_Intercept</code> and <code>b_g_Intercept</code> summaries look a lot like those for McElreath’s <code>b</code> and <code>g</code>, respectively. Our <code>b_a_Intercept</code> might look distressingly small, but that’s just because of how we parameterized our model. It’s actually very close to McElreath’s <code>a</code> after you exponentiate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fixef</span>(b14<span class="fl">.8</span>, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.055</span>, <span class="fl">0.945</span>))[<span class="st">"a_Intercept"</span>, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>)] <span class="sc">|&gt;</span> </span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">exp</span>() <span class="sc">|&gt;</span> </span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Estimate     Q5.5    Q94.5 
    1.37     0.35     5.34 </code></pre>
</div>
</div>
<p>Our Gaussian process parameters are different from McElreath’s. From the <code>gp</code> section of the <a href="https://cran.r-project.org/package=brms/brms.pdf"><strong>brms</strong> reference manual</a> <span class="citation" data-cites="brms2022RM">(<a href="references.html#ref-brms2022RM" role="doc-biblioref">Bürkner, 2022a</a>)</span>, we learn the <strong>brms</strong> parameterization follows the form</p>
<p><span class="math display">\[k(x_{i},x_{j}) = sdgp^2 \exp \big (-||x_i - x_j||^2 / (2 lscale^2) \big ),\]</span></p>
<p>where <span class="math inline">\(k(x_{i},x_{j})\)</span> is the same as McElreath’s <span class="math inline">\(\mathbf K_{ij}\)</span> and <span class="math inline">\(||x_i - x_j||^2\)</span> is the Euclidean distance, the same as McElreath’s <span class="math inline">\(D_{ij}^2\)</span>. Thus we could also express the <strong>brms</strong> parameterization as</p>
<p><span class="math display">\[\mathbf K_{ij} = sdgp^2 \exp \big (-D_{ij}^2 / (2 lscale^2) \big ),\]</span></p>
<p>which is much closer to McElreath’s</p>
<p><span class="math display">\[\mathbf K_{ij} = \eta^2 \exp \big (-\rho^2 D_{ij}^2 \big ) + \delta_{ij} \sigma^2\]</span></p>
<p>On page 470, McElreath explained that the final <span class="math inline">\(\delta_{ij} \sigma^2\)</span> term is mute with the Oceanic societies data. Thus we won’t consider it further. This reduces McElreath’s equation to</p>
<p><span class="math display">\[\mathbf K_{ij} = \eta^2 \exp \big (-\rho^2 D_{ij}^2 \big ).\]</span></p>
<p>Importantly, what McElreath called <span class="math inline">\(\eta\)</span>, Bürkner called <span class="math inline">\(sdgp\)</span>. While McElreath estimated <span class="math inline">\(\eta^2\)</span>, <strong>brms</strong> simply estimated <span class="math inline">\(sdgp\)</span>. So we’ll have to square our <code>sdgp(a_gplat_adjlon2_adj)</code> before it’s on the same scale as <code>etasq</code> in the text. Here it is.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>post <span class="ot">&lt;-</span> <span class="fu">as_draws_df</span>(b14<span class="fl">.8</span>) <span class="sc">|&gt;</span> </span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">etasq =</span> sdgp_a_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>post <span class="sc">|&gt;</span> </span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean_hdi</span>(etasq, <span class="at">.width =</span> <span class="fl">0.89</span>) <span class="sc">|&gt;</span> </span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_if</span>(is.double, round, <span class="at">digits =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Dropping 'draws_df' class as required metadata was removed.
Warning: Dropping 'draws_df' class as required metadata was removed.
Warning: Dropping 'draws_df' class as required metadata was removed.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  etasq .lower .upper .width .point .interval
  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    
1 0.325  0      0.619   0.89 mean   hdi      
2 0.325  0.649  0.675   0.89 mean   hdi      
3 0.325  0.754  0.756   0.89 mean   hdi      </code></pre>
</div>
</div>
<p>Though our posterior is a little bit larger than McElreath’s, we’re in the ballpark. You may have noticed that in our model <code>brm()</code> code, above, we just went with the flow and kept the <code>exponential(1)</code> prior on <code>sdgp</code>. The <strong>brms</strong> default would have been <code>student_t(3, 0, 15.6)</code>.</p>
<p>Now look at the denominator of the inner part of Bürkner’s equation, <span class="math inline">\(2 lscale^2\)</span>. This appears to be the <strong>brms</strong> equivalent to McElreath’s <span class="math inline">\(\rho^2\)</span>. Or at least it’s what we’ve got. Anyway, also note that McElreath estimated <span class="math inline">\(\rho^2\)</span> directly as <code>rhosq</code>. If I’m doing the algebra correctly, we might expect</p>
<p><span class="math display">\[\begin{align*}
\rho^2 &amp; = 1/(2 \cdot lscale^2) &amp; \text{and thus} \\
lscale &amp; = \sqrt{1 / (2 \cdot \rho^2)}.
\end{align*}\]</span></p>
<p>To get a sense of this relation, it might be helpful to plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="st">`</span><span class="at">rho^2</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">11</span>, <span class="at">by =</span> <span class="fl">0.01</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">lscale =</span> <span class="fu">sqrt</span>(<span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="st">`</span><span class="at">rho^2</span><span class="st">`</span>))) <span class="sc">|&gt;</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">rho^2</span><span class="st">`</span>, <span class="at">y =</span> lscale)) <span class="sc">+</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"#A65141"</span>) <span class="sc">+</span></span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">expression</span>(rho<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>),</span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>))</span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">lscale =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">11</span>, <span class="at">by =</span> <span class="fl">0.01</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb103-13"><a href="#cb103-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">rho^2</span><span class="st">`</span> <span class="ot">=</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> lscale<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb103-14"><a href="#cb103-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-15"><a href="#cb103-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> lscale, <span class="at">y =</span> <span class="st">`</span><span class="at">rho^2</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb103-16"><a href="#cb103-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb103-17"><a href="#cb103-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb103-18"><a href="#cb103-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"#80A0C7"</span>) <span class="sc">+</span></span>
<span id="cb103-19"><a href="#cb103-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(rho<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb103-20"><a href="#cb103-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>),</span>
<span id="cb103-21"><a href="#cb103-21" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>))</span>
<span id="cb103-22"><a href="#cb103-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-23"><a href="#cb103-23" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-64-1.png" class="img-fluid figure-img" width="624"></p>
</figure>
</div>
</div>
</div>
<p>The two aren’t quite inverses of one another, but the overall pattern is when one is large, the other is small. Now we have a sense of how they compare and how to covert one to the other, let’s see how our posterior for <span class="math inline">\(lscale\)</span> looks when we convert it to the scale of McElreath’s <span class="math inline">\(\rho^2\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>post <span class="ot">&lt;-</span> post <span class="sc">|&gt;</span> </span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">rhosq =</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> lscale_a_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>post <span class="sc">|&gt;</span> </span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean_hdi</span>(rhosq, <span class="at">.width =</span> <span class="fl">0.89</span>) <span class="sc">|&gt;</span> </span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_if</span>(is.double, round, <span class="at">digits =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  rhosq .lower .upper .width .point .interval
  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    
1 0.421   0.01  0.916   0.89 mean   hdi      </code></pre>
</div>
</div>
<p>This is about a third of the size of the McElreath’s <span class="math inline">\(\rho^2 = 1.31, 89 \text{% HDI } [0.08, 4.41]\)</span>. The plot deepens. If you look back, you’ll see we used a very different prior for <code>lscale</code>. Here it is: <code>inv_gamma(2.874624, 2.941204)</code>. Use <code>get_prior()</code> to discover where that came from.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_prior</span>(</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d, </span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">poisson</span>(<span class="at">link =</span> <span class="st">"identity"</span>),</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bf</span>(total_tools <span class="sc">~</span> <span class="fu">exp</span>(a) <span class="sc">*</span> population<span class="sc">^</span>b <span class="sc">/</span> g,</span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>     a  <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fu">gp</span>(lat_adj, lon2_adj, <span class="at">scale =</span> <span class="cn">FALSE</span>),</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>     b <span class="sc">+</span> g <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">nl =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         prior  class              coef group resp dpar nlpar lb ub tag       source
                        (flat)      b                                       a                default
                        (flat)      b         Intercept                     a           (vectorized)
                        (flat) lscale                                       a  0             default
 inv_gamma(2.874624, 2.941204) lscale gplat_adjlon2_adj                     a  0             default
         student_t(3, 0, 15.6)   sdgp                                       a  0             default
         student_t(3, 0, 15.6)   sdgp gplat_adjlon2_adj                     a  0        (vectorized)
                        (flat)      b                                       b                default
                        (flat)      b         Intercept                     b           (vectorized)
                        (flat)      b                                       g                default
                        (flat)      b         Intercept                     g           (vectorized)</code></pre>
</div>
</div>
<p>That is, we used the <strong>brms</strong> default prior for <span class="math inline">\(lscale\)</span>. In a <a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse/issues/8">GitHub exchange</a>, Bürkner pointed out that <strong>brms</strong> uses special priors for <span class="math inline">\(lscale\)</span> parameters based on Michael Betancourt’s <span class="citation" data-cites="betancourtRobustGaussianProcesses2017">(<a href="references.html#ref-betancourtRobustGaussianProcesses2017" role="doc-biblioref">2017</a>)</span> vignette, <a href="https://betanalpha.github.io/assets/case_studies/gp_part3/part3.html"><em>Robust Gaussian processes in Stan</em></a>. We can use the <code>dinvgamma()</code> function from the well-named <a href="https://CRAN.R-project.org/package=invgamma"><strong>invgamma</strong> package</a> <span class="citation" data-cites="R-invgamma">(<a href="references.html#ref-R-invgamma" role="doc-biblioref">Kahle &amp; Stamey, 2017</a>)</span> to get a sense of what that prior looks like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">lscale =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="fl">0.01</span>, <span class="at">to =</span> <span class="dv">9</span>, <span class="at">by =</span> <span class="fl">0.01</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">density =</span> invgamma<span class="sc">::</span><span class="fu">dinvgamma</span>(lscale, <span class="at">shape =</span> <span class="fl">2.874624</span>, <span class="at">rate =</span> <span class="fl">2.941204</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> lscale, <span class="at">y =</span> density)) <span class="sc">+</span></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_area</span>(<span class="at">fill =</span> <span class="st">"#80A0C7"</span>) <span class="sc">+</span></span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">"text"</span>, </span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="fl">4.75</span>, <span class="at">y =</span> <span class="fl">0.75</span>,</span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">label =</span> <span class="st">"inverse gamma(2.874624, 2.941204)"</span>,</span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb108-10"><a href="#cb108-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb108-11"><a href="#cb108-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="384"></p>
</figure>
</div>
</div>
</div>
<p>Anyways, let’s make the subplots for our version of Figure 14.11 to get a sense of what this all means. Start with the left panel, the prior predictive distribution for the covariance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For `slice_sample()`</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Wrangle</span></span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">prior_draws</span>(b14<span class="fl">.8</span>) <span class="sc">|&gt;</span> </span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transmute</span>(<span class="at">draw  =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">n</span>(),</span>
<span id="cb109-7"><a href="#cb109-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">etasq =</span> sdgp_a<span class="sc">^</span><span class="dv">2</span>,</span>
<span id="cb109-8"><a href="#cb109-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">rhosq =</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> lscale_a__1_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb109-9"><a href="#cb109-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">100</span>) <span class="sc">|&gt;</span></span>
<span id="cb109-10"><a href="#cb109-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">10</span>, <span class="at">by =</span> <span class="fl">0.05</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb109-11"><a href="#cb109-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">covariance =</span> etasq <span class="sc">*</span> <span class="fu">exp</span>(<span class="sc">-</span>rhosq <span class="sc">*</span> x<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb109-12"><a href="#cb109-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb109-13"><a href="#cb109-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot</span></span>
<span id="cb109-14"><a href="#cb109-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> covariance)) <span class="sc">+</span></span>
<span id="cb109-15"><a href="#cb109-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> draw),</span>
<span id="cb109-16"><a href="#cb109-16" aria-hidden="true" tabindex="-1"></a>            <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">color =</span> <span class="st">"#EEDA9D"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb109-17"><a href="#cb109-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">"distance (thousand km)"</span>, </span>
<span id="cb109-18"><a href="#cb109-18" aria-hidden="true" tabindex="-1"></a>                     <span class="at">breaks =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">5</span> <span class="sc">*</span> <span class="dv">2</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb109-19"><a href="#cb109-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>),</span>
<span id="cb109-20"><a href="#cb109-20" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb109-21"><a href="#cb109-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">subtitle =</span> <span class="st">"Gaussian process prior"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now make the right panel, the posterior distribution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For `slice_sample()`</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Wrangle</span></span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> post <span class="sc">|&gt;</span> </span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">etasq =</span> sdgp_a_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>,</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">rhosq =</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> lscale_a_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">50</span>) <span class="sc">|&gt;</span> </span>
<span id="cb110-9"><a href="#cb110-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">10</span>, <span class="at">by =</span> <span class="fl">0.05</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb110-10"><a href="#cb110-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">covariance =</span> etasq <span class="sc">*</span> <span class="fu">exp</span>(<span class="sc">-</span>rhosq <span class="sc">*</span> x<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb110-11"><a href="#cb110-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb110-12"><a href="#cb110-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot</span></span>
<span id="cb110-13"><a href="#cb110-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> covariance)) <span class="sc">+</span></span>
<span id="cb110-14"><a href="#cb110-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> .draw),</span>
<span id="cb110-15"><a href="#cb110-15" aria-hidden="true" tabindex="-1"></a>            <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">color =</span> <span class="st">"#EEDA9D"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb110-16"><a href="#cb110-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_function</span>(<span class="at">fun =</span> <span class="cf">function</span>(x) <span class="fu">mean</span>(post<span class="sc">$</span>sdgp_a_gplat_adjlon2_adj)<span class="sc">^</span><span class="dv">2</span> <span class="sc">*</span></span>
<span id="cb110-17"><a href="#cb110-17" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">exp</span>(<span class="sc">-</span>(<span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">mean</span>(post<span class="sc">$</span>lscale_a_gplat_adjlon2_adj)<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">*</span> x<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb110-18"><a href="#cb110-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">color =</span> <span class="st">"#DCA258"</span>, <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb110-19"><a href="#cb110-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">"distance (thousand km)"</span>,</span>
<span id="cb110-20"><a href="#cb110-20" aria-hidden="true" tabindex="-1"></a>                     <span class="at">breaks =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">5</span> <span class="sc">*</span> <span class="dv">2</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb110-21"><a href="#cb110-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>),</span>
<span id="cb110-22"><a href="#cb110-22" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb110-23"><a href="#cb110-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">subtitle =</span> <span class="st">"Gaussian process posterior"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Combine the two with <strong>patchwork</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">|</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Though the Gaussian process parameters from our <strong>brms</strong> parameterization looked different from McElreath’s, they resulted in a similar decline in spatial covariance.</p>
<p>Let’s finish this up and “push the parameters back through the function for <span class="math inline">\(\mathbf{K}\)</span>, the covariance matrix” (p.&nbsp;473).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute posterior median covariance among societies</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="dv">10</span>, <span class="at">ncol =</span> <span class="dv">10</span>)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>        k[i, j] <span class="ot">&lt;-</span> <span class="fu">median</span>(post<span class="sc">$</span>etasq) <span class="sc">*</span> </span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">exp</span>(<span class="sc">-</span><span class="fu">median</span>(post<span class="sc">$</span>rhosq) <span class="sc">*</span> islandsDistMatrix[i, j]<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-8"><a href="#cb112-8" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(k) <span class="ot">&lt;-</span> <span class="fu">median</span>(post<span class="sc">$</span>etasq) <span class="sc">+</span> <span class="fl">0.01</span></span>
<span id="cb112-9"><a href="#cb112-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-10"><a href="#cb112-10" aria-hidden="true" tabindex="-1"></a>k <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
 [1,] 0.17 0.15 0.14 0.00 0.11 0.06 0.01 0.02 0.07  0.00
 [2,] 0.15 0.17 0.15 0.00 0.11 0.06 0.02 0.03 0.06  0.00
 [3,] 0.14 0.15 0.17 0.00 0.09 0.08 0.03 0.04 0.05  0.00
 [4,] 0.00 0.00 0.00 0.17 0.00 0.04 0.09 0.08 0.00  0.00
 [5,] 0.11 0.11 0.09 0.00 0.17 0.01 0.00 0.00 0.14  0.00
 [6,] 0.06 0.06 0.08 0.04 0.01 0.17 0.07 0.13 0.00  0.00
 [7,] 0.01 0.02 0.03 0.09 0.00 0.07 0.17 0.11 0.00  0.00
 [8,] 0.02 0.03 0.04 0.08 0.00 0.13 0.11 0.17 0.00  0.00
 [9,] 0.07 0.06 0.05 0.00 0.14 0.00 0.00 0.00 0.17  0.00
[10,] 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00  0.17</code></pre>
</div>
</div>
<p>We’ll continue to follow suit and change these to a correlation matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to correlation matrix</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">cov2cor</span>(k), <span class="dv">2</span>)</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add row/col names for convenience</span></span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(rho) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ml"</span>, <span class="st">"Ti"</span>, <span class="st">"SC"</span>, <span class="st">"Ya"</span>, <span class="st">"Fi"</span>, <span class="st">"Tr"</span>, <span class="st">"Ch"</span>, <span class="st">"Mn"</span>, <span class="st">"To"</span>, <span class="st">"Ha"</span>)</span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(rho) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(rho)</span>
<span id="cb114-7"><a href="#cb114-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-8"><a href="#cb114-8" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ml   Ti   SC   Ya   Fi   Tr   Ch   Mn   To Ha
Ml 1.00 0.89 0.86 0.01 0.65 0.35 0.08 0.15 0.41  0
Ti 0.89 1.00 0.92 0.01 0.65 0.36 0.13 0.17 0.37  0
SC 0.86 0.92 1.00 0.03 0.53 0.47 0.19 0.25 0.27  0
Ya 0.01 0.01 0.03 1.00 0.00 0.22 0.53 0.50 0.00  0
Fi 0.65 0.65 0.53 0.00 1.00 0.08 0.02 0.02 0.82  0
Tr 0.35 0.36 0.47 0.22 0.08 1.00 0.43 0.79 0.03  0
Ch 0.08 0.13 0.19 0.53 0.02 0.43 1.00 0.66 0.00  0
Mn 0.15 0.17 0.25 0.50 0.02 0.79 0.66 1.00 0.01  0
To 0.41 0.37 0.27 0.00 0.82 0.03 0.00 0.01 1.00  0
Ha 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00  1</code></pre>
</div>
</div>
<p>Here are those correlations in a plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">row =</span> d<span class="sc">$</span>culture) <span class="sc">|&gt;</span> </span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>row, <span class="at">values_to =</span> <span class="st">"distance"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">column =</span> <span class="fu">factor</span>(name, <span class="at">levels =</span> <span class="fu">colnames</span>(d_mat)),</span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">row    =</span> <span class="fu">factor</span>(row, <span class="at">levels =</span> <span class="fu">rownames</span>(d_mat)) <span class="sc">|&gt;</span> <span class="fu">fct_rev</span>(),</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">label  =</span> <span class="fu">formatC</span>(distance, <span class="at">format =</span> <span class="st">'f'</span>, <span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">"0."</span>, <span class="st">"."</span>)) <span class="sc">|&gt;</span></span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Omit this line to keep the diagonal of 1's</span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(distance <span class="sc">!=</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> column, <span class="at">y =</span> row)) <span class="sc">+</span> </span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill =</span> distance)) <span class="sc">+</span> </span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> label),</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="fl">2.75</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">color =</span> <span class="st">"#100F14"</span>) <span class="sc">+</span></span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">position =</span> <span class="st">"top"</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="fu">expression</span>(rho), <span class="at">low =</span> <span class="st">"#FCF9F0"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb116-18"><a href="#cb116-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_earring</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb116-19"><a href="#cb116-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.ticks =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid figure-img" width="456"></p>
</figure>
</div>
</div>
</div>
<p>The correlations in our <code>rho</code> matrix look a little higher than those in the text (p.&nbsp;474). Before we move on to the next plot, let’s consider <code>psize</code>. If you really want to scale the points in Figure 14.12.a like McElreath did, you can make the <code>psize</code> variable in a <strong>tidyverse</strong> sort of way as follows. However, if you compare the <code>psize</code> method and the default <strong>ggplot2</strong> method using just <code>logpop</code>, you’ll see the difference is negligible. In that light, I’m going to be lazy and just use <code>logpop</code> in my plots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>d <span class="sc">|&gt;</span> </span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transmute</span>(<span class="at">psize =</span> logpop <span class="sc">/</span> <span class="fu">max</span>(logpop)) <span class="sc">|&gt;</span> </span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transmute</span>(<span class="at">psize =</span> <span class="fu">exp</span>(psize <span class="sc">*</span> <span class="fl">1.5</span>) <span class="sc">-</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       psize
1  0.3134090
2  0.4009582
3  0.6663711
4  0.7592196
5  0.9066890
6  0.9339560
7  0.9834797
8  1.1096138
9  1.2223112
10 2.4816891</code></pre>
</div>
</div>
<p>As far as I can figure, you still have to get <code>rho</code> into a tidy data frame before feeding it into <strong>ggplot2</strong>. Here’s my attempt at doing so.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>tidy_rho <span class="ot">&lt;-</span> rho <span class="sc">|&gt;</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">|&gt;</span> </span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(d <span class="sc">|&gt;</span> <span class="fu">select</span>(culture, logpop, total_tools, lon2, lat)) <span class="sc">|&gt;</span> </span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(Ml<span class="sc">:</span>Ha,</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"colname"</span>, </span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"correlation"</span>) <span class="sc">|&gt;</span></span>
<span id="cb119-8"><a href="#cb119-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">str_c</span>(<span class="fu">pmin</span>(rowname, colname), <span class="fu">pmax</span>(rowname, colname))) <span class="sc">|&gt;</span></span>
<span id="cb119-9"><a href="#cb119-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(rowname, colname, group, culture, <span class="fu">everything</span>())  </span>
<span id="cb119-10"><a href="#cb119-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-11"><a href="#cb119-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tidy_rho)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  rowname colname group culture  logpop total_tools  lon2   lat correlation
  &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt; &lt;fct&gt;     &lt;dbl&gt;       &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;       &lt;dbl&gt;
1 Ml      Ml      MlMl  Malekula   7.00          13 -12.5 -16.3        1   
2 Ml      Ti      MlTi  Malekula   7.00          13 -12.5 -16.3        0.89
3 Ml      SC      MlSC  Malekula   7.00          13 -12.5 -16.3        0.86
4 Ml      Ya      MlYa  Malekula   7.00          13 -12.5 -16.3        0.01
5 Ml      Fi      FiMl  Malekula   7.00          13 -12.5 -16.3        0.65
6 Ml      Tr      MlTr  Malekula   7.00          13 -12.5 -16.3        0.35</code></pre>
</div>
</div>
<p>Okay, here’s the code for our version of Figure 14.12.a.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> tidy_rho <span class="sc">|&gt;</span>       </span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> lon2, <span class="at">y =</span> lat)) <span class="sc">+</span></span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> d, </span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">size =</span> logpop), <span class="at">color =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> group, <span class="at">alpha =</span> correlation<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> d, <span class="fu">aes</span>(<span class="at">label =</span> culture), </span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">point.padding =</span> <span class="fl">0.2</span>, <span class="at">seed =</span> <span class="dv">14</span>, <span class="at">size =</span> <span class="fl">2.75</span>) <span class="sc">+</span></span>
<span id="cb121-11"><a href="#cb121-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_alpha_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb121-12"><a href="#cb121-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"longitude"</span>,</span>
<span id="cb121-13"><a href="#cb121-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"latitude"</span>,</span>
<span id="cb121-14"><a href="#cb121-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Among societies in geographic space</span><span class="sc">\n</span><span class="st">"</span>) <span class="sc">+</span></span>
<span id="cb121-15"><a href="#cb121-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">range</span>(d<span class="sc">$</span>lon2),</span>
<span id="cb121-16"><a href="#cb121-16" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">range</span>(d<span class="sc">$</span>lat)) <span class="sc">+</span></span>
<span id="cb121-17"><a href="#cb121-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here’s our the code for our version of Figure 14.12.b.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the average posterior predictive relationship between </span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="co"># log population and total tools, summarized by the median and 80% interval</span></span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> post <span class="sc">|&gt;</span> </span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(<span class="at">logpop =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">6</span>, <span class="at">to =</span> <span class="dv">14</span>, <span class="at">length.out =</span> <span class="dv">30</span>)) <span class="sc">|&gt;</span></span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">population =</span> <span class="fu">exp</span>(logpop)) <span class="sc">|&gt;</span> </span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">lambda =</span> <span class="fu">exp</span>(b_a_Intercept) <span class="sc">*</span> population<span class="sc">^</span>b_b_Intercept <span class="sc">/</span> b_g_Intercept) <span class="sc">|&gt;</span></span>
<span id="cb122-7"><a href="#cb122-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(logpop) <span class="sc">|&gt;</span> </span>
<span id="cb122-8"><a href="#cb122-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">median_qi</span>(lambda, <span class="at">.width =</span> <span class="fl">0.8</span>)</span>
<span id="cb122-9"><a href="#cb122-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-10"><a href="#cb122-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb122-11"><a href="#cb122-11" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> tidy_rho <span class="sc">|&gt;</span> </span>
<span id="cb122-12"><a href="#cb122-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> logpop)) <span class="sc">+</span></span>
<span id="cb122-13"><a href="#cb122-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">data =</span> f,</span>
<span id="cb122-14"><a href="#cb122-14" aria-hidden="true" tabindex="-1"></a>              <span class="fu">aes</span>(<span class="at">y =</span> lambda, <span class="at">ymin =</span> .lower, <span class="at">ymax =</span> .upper),</span>
<span id="cb122-15"><a href="#cb122-15" aria-hidden="true" tabindex="-1"></a>              <span class="at">stat =</span> <span class="st">"identity"</span>,</span>
<span id="cb122-16"><a href="#cb122-16" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">fill =</span> <span class="st">"#394165"</span>, <span class="at">linewidth =</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb122-17"><a href="#cb122-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> d, </span>
<span id="cb122-18"><a href="#cb122-18" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">y =</span> total_tools, <span class="at">size =</span> logpop), </span>
<span id="cb122-19"><a href="#cb122-19" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb122-20"><a href="#cb122-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> total_tools, <span class="at">group =</span> group, <span class="at">alpha =</span> correlation<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb122-21"><a href="#cb122-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb122-22"><a href="#cb122-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> d, </span>
<span id="cb122-23"><a href="#cb122-23" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">y =</span> total_tools, <span class="at">label =</span> culture), </span>
<span id="cb122-24"><a href="#cb122-24" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">point.padding =</span> <span class="fl">0.2</span>, <span class="at">seed =</span> <span class="dv">14</span>, <span class="at">size =</span> <span class="fl">2.75</span>) <span class="sc">+</span></span>
<span id="cb122-25"><a href="#cb122-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_alpha_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb122-26"><a href="#cb122-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"log population"</span>,</span>
<span id="cb122-27"><a href="#cb122-27" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"total tools"</span>,</span>
<span id="cb122-28"><a href="#cb122-28" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Shown against the relation between</span><span class="sc">\n</span><span class="st">total tools and log pop"</span>) <span class="sc">+</span></span>
<span id="cb122-29"><a href="#cb122-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">range</span>(d<span class="sc">$</span>logpop),</span>
<span id="cb122-30"><a href="#cb122-30" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">range</span>(d<span class="sc">$</span>total_tools)) <span class="sc">+</span></span>
<span id="cb122-31"><a href="#cb122-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we combine them to make the full version of Figure 14.12.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> </span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Posterior median correlations"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-78-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>As expressed by the intensity of the colors of those connecting lines, our correlations are a bit more pronounced than those in the text, making for a more densely webbed plot. It is still the case that the correlations among Malekula, Tikopia, and Santa Cruz are the most pronounced. The next two notable correlations are between Manus and Trobriand and between Lau Fiji and Tonga.</p>
<section id="rethinking-dispersion-by-other-names" class="level5" data-number="14.5.1.0.1">
<h5 data-number="14.5.1.0.1" class="anchored" data-anchor-id="rethinking-dispersion-by-other-names"><span class="header-section-number">14.5.1.0.1</span> Rethinking: Dispersion by other names</h5>
<p>McElreath remarked it might be a good idea to fit an alternative of this model using the gamma-Poisson likelihood. Let’s take him up on the challenge. Remember that gamma-Poisson models are also referred to as negative binomial models. When using <strong>brms</strong>, you specify this using <code>family = negbinomial</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.8</span>_nb <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d, </span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">negbinomial</span>(<span class="at">link =</span> <span class="st">"identity"</span>),</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bf</span>(total_tools <span class="sc">~</span> <span class="fu">exp</span>(a) <span class="sc">*</span> population<span class="sc">^</span>b <span class="sc">/</span> g,</span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a>     a <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fu">gp</span>(lat_adj, lon2_adj, <span class="at">scale =</span> <span class="cn">FALSE</span>),</span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a>     b <span class="sc">+</span> g <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb124-7"><a href="#cb124-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">nl =</span> <span class="cn">TRUE</span>),</span>
<span id="cb124-8"><a href="#cb124-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">nlpar =</span> a),</span>
<span id="cb124-9"><a href="#cb124-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">nlpar =</span> b, <span class="at">lb =</span> <span class="dv">0</span>),</span>
<span id="cb124-10"><a href="#cb124-10" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">nlpar =</span> g, <span class="at">lb =</span> <span class="dv">0</span>),</span>
<span id="cb124-11"><a href="#cb124-11" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">inv_gamma</span>(<span class="fl">2.874624</span>, <span class="fl">2.941204</span>), <span class="at">class =</span> lscale, <span class="at">coef =</span> gplat_adjlon2_adj, <span class="at">nlpar =</span> a),</span>
<span id="cb124-12"><a href="#cb124-12" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sdgp, <span class="at">coef =</span> gplat_adjlon2_adj, <span class="at">nlpar =</span> a),</span>
<span id="cb124-13"><a href="#cb124-13" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Default prior</span></span>
<span id="cb124-14"><a href="#cb124-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">inv_gamma</span>(<span class="fl">0.4</span>, <span class="fl">0.3</span>), <span class="at">class =</span> shape)),</span>
<span id="cb124-15"><a href="#cb124-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb124-16"><a href="#cb124-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb124-17"><a href="#cb124-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">control =</span> <span class="fu">list</span>(<span class="at">adapt_delta =</span> <span class="fl">0.95</span>),</span>
<span id="cb124-18"><a href="#cb124-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.08_nb"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the summary.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.8</span>_nb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: negbinomial 
  Links: mu = identity 
Formula: total_tools ~ exp(a) * population^b/g 
         a ~ 1 + gp(lat_adj, lon2_adj, scale = FALSE)
         b ~ 1
         g ~ 1
   Data: d (Number of observations: 10) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Gaussian Process Hyperparameters:
                            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sdgp(a_gplat_adjlon2_adj)       0.40      0.31     0.02     1.19 1.00      836     1333
lscale(a_gplat_adjlon2_adj)     1.69      1.21     0.50     4.59 1.00     1937     2298

Regression Coefficients:
            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
a_Intercept     0.33      0.87    -1.44     1.97 1.00     2891     1919
b_Intercept     0.27      0.09     0.10     0.44 1.00     1876     1273
g_Intercept     0.69      0.64     0.06     2.35 1.00     2585     2084

Further Distributional Parameters:
           Estimate       Est.Error l-95% CI  u-95% CI Rhat Bulk_ESS Tail_ESS
shape 4472764831.05 268871223158.15     4.12 128492.12 1.00      833     1163

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>This resulted in a slightly less pronounced correlation matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>post <span class="ot">&lt;-</span> <span class="fu">as_draws_df</span>(b14<span class="fl">.8</span>_nb)</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute posterior median covariance among societies</span></span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="dv">10</span>, <span class="at">ncol =</span> <span class="dv">10</span>)</span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb127-6"><a href="#cb127-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb127-7"><a href="#cb127-7" aria-hidden="true" tabindex="-1"></a>        k[i, j] <span class="ot">&lt;-</span> <span class="fu">median</span>(post<span class="sc">$</span>sdgp_a_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> </span>
<span id="cb127-8"><a href="#cb127-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">exp</span>(<span class="sc">-</span>islandsDistMatrix[i, j]<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">median</span>(post<span class="sc">$</span>lscale_a_gplat_adjlon2_adj)<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb127-9"><a href="#cb127-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-10"><a href="#cb127-10" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(k) <span class="ot">&lt;-</span> <span class="fu">median</span>(post<span class="sc">$</span>sdgp_a_gplat_adjlon2_adj<span class="sc">^</span><span class="dv">2</span>) <span class="sc">+</span> <span class="fl">0.01</span></span>
<span id="cb127-11"><a href="#cb127-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-12"><a href="#cb127-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to correlation matrix</span></span>
<span id="cb127-13"><a href="#cb127-13" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">cov2cor</span>(k), <span class="dv">2</span>)</span>
<span id="cb127-14"><a href="#cb127-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-15"><a href="#cb127-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add row/col names for convenience</span></span>
<span id="cb127-16"><a href="#cb127-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(rho) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ml"</span>, <span class="st">"Ti"</span>, <span class="st">"SC"</span>, <span class="st">"Ya"</span>, <span class="st">"Fi"</span>, <span class="st">"Tr"</span>, <span class="st">"Ch"</span>, <span class="st">"Mn"</span>, <span class="st">"To"</span>, <span class="st">"Ha"</span>)</span>
<span id="cb127-17"><a href="#cb127-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(rho) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(rho)</span>
<span id="cb127-18"><a href="#cb127-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-19"><a href="#cb127-19" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ml   Ti   SC   Ya   Fi   Tr   Ch   Mn   To Ha
Ml 1.00 0.86 0.83 0.01 0.61 0.31 0.06 0.12 0.37  0
Ti 0.86 1.00 0.89 0.01 0.61 0.32 0.10 0.14 0.33  0
SC 0.83 0.89 1.00 0.02 0.49 0.42 0.16 0.21 0.23  0
Ya 0.01 0.01 0.02 1.00 0.00 0.18 0.48 0.46 0.00  0
Fi 0.61 0.61 0.49 0.00 1.00 0.06 0.01 0.02 0.79  0
Tr 0.31 0.32 0.42 0.18 0.06 1.00 0.39 0.76 0.02  0
Ch 0.06 0.10 0.16 0.48 0.01 0.39 1.00 0.62 0.00  0
Mn 0.12 0.14 0.21 0.46 0.02 0.76 0.62 1.00 0.00  0
To 0.37 0.33 0.23 0.00 0.79 0.02 0.00 0.00 1.00  0
Ha 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00  1</code></pre>
</div>
</div>
<p>Like before, we’ll view them in a plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">row =</span> d<span class="sc">$</span>culture) <span class="sc">|&gt;</span> </span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>row, <span class="at">values_to =</span> <span class="st">"distance"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">column =</span> <span class="fu">factor</span>(name, <span class="at">levels =</span> <span class="fu">colnames</span>(d_mat)),</span>
<span id="cb129-6"><a href="#cb129-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">row    =</span> <span class="fu">factor</span>(row, <span class="at">levels =</span> <span class="fu">rownames</span>(d_mat)) <span class="sc">|&gt;</span> <span class="fu">fct_rev</span>(),</span>
<span id="cb129-7"><a href="#cb129-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">label  =</span> <span class="fu">formatC</span>(distance, <span class="at">format =</span> <span class="st">'f'</span>, <span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">"0."</span>, <span class="st">"."</span>)) <span class="sc">|&gt;</span></span>
<span id="cb129-8"><a href="#cb129-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Omit this line to keep the diagonal of 1's</span></span>
<span id="cb129-9"><a href="#cb129-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(distance <span class="sc">!=</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb129-10"><a href="#cb129-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb129-11"><a href="#cb129-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> column, <span class="at">y =</span> row)) <span class="sc">+</span> </span>
<span id="cb129-12"><a href="#cb129-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill =</span> distance)) <span class="sc">+</span> </span>
<span id="cb129-13"><a href="#cb129-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> label),</span>
<span id="cb129-14"><a href="#cb129-14" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">size =</span> <span class="fl">2.75</span>) <span class="sc">+</span></span>
<span id="cb129-15"><a href="#cb129-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">position =</span> <span class="st">"top"</span>) <span class="sc">+</span></span>
<span id="cb129-16"><a href="#cb129-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb129-17"><a href="#cb129-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="fu">expression</span>(rho), <span class="at">low =</span> <span class="st">"#FCF9F0"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb129-18"><a href="#cb129-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_earring</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb129-19"><a href="#cb129-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.ticks =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-82-1.png" class="img-fluid figure-img" width="456"></p>
</figure>
</div>
</div>
</div>
<p>On the whole, the correlation medians appear a just little bit lower than from the Poisson model. Now let’s make a gamma-Poisson alternative to Figure 14.12.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tidy up rho</span></span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>tidy_rho <span class="ot">&lt;-</span> rho <span class="sc">|&gt;</span></span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">|&gt;</span> </span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(d <span class="sc">|&gt;</span> <span class="fu">select</span>(culture, logpop, total_tools, lon2, lat)) <span class="sc">|&gt;</span> </span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(Ml<span class="sc">:</span>Ha,</span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"colname"</span>, </span>
<span id="cb130-8"><a href="#cb130-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"correlation"</span>) <span class="sc">|&gt;</span></span>
<span id="cb130-9"><a href="#cb130-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">str_c</span>(<span class="fu">pmin</span>(rowname, colname), <span class="fu">pmax</span>(rowname, colname))) </span>
<span id="cb130-10"><a href="#cb130-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-11"><a href="#cb130-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Left panel</span></span>
<span id="cb130-12"><a href="#cb130-12" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> tidy_rho <span class="sc">|&gt;</span>       </span>
<span id="cb130-13"><a href="#cb130-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> lon2, <span class="at">y =</span> lat)) <span class="sc">+</span></span>
<span id="cb130-14"><a href="#cb130-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> d, </span>
<span id="cb130-15"><a href="#cb130-15" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">size =</span> logpop), <span class="at">color =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb130-16"><a href="#cb130-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> group, <span class="at">alpha =</span> correlation<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb130-17"><a href="#cb130-17" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb130-18"><a href="#cb130-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> d, <span class="fu">aes</span>(<span class="at">label =</span> culture), </span>
<span id="cb130-19"><a href="#cb130-19" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">point.padding =</span> <span class="fl">0.2</span>, <span class="at">seed =</span> <span class="dv">14</span>, <span class="at">size =</span> <span class="fl">2.75</span>) <span class="sc">+</span></span>
<span id="cb130-20"><a href="#cb130-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_alpha_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb130-21"><a href="#cb130-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"longitude"</span>,</span>
<span id="cb130-22"><a href="#cb130-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"latitude"</span>,</span>
<span id="cb130-23"><a href="#cb130-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Among societies in geographic space</span><span class="sc">\n</span><span class="st">"</span>) <span class="sc">+</span></span>
<span id="cb130-24"><a href="#cb130-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">range</span>(d<span class="sc">$</span>lon2),</span>
<span id="cb130-25"><a href="#cb130-25" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">range</span>(d<span class="sc">$</span>lat)) <span class="sc">+</span></span>
<span id="cb130-26"><a href="#cb130-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb130-27"><a href="#cb130-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-28"><a href="#cb130-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the average posterior predictive relationship between </span></span>
<span id="cb130-29"><a href="#cb130-29" aria-hidden="true" tabindex="-1"></a><span class="co"># log population and total tools, summarized by the median and 80% interval</span></span>
<span id="cb130-30"><a href="#cb130-30" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> post <span class="sc">|&gt;</span> </span>
<span id="cb130-31"><a href="#cb130-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(<span class="at">logpop =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">6</span>, <span class="at">to =</span> <span class="dv">14</span>, <span class="at">length.out =</span> <span class="dv">30</span>)) <span class="sc">|&gt;</span></span>
<span id="cb130-32"><a href="#cb130-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">population =</span> <span class="fu">exp</span>(logpop)) <span class="sc">|&gt;</span> </span>
<span id="cb130-33"><a href="#cb130-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">lambda =</span> <span class="fu">exp</span>(b_a_Intercept) <span class="sc">*</span> population<span class="sc">^</span>b_b_Intercept <span class="sc">/</span> b_g_Intercept) <span class="sc">|&gt;</span></span>
<span id="cb130-34"><a href="#cb130-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(logpop) <span class="sc">|&gt;</span> </span>
<span id="cb130-35"><a href="#cb130-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">median_qi</span>(lambda, <span class="at">.width =</span> <span class="fl">0.8</span>)</span>
<span id="cb130-36"><a href="#cb130-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-37"><a href="#cb130-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Right panel</span></span>
<span id="cb130-38"><a href="#cb130-38" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> tidy_rho <span class="sc">|&gt;</span> </span>
<span id="cb130-39"><a href="#cb130-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> logpop)) <span class="sc">+</span></span>
<span id="cb130-40"><a href="#cb130-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">data =</span> f,</span>
<span id="cb130-41"><a href="#cb130-41" aria-hidden="true" tabindex="-1"></a>              <span class="fu">aes</span>(<span class="at">y =</span> lambda, <span class="at">ymin =</span> .lower, <span class="at">ymax =</span> .upper),</span>
<span id="cb130-42"><a href="#cb130-42" aria-hidden="true" tabindex="-1"></a>              <span class="at">stat =</span> <span class="st">"identity"</span>,</span>
<span id="cb130-43"><a href="#cb130-43" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">fill =</span> <span class="st">"#394165"</span>, <span class="at">linewidth =</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb130-44"><a href="#cb130-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> d, </span>
<span id="cb130-45"><a href="#cb130-45" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">y =</span> total_tools, <span class="at">size =</span> logpop), </span>
<span id="cb130-46"><a href="#cb130-46" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#DCA258"</span>) <span class="sc">+</span></span>
<span id="cb130-47"><a href="#cb130-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> total_tools, <span class="at">group =</span> group, <span class="at">alpha =</span> correlation<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb130-48"><a href="#cb130-48" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb130-49"><a href="#cb130-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> d, </span>
<span id="cb130-50"><a href="#cb130-50" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">y =</span> total_tools, <span class="at">label =</span> culture), </span>
<span id="cb130-51"><a href="#cb130-51" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">point.padding =</span> <span class="fl">0.2</span>, <span class="at">seed =</span> <span class="dv">14</span>, <span class="at">size =</span> <span class="fl">2.75</span>) <span class="sc">+</span></span>
<span id="cb130-52"><a href="#cb130-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_alpha_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb130-53"><a href="#cb130-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"log population"</span>,</span>
<span id="cb130-54"><a href="#cb130-54" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"total tools"</span>,</span>
<span id="cb130-55"><a href="#cb130-55" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Shown against the relation between</span><span class="sc">\n</span><span class="st">total tools and log pop"</span>) <span class="sc">+</span></span>
<span id="cb130-56"><a href="#cb130-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">range</span>(d<span class="sc">$</span>logpop),</span>
<span id="cb130-57"><a href="#cb130-57" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ylim =</span> <span class="fu">range</span>(d<span class="sc">$</span>total_tools)) <span class="sc">+</span></span>
<span id="cb130-58"><a href="#cb130-58" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb130-59"><a href="#cb130-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-60"><a href="#cb130-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine, entitle, and display</span></span>
<span id="cb130-61"><a href="#cb130-61" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> </span>
<span id="cb130-62"><a href="#cb130-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Posterior median correlations based on the gamma-Poisson model"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-83-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>There is very little difference. Finish off by comparing the two models with the WAIC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.8</span>    <span class="ot">&lt;-</span> <span class="fu">add_criterion</span>(b14<span class="fl">.8</span>,    <span class="at">criterion =</span> <span class="st">"waic"</span>)</span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.8</span>_nb <span class="ot">&lt;-</span> <span class="fu">add_criterion</span>(b14<span class="fl">.8</span>_nb, <span class="at">criterion =</span> <span class="st">"waic"</span>)</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a><span class="fu">loo_compare</span>(b14<span class="fl">.8</span>, b14<span class="fl">.8</span>_nb, <span class="at">criterion =</span> <span class="st">"waic"</span>) <span class="sc">|&gt;</span> <span class="fu">print</span>(<span class="at">simplify =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         elpd_diff se_diff elpd_waic se_elpd_waic p_waic se_p_waic waic  se_waic
b14.8      0.0       0.0   -33.7       1.4          3.9    0.7      67.3   2.8  
b14.8_nb  -2.4       0.4   -36.1       1.7          3.8    0.7      72.2   3.4  </code></pre>
</div>
</div>
<p>The WAIC comparison suggests we gain little by switching to the gamma-Poisson. If anything, we may have overfit.</p>
</section>
</section>
<section id="example-phylogenetic-distance" class="level3" data-number="14.5.2">
<h3 data-number="14.5.2" class="anchored" data-anchor-id="example-phylogenetic-distance"><span class="header-section-number">14.5.2</span> Example: Phylogenetic distance</h3>
<blockquote class="blockquote">
<p>Consider as an example the causal influence of group size (<span class="math inline">\(G\)</span>) on brain size (<span class="math inline">\(B\)</span>). Hypotheses connecting these variables are popular, because primates (including humans) are unusual in both. Most primates live in social groups. Most mammals do not. Second, primates have relatively large brains. There is a family of hypotheses linking these two features. Suppose for example that group living, whatever its cause, could select for larger brains, because once you live with others, a larger brain helps to cope with the complexity of cooperation and manipulation. (p.&nbsp;477)</p>
</blockquote>
<p>There are also many potential confounds, which we’ll call <span class="math inline">\(U\)</span>. Also imagine there are two time points, indicated by subscripts 1 and 2. Here is the basic DAG.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>dag_coords <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"G1"</span>, <span class="st">"B1"</span>, <span class="st">"U1"</span>, <span class="st">"G2"</span>, <span class="st">"B2"</span>, <span class="st">"U2"</span>),</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x    =</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">each =</span> <span class="dv">3</span>),</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y    =</span> <span class="fu">rep</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">1</span>, <span class="at">times =</span> <span class="dv">2</span>))</span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dagify</span>(G2 <span class="sc">~</span> G1 <span class="sc">+</span> U1,</span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>       B2 <span class="sc">~</span> G1 <span class="sc">+</span> B1 <span class="sc">+</span> U1,</span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a>       U2 <span class="sc">~</span> U1,</span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">coords =</span> dag_coords) <span class="sc">|&gt;</span></span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tidy_dagitty</span>() <span class="sc">|&gt;</span> </span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">color =</span> <span class="fu">ifelse</span>(name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"G1"</span>, <span class="st">"B1"</span>), <span class="st">"a"</span>,</span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">ifelse</span>(name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"G2"</span>, <span class="st">"B2"</span>), <span class="st">"b"</span>, <span class="st">"c"</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb133-13"><a href="#cb133-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb133-14"><a href="#cb133-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">xend =</span> xend, <span class="at">yend =</span> yend)) <span class="sc">+</span></span>
<span id="cb133-15"><a href="#cb133-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> color),</span>
<span id="cb133-16"><a href="#cb133-16" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">"#FCF9F0"</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">7</span>, <span class="at">stroke =</span> <span class="dv">2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb133-17"><a href="#cb133-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_text</span>(<span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">parse =</span> T,</span>
<span id="cb133-18"><a href="#cb133-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">label =</span> <span class="fu">c</span>(<span class="fu">expression</span>(B[<span class="dv">1</span>]), <span class="fu">expression</span>(G[<span class="dv">1</span>]), <span class="fu">expression</span>(U[<span class="dv">1</span>]), </span>
<span id="cb133-19"><a href="#cb133-19" aria-hidden="true" tabindex="-1"></a>                          <span class="fu">expression</span>(B[<span class="dv">2</span>]), <span class="fu">expression</span>(G[<span class="dv">2</span>]), <span class="fu">expression</span>(U[<span class="dv">2</span>]))) <span class="sc">+</span></span>
<span id="cb133-20"><a href="#cb133-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_edges</span>(<span class="at">edge_colour =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb133-21"><a href="#cb133-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#EEDA9D"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb133-22"><a href="#cb133-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_dag</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-85-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<p>However, this will not be our model. Rather, we’ll consider this one.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>dag_coords <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"G"</span>, <span class="st">"U"</span>, <span class="st">"M"</span>, <span class="st">"P"</span>, <span class="st">"B"</span>),</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x    =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>),</span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y    =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dagify</span>(G <span class="sc">~</span> U <span class="sc">+</span> M,</span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a>       U <span class="sc">~</span> P,</span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a>       M <span class="sc">~</span> U,</span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a>       B <span class="sc">~</span> G <span class="sc">+</span> M <span class="sc">+</span> U,</span>
<span id="cb134-10"><a href="#cb134-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">coords =</span> dag_coords) <span class="sc">|&gt;</span></span>
<span id="cb134-11"><a href="#cb134-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb134-12"><a href="#cb134-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">xend =</span> xend, <span class="at">yend =</span> yend)) <span class="sc">+</span></span>
<span id="cb134-13"><a href="#cb134-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> name <span class="sc">==</span> <span class="st">"U"</span>),</span>
<span id="cb134-14"><a href="#cb134-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">"#FCF9F0"</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">6</span>, <span class="at">stroke =</span> <span class="dv">2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb134-15"><a href="#cb134-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_text</span>(<span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">parse =</span> T) <span class="sc">+</span></span>
<span id="cb134-16"><a href="#cb134-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_dag_edges</span>(<span class="at">edge_colour =</span> <span class="st">"#FCF9F0"</span>) <span class="sc">+</span></span>
<span id="cb134-17"><a href="#cb134-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#EEDA9D"</span>, <span class="st">"#A65141"</span>)) <span class="sc">+</span></span>
<span id="cb134-18"><a href="#cb134-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_dag</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-86-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>There’s a lot going on here, but we can take it one piece at a time. Again, we’re interested in <span class="math inline">\(G \rightarrow B\)</span>. There is one confound we know for sure, body mass (<span class="math inline">\(M\)</span>). It possibly influences both <span class="math inline">\(G\)</span> and <span class="math inline">\(B\)</span>. So we’ll include that in the model. The unobserved confounds <span class="math inline">\(U\)</span> could potentially influence all three variables. Finally, we let the phylogenetic relationships (<span class="math inline">\(P\)</span>) influence <span class="math inline">\(U\)</span>. How is <span class="math inline">\(P\)</span> causal? If we traveled back in time and delayed a split between two species, it could influence the expected differences in their traits. So it is really the timing of the split that is causal, not the phylogeny. Of course <span class="math inline">\(P\)</span> may also influence <span class="math inline">\(G\)</span> and <span class="math inline">\(B\)</span> and <span class="math inline">\(M\)</span> directly. But those arrows aren’t our concern right now, so I’ve omitted them for clarity. (p.&nbsp;478)</p>
</blockquote>
<p><strong>Phylogenetic regression</strong> models, which “use some function of phylogenetic distance to model the covariation among species” (p.&nbsp;478) attempt to grapple with all this. To see how, load the primates data and its phylogeny <span class="citation" data-cites="streetCoevolutionCulturalIntelligence2017">(see <a href="references.html#ref-streetCoevolutionCulturalIntelligence2017" role="doc-biblioref">Street et al., 2017</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Primates301, <span class="at">package =</span> <span class="st">"rethinking"</span>) </span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Primates301_nex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When working within the <strong>ggplot2</strong> framework, one can plot a phylogeny with help from the <a href="https://github.com/YuLab-SMU/ggtree"><strong>ggtree</strong> package</a> <span class="citation" data-cites="yuDataIntegrationManipulation2020 yuUsingGgtreeVisualize2020 yuTwoMethodsMapping2018 yuGgtreePackageVisualization2017">(<a href="references.html#ref-yuGgtreePackageVisualization2017" role="doc-biblioref">Yu et al., 2017</a>, <a href="references.html#ref-yuTwoMethodsMapping2018" role="doc-biblioref">2018</a>; <a href="references.html#ref-yuUsingGgtreeVisualize2020" role="doc-biblioref">Yu, 2020a</a>, <a href="references.html#ref-yuDataIntegrationManipulation2020" role="doc-biblioref">2020b</a>)</span>. Here’s a basic plot for our version of Figure 14.13.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>Primates301_nex <span class="sc">|&gt;</span></span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtree</span>(<span class="at">layout =</span> <span class="st">"circular"</span>, <span class="at">color =</span> <span class="st">"#394165"</span>, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span> </span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="at">color =</span> <span class="st">"#100F14"</span>, <span class="at">size =</span> <span class="dv">5</span><span class="sc">/</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-88-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s format the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> Primates301 <span class="sc">|&gt;</span> </span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">name =</span> <span class="fu">as.character</span>(name)) <span class="sc">|&gt;</span> </span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>(group_size, body, brain) <span class="sc">|&gt;</span> </span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">m =</span> <span class="fu">log</span>(body) <span class="sc">|&gt;</span> <span class="fu">standardize</span>(),</span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">b =</span> <span class="fu">log</span>(brain) <span class="sc">|&gt;</span> <span class="fu">standardize</span>(),</span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">g =</span> <span class="fu">log</span>(group_size) <span class="sc">|&gt;</span> <span class="fu">standardize</span>())</span>
<span id="cb137-7"><a href="#cb137-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-8"><a href="#cb137-8" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 151
Columns: 19
$ name                &lt;chr&gt; "Allenopithecus_nigroviridis", "Alouatta_belzebul", "Alouatta_caraya", "Alouatta_guariba", "Alouatta…
$ genus               &lt;fct&gt; Allenopithecus, Alouatta, Alouatta, Alouatta, Alouatta, Alouatta, Alouatta, Aotus, Aotus, Arctocebus…
$ species             &lt;fct&gt; nigroviridis, belzebul, caraya, guariba, palliata, pigra, seniculus, azarai, trivirgatus, calabarens…
$ subspecies          &lt;fct&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ spp_id              &lt;int&gt; 1, 3, 4, 5, 6, 7, 9, 10, 18, 22, 23, 25, 26, 28, 29, 32, 33, 34, 40, 41, 46, 49, 50, 51, 52, 53, 54,…
$ genus_id            &lt;int&gt; 1, 3, 3, 3, 3, 3, 3, 4, 4, 6, 7, 7, 7, 8, 8, 10, 11, 11, 13, 14, 14, 14, 14, 15, 15, 15, 15, 16, 16,…
$ social_learning     &lt;int&gt; 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 1, 17, 5, 0, 0, 1, NA, NA, 1, 0…
$ research_effort     &lt;int&gt; 6, 15, 45, 37, 79, 25, 82, 22, 58, 1, 12, 58, 30, 10, 6, 24, 11, 8, 43, 16, 161, NA, 36, 13, 249, 60…
$ brain               &lt;dbl&gt; 58.02, 52.84, 52.63, 51.70, 49.88, 51.13, 55.22, 20.67, 16.85, 6.92, 117.02, 105.09, 103.85, 9.86, 7…
$ body                &lt;dbl&gt; 4655, 6395, 5383, 5175, 6250, 8915, 5950, 1205, 989, 309, 8167, 7535, 8280, 1207, 801, 6728, 3165, 2…
$ group_size          &lt;dbl&gt; 40.00, 7.40, 8.90, 7.40, 13.10, 5.50, 7.90, 4.10, 3.15, 1.00, 14.50, 42.00, 20.00, 2.00, 3.00, 3.20,…
$ gestation           &lt;dbl&gt; NA, NA, 185.92, NA, 185.42, 185.92, 189.90, NA, 133.47, 133.74, 138.20, 226.37, 228.18, 136.15, NA, …
$ weaning             &lt;dbl&gt; 106.15, NA, 323.16, NA, 495.60, NA, 370.04, 229.69, 76.21, 109.26, NA, 816.35, 805.41, 149.15, NA, 6…
$ longevity           &lt;dbl&gt; 276.0, NA, 243.6, NA, 300.0, 240.0, 300.0, NA, 303.6, 156.0, 336.0, 327.6, 453.6, NA, NA, NA, 324.0,…
$ sex_maturity        &lt;dbl&gt; NA, NA, 1276.72, NA, 1578.42, NA, 1690.22, NA, 736.60, 298.91, NA, 2104.57, 2104.57, NA, NA, 2689.08…
$ maternal_investment &lt;dbl&gt; NA, NA, 509.08, NA, 681.02, NA, 559.94, NA, 209.68, 243.00, NA, 1042.72, 1033.59, 285.30, NA, 867.63…
$ m                   &lt;dbl&gt; 0.36958768, 0.57601585, 0.46403740, 0.43842259, 0.56110778, 0.79196303, 0.52913339, -0.50888297, -0.…
$ b                   &lt;dbl&gt; 0.4039485, 0.3285765, 0.3253670, 0.3109981, 0.2821147, 0.3020630, 0.3640841, -0.4278779, -0.5925603,…
$ g                   &lt;dbl&gt; 1.397272713, 0.003132082, 0.155626096, 0.003132082, 0.475005322, -0.242029791, 0.057151751, -0.48473…</code></pre>
</div>
</div>
<p>Our first model, which we might call the naïve model, explores the conditional relations of <span class="math inline">\(\log(\text{body mass})\)</span> and <span class="math inline">\(\log(\text{group size})\)</span> on <span class="math inline">\(\log(\text{brain size})\)</span> without accounting for phylogenetic relationships.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.9</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d,</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  b <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> m <span class="sc">+</span> g,</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b),</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma)),</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.09"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the summary of the naïve model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity 
Formula: b ~ 1 + m + g 
   Data: d (Number of observations: 151) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Regression Coefficients:
          Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept     0.00      0.02    -0.03     0.04 1.00     3964     2812
m             0.89      0.02     0.85     0.94 1.00     3163     2807
g             0.12      0.02     0.08     0.17 1.00     3314     2647

Further Distributional Parameters:
      Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma     0.22      0.01     0.19     0.24 1.00     4013     2656

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>If you want <span class="math inline">\(\sigma\)</span> in the <span class="math inline">\(\sigma^2\)</span> metric, you can square that by hand.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_draws_df</span>(b14<span class="fl">.9</span>) <span class="sc">|&gt;</span> </span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sigma_sq =</span> sigma<span class="sc">^</span><span class="dv">2</span>) <span class="sc">|&gt;</span> </span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean_qi</span>(sigma_sq) <span class="sc">|&gt;</span> </span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_if</span>(is.double, round, <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  sigma_sq .lower .upper .width .point .interval
     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    
1     0.05   0.04   0.06   0.95 mean   qi       </code></pre>
</div>
</div>
<p>The oldest and most conservative way to include information about phylogenetic relationships is with a Brownian motion model.</p>
<blockquote class="blockquote">
<p>Brownian motion just means Gaussian random walks. If species traits drift randomly with respect to one another after speciation, then the covariance between a pair of species ends up being linearly related to the phylogenetic branch distance between them–the further apart, the less covariance, as a proportion of distance. (p.&nbsp;481)</p>
</blockquote>
<p>We’ll use functions from the <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong> package</a> <span class="citation" data-cites="R-ape ape2019">(<a href="references.html#ref-R-ape" role="doc-biblioref">Emmanuel Paradis et al., 2022</a>; <a href="references.html#ref-ape2019" role="doc-biblioref">E. Paradis &amp; Schliep, 2019</a>)</span> to make the covariance matrix (<code>V</code>) and the distance matrix (<code>Dmat</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>spp_obs <span class="ot">&lt;-</span> d<span class="sc">$</span>name</span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a>tree_trimmed <span class="ot">&lt;-</span> <span class="fu">keep.tip</span>(Primates301_nex, spp_obs)</span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a>Rbm <span class="ot">&lt;-</span> <span class="fu">corBrownian</span>(<span class="at">phy =</span> tree_trimmed)</span>
<span id="cb144-7"><a href="#cb144-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-8"><a href="#cb144-8" aria-hidden="true" tabindex="-1"></a>V <span class="ot">&lt;-</span> <span class="fu">vcv</span>(Rbm)</span>
<span id="cb144-9"><a href="#cb144-9" aria-hidden="true" tabindex="-1"></a>Dmat <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>( tree_trimmed )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here’s the distance by covariance scatter plot McElreath alluded to but did not show in the text.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">full_join</span>(</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>  Dmat <span class="sc">|&gt;</span> </span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"row"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="sc">-</span>row,</span>
<span id="cb145-5"><a href="#cb145-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">names_to =</span> <span class="st">"col"</span>,</span>
<span id="cb145-6"><a href="#cb145-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">values_to =</span> <span class="st">"distance"</span>),</span>
<span id="cb145-7"><a href="#cb145-7" aria-hidden="true" tabindex="-1"></a>  V <span class="sc">|&gt;</span> </span>
<span id="cb145-8"><a href="#cb145-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"row"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb145-9"><a href="#cb145-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="sc">-</span>row,</span>
<span id="cb145-10"><a href="#cb145-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">names_to =</span> <span class="st">"col"</span>,</span>
<span id="cb145-11"><a href="#cb145-11" aria-hidden="true" tabindex="-1"></a>                 <span class="at">values_to =</span> <span class="st">"covariance"</span>),</span>
<span id="cb145-12"><a href="#cb145-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"row"</span>, <span class="st">"col"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb145-13"><a href="#cb145-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb145-14"><a href="#cb145-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> distance, <span class="at">y =</span> covariance)) <span class="sc">+</span></span>
<span id="cb145-15"><a href="#cb145-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>, <span class="at">color =</span> <span class="st">"#80A0C7"</span>) <span class="sc">+</span></span>
<span id="cb145-16"><a href="#cb145-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"phylogenetic distance"</span>, </span>
<span id="cb145-17"><a href="#cb145-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"covariance"</span>,</span>
<span id="cb145-18"><a href="#cb145-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"These variables are the</span><span class="sc">\n</span><span class="st">inverse of one another."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-93-1.png" class="img-fluid figure-img" width="312"></p>
</figure>
</div>
</div>
</div>
<p>McElreath suggested executing <code>image(V)</code> and <code>image(Dmat)</code> to plot heat maps of each matrix. We’ll have to work a little harder to make decent-looking head maps within our <strong>tidyverse</strong> workflow.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Headmap of `Dmat`</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> Dmat <span class="sc">|&gt;</span> </span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"row"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>row,</span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"col"</span>,</span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"distance"</span>) <span class="sc">|&gt;</span></span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb146-8"><a href="#cb146-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> col, <span class="at">y =</span> row, <span class="at">fill =</span> distance)) <span class="sc">+</span></span>
<span id="cb146-9"><a href="#cb146-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb146-10"><a href="#cb146-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb146-11"><a href="#cb146-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb146-12"><a href="#cb146-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"#100F14"</span>, <span class="at">high =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb146-13"><a href="#cb146-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"top"</span>)</span>
<span id="cb146-14"><a href="#cb146-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-15"><a href="#cb146-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Headmap of `V`</span></span>
<span id="cb146-16"><a href="#cb146-16" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> V <span class="sc">|&gt;</span> </span>
<span id="cb146-17"><a href="#cb146-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"row"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb146-18"><a href="#cb146-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>row,</span>
<span id="cb146-19"><a href="#cb146-19" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"col"</span>,</span>
<span id="cb146-20"><a href="#cb146-20" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"covariance"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb146-21"><a href="#cb146-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb146-22"><a href="#cb146-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> col, <span class="at">y =</span> row, <span class="at">fill =</span> covariance)) <span class="sc">+</span></span>
<span id="cb146-23"><a href="#cb146-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb146-24"><a href="#cb146-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb146-25"><a href="#cb146-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb146-26"><a href="#cb146-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"#100F14"</span>, <span class="at">high =</span> <span class="st">"#EEDA9D"</span>) <span class="sc">+</span></span>
<span id="cb146-27"><a href="#cb146-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"top"</span>)</span>
<span id="cb146-28"><a href="#cb146-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-29"><a href="#cb146-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine, entitle, and display</span></span>
<span id="cb146-30"><a href="#cb146-30" aria-hidden="true" tabindex="-1"></a>(p1 <span class="sc">|</span> p2) <span class="sc">+</span> <span class="fu">plot_annotation</span>(<span class="at">subtitle =</span> <span class="st">"Again, distance is the inverse of covariance."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-94-1.png" class="img-fluid figure-img" width="600"></p>
</figure>
</div>
</div>
</div>
<p>Within the <strong>brms</strong> paradigm, one inserts a known covariance matrix into a model using the <code>fcor()</code> function. For any longer-term <strong>brms</strong> users, <code>fcor()</code> is the replacement for the now-depreciated <code>cor_fixed()</code> function. Along with <code>fcor()</code>, one tells <strong>brms</strong> about the data with the <code>data2</code> function. Here’s how to use it for our version of McElreath’s <code>m14.10</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>R <span class="ot">&lt;-</span> V[spp_obs, spp_obs] <span class="sc">/</span> <span class="fu">max</span>(V)</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.10</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> d,</span>
<span id="cb147-5"><a href="#cb147-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data2 =</span> <span class="fu">list</span>(<span class="at">R =</span> R),</span>
<span id="cb147-6"><a href="#cb147-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb147-7"><a href="#cb147-7" aria-hidden="true" tabindex="-1"></a>  b <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> m <span class="sc">+</span> g <span class="sc">+</span> <span class="fu">fcor</span>(R),</span>
<span id="cb147-8"><a href="#cb147-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb147-9"><a href="#cb147-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">class =</span> b),</span>
<span id="cb147-10"><a href="#cb147-10" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma)),</span>
<span id="cb147-11"><a href="#cb147-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">2000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb147-12"><a href="#cb147-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb147-13"><a href="#cb147-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.10"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the summary of the Brownian model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity 
Formula: b ~ 1 + m + g + fcor(R) 
   Data: d (Number of observations: 151) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Regression Coefficients:
          Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept    -0.19      0.16    -0.52     0.12 1.00     4252     2997
m             0.70      0.04     0.63     0.77 1.00     4253     2619
g            -0.01      0.02    -0.05     0.03 1.00     3981     3083

Further Distributional Parameters:
      Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma     0.40      0.02     0.36     0.45 1.00     4190     2918

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Since our residual variance is still in the <span class="math inline">\(\sigma\)</span> metric, it will be easier to compare it to McElreath’s <code>sigma_sq</code> parameter after transforming the posterior samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_draws_df</span>(b14<span class="fl">.10</span>) <span class="sc">|&gt;</span> </span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sigma_sq =</span> sigma<span class="sc">^</span><span class="dv">2</span>) <span class="sc">|&gt;</span> </span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean_hdi</span>(sigma_sq, <span class="at">.width =</span> <span class="fl">0.89</span>) <span class="sc">|&gt;</span> </span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_if</span>(is.double, round, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  sigma_sq .lower .upper .width .point .interval
     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    
1     0.16   0.13   0.19   0.89 mean   hdi      </code></pre>
</div>
</div>
<p>McElreath introduced the Ornstein–Uhlenbeck (OU) process as an alternative to the Brownian motion model. The OU model proposes the covariance between two species <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> is</p>
<p><span class="math display">\[K(i, j) = \eta^2 \exp(\rho^2 D_{ij}),\]</span></p>
<p>where, in our case, <span class="math inline">\(D_{ij}\)</span> is the distance matrix we’ve saved above as <code>Dmat</code>. Sadly for us, <strong>brms</strong> only supports the exponentiated-quadratic kernel for Gaussian process models, at this time. However, the Ornstein–Uhlenbeck kernel is one of the alternative kernels Bürkner has on his to-do list (see <a href="https://github.com/paul-buerkner/brms/issues/234">GitHub issue #234</a>). To follow along with McElreath, we will have to use <strong>rethinking</strong> or raw Stan. Our approach will be the former. First we make the <code>dat_list</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>dat_list <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">N_spp =</span> <span class="fu">nrow</span>(d),</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">M     =</span> <span class="fu">standardize</span>(<span class="fu">log</span>(d<span class="sc">$</span>body)),</span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">B     =</span> <span class="fu">standardize</span>(<span class="fu">log</span>(d<span class="sc">$</span>brain)),</span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">G     =</span> <span class="fu">standardize</span>(<span class="fu">log</span>(d<span class="sc">$</span>group_size)), <span class="at">Imat =</span> <span class="fu">diag</span>(<span class="fu">nrow</span>(d)),</span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">V     =</span> V[spp_obs, spp_obs],</span>
<span id="cb152-7"><a href="#cb152-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">R     =</span> V[spp_obs, spp_obs] <span class="sc">/</span> <span class="fu">max</span>(V[spp_obs, spp_obs]),</span>
<span id="cb152-8"><a href="#cb152-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">Dmat  =</span> Dmat[spp_obs, spp_obs] <span class="sc">/</span> <span class="fu">max</span>(Dmat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now fit the OU model with <code>rethinking::ulam()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>m14<span class="fl">.11</span> <span class="ot">&lt;-</span> <span class="fu">ulam</span>( </span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">alist</span>(</span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>    B <span class="sc">~</span> <span class="fu">multi_normal</span>(mu, SIGMA),</span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a>    mu <span class="ot">&lt;-</span> a <span class="sc">+</span> bM <span class="sc">*</span> M <span class="sc">+</span> bG <span class="sc">*</span> G,</span>
<span id="cb153-5"><a href="#cb153-5" aria-hidden="true" tabindex="-1"></a>    matrix[N_spp,N_spp]<span class="sc">:</span> SIGMA <span class="ot">&lt;-</span> <span class="fu">cov_GPL1</span>(Dmat, etasq, rhosq, <span class="fl">0.01</span>), </span>
<span id="cb153-6"><a href="#cb153-6" aria-hidden="true" tabindex="-1"></a>    a <span class="sc">~</span> <span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb153-7"><a href="#cb153-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(bM,bG) <span class="sc">~</span> <span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.5</span>),</span>
<span id="cb153-8"><a href="#cb153-8" aria-hidden="true" tabindex="-1"></a>    etasq <span class="sc">~</span> <span class="fu">half_normal</span>(<span class="dv">1</span>, <span class="fl">0.25</span>),</span>
<span id="cb153-9"><a href="#cb153-9" aria-hidden="true" tabindex="-1"></a>    rhosq <span class="sc">~</span> <span class="fu">half_normal</span>(<span class="dv">3</span>, <span class="fl">0.25</span>)</span>
<span id="cb153-10"><a href="#cb153-10" aria-hidden="true" tabindex="-1"></a>  ), </span>
<span id="cb153-11"><a href="#cb153-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat_list, </span>
<span id="cb153-12"><a href="#cb153-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Happily, our results are much like those in the text.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="fu">precis</span>(m14<span class="fl">.11</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>       mean   sd  5.5% 94.5% rhat ess_bulk
a     -0.06 0.08 -0.18  0.06    1  2085.50
bG     0.05 0.02  0.01  0.09    1  2096.53
bM     0.83 0.03  0.79  0.88    1  2301.19
etasq  0.03 0.01  0.03  0.05    1  1792.25
rhosq  2.80 0.24  2.41  3.18    1  1759.88</code></pre>
</div>
</div>
<p>Next we extract the posterior draws with <code>extract.samples()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>post <span class="ot">&lt;-</span> <span class="fu">extract.samples</span>(m14<span class="fl">.11</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To get a sense of the covariance function implied by <code>etasq</code> and <code>rhosq</code>, here we’ll plot the posterior against the prior for our <strong>tidyverse</strong> variant of Figure 14.14.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a>d_fig <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Posterior</span></span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>  post <span class="sc">|&gt;</span> </span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb157-7"><a href="#cb157-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">30</span>) <span class="sc">|&gt;</span> </span>
<span id="cb157-8"><a href="#cb157-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">draw =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">n</span>()) <span class="sc">|&gt;</span> </span>
<span id="cb157-9"><a href="#cb157-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(draw, etasq, rhosq) <span class="sc">|&gt;</span> </span>
<span id="cb157-10"><a href="#cb157-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">expand_grid</span>(<span class="at">d_seq =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">50</span>)),</span>
<span id="cb157-11"><a href="#cb157-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Prior</span></span>
<span id="cb157-12"><a href="#cb157-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">eta =</span> <span class="fu">abs</span>(<span class="fu">rnorm</span>(<span class="fl">1e5</span>, <span class="at">mean =</span> <span class="dv">1</span>, <span class="at">sd =</span> <span class="fl">0.25</span>)),</span>
<span id="cb157-13"><a href="#cb157-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">rho =</span> <span class="fu">abs</span>(<span class="fu">rnorm</span>(<span class="fl">1e5</span>, <span class="at">mean =</span> <span class="dv">3</span>, <span class="at">sd =</span> <span class="fl">0.25</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb157-14"><a href="#cb157-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">expand_grid</span>(<span class="at">d_seq =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">50</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb157-15"><a href="#cb157-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">k =</span> eta <span class="sc">*</span> <span class="fu">exp</span>(<span class="sc">-</span>rho <span class="sc">*</span> d_seq)) <span class="sc">|&gt;</span> </span>
<span id="cb157-16"><a href="#cb157-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(d_seq) <span class="sc">|&gt;</span> </span>
<span id="cb157-17"><a href="#cb157-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean_hdi</span>(k, <span class="at">.width =</span> <span class="fl">0.89</span>),</span>
<span id="cb157-18"><a href="#cb157-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Join them</span></span>
<span id="cb157-19"><a href="#cb157-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="st">"d_seq"</span>)</span>
<span id="cb157-20"><a href="#cb157-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-21"><a href="#cb157-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot!</span></span>
<span id="cb157-22"><a href="#cb157-22" aria-hidden="true" tabindex="-1"></a>d_fig <span class="sc">|&gt;</span> </span>
<span id="cb157-23"><a href="#cb157-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> d_seq)) <span class="sc">+</span></span>
<span id="cb157-24"><a href="#cb157-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> etasq <span class="sc">*</span> <span class="fu">exp</span>(<span class="sc">-</span>rhosq <span class="sc">*</span> d_seq), <span class="at">group =</span> draw),</span>
<span id="cb157-25"><a href="#cb157-25" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb157-26"><a href="#cb157-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_lineribbon</span>(<span class="at">data =</span> d_fig <span class="sc">|&gt;</span> <span class="fu">filter</span>(draw <span class="sc">==</span> <span class="dv">1</span>),</span>
<span id="cb157-27"><a href="#cb157-27" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">y =</span> k, <span class="at">ymin =</span> .lower, <span class="at">ymax =</span> .upper),</span>
<span id="cb157-28"><a href="#cb157-28" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"#A65141"</span>, <span class="at">fill =</span> <span class="st">"#E7CDC2"</span>) <span class="sc">+</span></span>
<span id="cb157-29"><a href="#cb157-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">"text"</span>,</span>
<span id="cb157-30"><a href="#cb157-30" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.5</span>), <span class="at">y =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.5</span>),</span>
<span id="cb157-31"><a href="#cb157-31" aria-hidden="true" tabindex="-1"></a>           <span class="at">label =</span> <span class="fu">c</span>(<span class="st">"posterior"</span>, <span class="st">"prior"</span>),</span>
<span id="cb157-32"><a href="#cb157-32" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="fu">c</span>(<span class="st">"#80A0C7"</span>, <span class="st">"#E7CDC2"</span>),</span>
<span id="cb157-33"><a href="#cb157-33" aria-hidden="true" tabindex="-1"></a>           <span class="at">family =</span> <span class="st">"Courier"</span>) <span class="sc">+</span></span>
<span id="cb157-34"><a href="#cb157-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"phylogenetic distance"</span>, </span>
<span id="cb157-35"><a href="#cb157-35" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"covariance"</span>) <span class="sc">+</span></span>
<span id="cb157-36"><a href="#cb157-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="fl">1.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-99-1.png" class="img-fluid figure-img" width="312"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>There just isn’t a lot of phylogenetic covariance for brain sizes, at least according to this model and these data. As a result, the phylogenetic distance doesn’t completely explain away the association between group size and brain size, as it did in the Brownian motion model. (p.&nbsp;485)</p>
</blockquote>
<p>For a thorough discussion on how to fit phylogenetic models with <strong>brms</strong>, particularly within the multilevel paradigm, see Bürkner’s <span class="citation" data-cites="Bürkner2022Phylogenetic">(<a href="references.html#ref-Bürkner2022Phylogenetic" role="doc-biblioref">2022d</a>)</span> vignette, <a href="https://CRAN.R-project.org/package=brms/vignettes/brms_phylogenetics.html"><em>Estimating phylogenetic multilevel models with brms</em></a>.</p>
</section>
</section>
<section id="summary-bonus-multilevel-growth-models-and-the-melsm" class="level2" data-number="14.6">
<h2 data-number="14.6" class="anchored" data-anchor-id="summary-bonus-multilevel-growth-models-and-the-melsm"><span class="header-section-number">14.6</span> <del>Summary</del> Bonus: Multilevel growth models and the MELSM</h2>
<p>To this point in the chapter and most of the text, the data have largely had a cross-sectional feel. In fairness, we did incorporate an element of time with the café example from model <code>b14.1</code> by looking at the differences between mornings and evenings. However, even then we collapsed across longer time spans, such as days, weeks, months, and so on. One of the two goals of this bonus section to provide a brief introduction multilevel models designed to express change over time. The particular brand of multilevel models we’ll focus on are often called multilevel growth models. Though we will focus on simple linear models, this basic framework can be generalized along many lines. The second goal is to build on our appreciation of covariance structures by introducing a class of multilevel models designed to investigate variation in variation called the mixed-effects location scale models (MELSM). For our final model, we get a little fancy and fit a multivariate MELSM.</p>
<section id="borrow-some-data" class="level3" data-number="14.6.1">
<h3 data-number="14.6.1" class="anchored" data-anchor-id="borrow-some-data"><span class="header-section-number">14.6.1</span> Borrow some data</h3>
<p>All the models in this bonus section are based on the paper by <span class="citation" data-cites="williamsBayesianMultivariateMixedeffects2021">Donald R. Williams, Martin, et al. (<a href="references.html#ref-williamsBayesianMultivariateMixedeffects2021" role="doc-biblioref">2021</a>)</span>, <em>Bayesian multivariate mixed-effects location scale modeling of longitudinal relations among affective traits, states, and physical activity</em> (you can find the preprint <a href="https://psyarxiv.com/4kfjp">here</a>). Williams and colleagues’ data and supporting scripts are available in the <code>example_analyses</code> folder from their OSF project at <a href="https://osf.io/3bmdh/">https://osf.io/3bmdh/</a>. You can also download the data from the <a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed/tree/master/data"><code>data</code> folder</a> of this ebook’s GitHub repo.</p>
<p>Load the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data from GitHub</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">url</span>(<span class="st">"https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed/raw/master/data/m_melsm_dat.rda"</span>))</span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb158-4"><a href="#cb158-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a scaled time variable</span></span>
<span id="cb158-5"><a href="#cb158-5" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> dat <span class="sc">|&gt;</span> </span>
<span id="cb158-6"><a href="#cb158-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">day01 =</span> (day <span class="sc">-</span> <span class="dv">2</span>) <span class="sc">/</span> <span class="fu">max</span>((day <span class="sc">-</span> <span class="dv">2</span>)))</span>
<span id="cb158-7"><a href="#cb158-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb158-8"><a href="#cb158-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look</span></span>
<span id="cb158-9"><a href="#cb158-9" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 13,033
Columns: 9
$ P_A.std   &lt;dbl&gt; 1.74740876, -0.23109384, 0.34155950, 0.45664827, -0.23484069, 1.12785344, 1.11272629, 0.55245938, 1.52371746, …
$ day       &lt;dbl&gt; 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 24, 25, 26, 27, 28, 29, 30, 33, 34, 35…
$ P_A.lag   &lt;dbl&gt; 0.7478597, 1.4674156, -0.3772641, 0.1286055, 0.3292090, 1.4107233, 0.7696644, 0.7304159, 1.0444039, 0.6519189,…
$ N_A.lag   &lt;dbl&gt; 0.25399356, -0.85363386, 0.96144592, -0.19620339, -0.16047348, -0.90365575, -0.77502804, -0.31053913, -0.68213…
$ steps.pm  &lt;dbl&gt; 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, 0.955171, …
$ steps.pmd &lt;dbl&gt; 0.59955783, -0.39471676, -1.51935866, -1.34423352, 0.41759695, -0.32310425, 0.37641976, 0.31766495, -0.4459848…
$ record_id &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ N_A.std   &lt;dbl&gt; -0.73357975, 0.53856559, 0.60161616, 0.27807249, 0.54674641, 0.05660701, -0.08417053, 0.10458994, -0.48557629,…
$ day01     &lt;dbl&gt; 0.00000000, 0.01020408, 0.02040816, 0.03061224, 0.04081633, 0.05102041, 0.06122449, 0.07142857, 0.08163265, 0.…</code></pre>
</div>
</div>
<p>These data are from 193 participants.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">distinct</span>(dat, record_id) <span class="sc">|&gt;</span> </span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
      n
  &lt;int&gt;
1   193</code></pre>
</div>
</div>
<p>Participants were asked to complete self-report ratings once a day for a few months. People varied by how many days they participated in the study, with number of days ranging from 8 to 99 and a median of 74.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>dat <span class="sc">|&gt;</span> </span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(record_id) <span class="sc">|&gt;</span> </span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">median =</span> <span class="fu">median</span>(n),</span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">min =</span> <span class="fu">min</span>(n),</span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  median   min   max
   &lt;int&gt; &lt;int&gt; &lt;int&gt;
1     74     8    99</code></pre>
</div>
</div>
<p>Here is a plot of that distribution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>dat <span class="sc">|&gt;</span> </span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(record_id) <span class="sc">|&gt;</span> </span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb164-4"><a href="#cb164-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> n)) <span class="sc">+</span></span>
<span id="cb164-5"><a href="#cb164-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">fill =</span> <span class="st">"#B1934A"</span>) <span class="sc">+</span></span>
<span id="cb164-6"><a href="#cb164-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">"number of days"</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb164-7"><a href="#cb164-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pearl_earring</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-104-1.png" class="img-fluid figure-img" width="384"></p>
</figure>
</div>
</div>
</div>
<p>Our primary variables of interest were taken from the Positive and Negative Affect Schedule <span class="citation" data-cites="watsonPANASDevelopment1988">(PANAS, <a href="references.html#ref-watsonPANASDevelopment1988" role="doc-biblioref">Watson et al., 1988</a>)</span>, which is widely used in certain areas of psychology to measure mood or emotion. In this study, participants completed the PANAS once a day by endorsing the extent to which they experienced various positive (e.g., excited, inspired) and negative (e.g., upset, afraid) emotional states. These responses are summed into two composite scores: positive affect (PA) and negative affect (NA). In the current data, the standardized versions of these scores are in the <code>P_A.std</code> and <code>N_A.std</code> columns, respectively. To get a sense of what these look like, here are the daily <code>N_A.std</code> scores from a random sample of 16 participants.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">14</span>)</span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a>dat <span class="sc">|&gt;</span> </span>
<span id="cb165-4"><a href="#cb165-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nest</span>(<span class="at">data =</span> <span class="fu">c</span>(P_A.std, day, P_A.lag, N_A.lag, steps.pm, steps.pmd, N_A.std, day01)) <span class="sc">|&gt;</span> </span>
<span id="cb165-5"><a href="#cb165-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">16</span>) <span class="sc">|&gt;</span> </span>
<span id="cb165-6"><a href="#cb165-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(data) <span class="sc">|&gt;</span> </span>
<span id="cb165-7"><a href="#cb165-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb165-8"><a href="#cb165-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> day, <span class="at">y =</span> N_A.lag)) <span class="sc">+</span></span>
<span id="cb165-9"><a href="#cb165-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"#80A0C7"</span>) <span class="sc">+</span></span>
<span id="cb165-10"><a href="#cb165-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"#FCF9F0"</span>, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb165-11"><a href="#cb165-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"negative affect (standardized)"</span>) <span class="sc">+</span></span>
<span id="cb165-12"><a href="#cb165-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> record_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-105-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="conventional-multilevel-growth-model" class="level3" data-number="14.6.2">
<h3 data-number="14.6.2" class="anchored" data-anchor-id="conventional-multilevel-growth-model"><span class="header-section-number">14.6.2</span> Conventional multilevel growth model</h3>
<p>In the social sciences, a typical way to analyze data like these is with a multilevel growth model in which participants vary in their intercepts (starting point) and time slopes (change over time). In the sample of the data, above, it looks like most participants have fairly constant levels of NA over time (i.e., near-zero slopes) but some (e.g., #128 and 147) show some evidence of systemic decreases in NA (i.e., negative slopes). There is also some variation in starting points, though most of the participants in this subset of the data seemed to have endorsed relatively low levels of NA both at baseline and throughout the study.</p>
<p>We want a model that can capture all these kinds of variation. Eventually, we will fit a model that accounts for both PA and NA. But to keep things simple while we’re warming up, we will restrict our focus to NA. If we let <span class="math inline">\(\text{NA}_{ij}\)</span> be the standardized NA score for the <span class="math inline">\(i\)</span>th participant on the <span class="math inline">\(j\)</span>th day, our first Bayesian multilevel growth model will follow the form</p>
<p><span class="math display">\[\begin{align*}
\text{NA}_{ij} &amp; \sim \operatorname{Normal}(\mu_{ij}, \sigma) \\
\mu_{ij}       &amp; = \beta_0 + \beta_1 \text{time}_{ij} + \color{#A65141}{u_{0i} + u_{1i} \text{time}_{ij}} \\
\sigma &amp; = \sigma_\epsilon \\
\color{#A65141}{\begin{bmatrix} u_{0i} \\ u_{1i} \end{bmatrix}} &amp; \color{#A65141}\sim \color{#A65141}{\operatorname{MVNormal}\begin{pmatrix} \begin{bmatrix} 0 \\ 0 \end{bmatrix}, \mathbf S \mathbf R \mathbf S \end{pmatrix}} \\
\mathbf S &amp; = \begin{bmatrix} \sigma_0 &amp; 0 \\ 0 &amp; \sigma_1 \end{bmatrix} \\
\mathbf R &amp; = \begin{bmatrix} 1 &amp; \rho_{12} \\ \rho_{21} &amp; 1 \end{bmatrix} \\
\beta_0   &amp; \sim \operatorname{Normal}(0, 0.2) \\
\beta_1   &amp; \sim \operatorname{Normal}(0, 1) \\
\sigma_0 \text{ and } \sigma_1 &amp; \sim \operatorname{Exponential}(1) \\
\sigma_\epsilon &amp; \sim \operatorname{Exponential}(1) \\
\mathbf R &amp; \sim \operatorname{LKJ}(2),
\end{align*}\]</span></p>
<p>where <span class="math inline">\(\beta_0\)</span> is the intercept (i.e., starting point) and <span class="math inline">\(u_{0i}\)</span> captures variations in that intercept across participants. Similarly, <span class="math inline">\(\beta_1\)</span> is the slope depicting linear change in <span class="math inline">\(\text{NA}\)</span> across time and <span class="math inline">\(u_{1i}\)</span> captures variations in that linear change across participants. The <span class="math inline">\(u_{0i}\)</span> and <span class="math inline">\(u_{1i}\)</span> parameters are modeled as multivariate normal with zero means (i.e., they are deviations from the population parameters) and standard deviations <span class="math inline">\(\sigma_0\)</span> and <span class="math inline">\(\sigma_1\)</span>. We express the correlation between those two group-level <span class="math inline">\(\sigma\)</span> parameters with <span class="math inline">\(\mathbf R\)</span>, the symmetric correlation matrix. Here we just have one correlation, <span class="math inline">\(\rho_{21}\)</span>, which is the same as <span class="math inline">\(\rho_{12}\)</span>. Finally, variation not accounted for by the other parameters is captured by the single parameter <span class="math inline">\(\sigma_\epsilon\)</span>, which is often just called <span class="math inline">\(\epsilon\)</span>.</p>
<p>We have two variables measuring time in these data. The <code>day</code> variable measures time by integers, ranging from 2 to 100. To make it a little easier to set the priors and fit the model with Stan, we have a rescaled version of the variable, <code>day01</code>, which ranges from 0 to 1. In this way, <span class="math inline">\(\beta_0\)</span> is the value for the first day in the data set and <span class="math inline">\(\beta_1\)</span> is the expected change by the end of the collection (i.e., the 100th day). For consistency, we are largely following McElreath’s weakly-regularizing approach to priors.</p>
<p>You may have noticed my statistical notation differs a bit from McElreath’s, here. It follows a blend of sensibilities from McElreath, Williams, and from the notation I used in my <span class="citation" data-cites="kurzAppliedLongitudinalDataAnalysis2026">(<a href="references.html#ref-kurzAppliedLongitudinalDataAnalysis2026" role="doc-biblioref">2026</a>)</span> translation](https://solomon.quarto.pub/alda/) of Singer and Willett’s <span class="citation" data-cites="singerAppliedLongitudinalData2003">(<a href="references.html#ref-singerAppliedLongitudinalData2003" role="doc-biblioref">2003</a>)</span> text, <a href="https://oxford.universitypressscholarship.com/view/10.1093/acprof:oso/9780195152968.001.0001/acprof-9780195152968"><em>Applied longitudinal data analysis: Modeling change and event occurrence</em></a>. I hope it’s clear.</p>
<p>Here is how to fit our simple multilevel growth model with <strong>brms</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.12</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat,</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a>  N_A.std <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">|</span> record_id),</span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb166-6"><a href="#cb166-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b),</span>
<span id="cb166-7"><a href="#cb166-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd),</span>
<span id="cb166-8"><a href="#cb166-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma),</span>
<span id="cb166-9"><a href="#cb166-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor)),</span>
<span id="cb166-10"><a href="#cb166-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">3000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb166-11"><a href="#cb166-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb166-12"><a href="#cb166-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.12"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the summary.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity 
Formula: N_A.std ~ 1 + day01 + (1 + day01 | record_id) 
   Data: dat (Number of observations: 13033) 
  Draws: 4 chains, each with iter = 3000; warmup = 1000; thin = 1;
         total post-warmup draws = 8000

Multilevel Hyperparameters:
~record_id (Number of levels: 193) 
                     Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sd(Intercept)            0.78      0.04     0.70     0.86 1.00     1741     3350
sd(day01)                0.65      0.05     0.57     0.75 1.00     3211     5311
cor(Intercept,day01)    -0.34      0.08    -0.49    -0.18 1.00     3101     4725

Regression Coefficients:
          Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept     0.03      0.05    -0.07     0.14 1.01      847     1547
day01        -0.16      0.06    -0.26    -0.05 1.00     2872     4656

Further Distributional Parameters:
      Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma     0.61      0.00     0.60     0.62 1.00    13236     5452

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Hopefully it makes sense that the population-level intercept (<span class="math inline">\(\beta_0\)</span>) is near zero. It would be odd if it wasn’t given these are standardized data. The coefficient for <code>day01</code> (<span class="math inline">\(\beta_1\)</span>) is mildly negative, suggesting an overall trend for participants to endorse lower NA scores over time.</p>
<p>The two group-level <span class="math inline">\(\sigma\)</span> parameters are fairly large given the scale of the data. They suggest participants varied quite a bit in terms of both intercepts and slopes. They also have a moderate negative correlation, suggesting that participants with higher intercepts tended to have more negative slopes.</p>
<p>To get a sense of the model, we’ll plot the posterior means for each participants’ fitted trajectory across time (thin lines), along with the population-average trajectory (thick line).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>nd <span class="ot">&lt;-</span> dat <span class="sc">|&gt;</span> </span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(record_id, day01)</span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a><span class="fu">fitted</span>(b14<span class="fl">.12</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb169-5"><a href="#cb169-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb169-6"><a href="#cb169-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(nd) <span class="sc">|&gt;</span> </span>
<span id="cb169-7"><a href="#cb169-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb169-8"><a href="#cb169-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> day01, <span class="at">y =</span> Estimate, <span class="at">group =</span> record_id)) <span class="sc">+</span></span>
<span id="cb169-9"><a href="#cb169-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb169-10"><a href="#cb169-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">xend =</span> <span class="dv">1</span>,</span>
<span id="cb169-11"><a href="#cb169-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">y =</span> <span class="fu">fixef</span>(b14<span class="fl">.12</span>)[<span class="dv">1</span>, <span class="dv">1</span>],</span>
<span id="cb169-12"><a href="#cb169-12" aria-hidden="true" tabindex="-1"></a>               <span class="at">yend =</span> <span class="fu">fixef</span>(b14<span class="fl">.12</span>)[<span class="dv">1</span>, <span class="dv">1</span>] <span class="sc">+</span> <span class="fu">fixef</span>(b14<span class="fl">.12</span>)[<span class="dv">2</span>, <span class="dv">1</span>],</span>
<span id="cb169-13"><a href="#cb169-13" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="st">"#80A0C7"</span>, <span class="at">linewidth =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb169-14"><a href="#cb169-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb169-15"><a href="#cb169-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"negative affect (standardized)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-108-1.png" class="img-fluid figure-img" width="384"></p>
</figure>
</div>
</div>
</div>
<p>If you look back up to the model summary from before the plot, the one parameter we didn’t focus on was the lone <code>sigma</code> parameter at the bottom. That’s our <span class="math inline">\(\sigma_\epsilon\)</span>, which captures the variation in NA not accounted for by the intercepts, slopes, and their correlation. An important characteristic of the conventional multilevel growth model is that <span class="math inline">\(\sigma_\epsilon\)</span> does not vary across persons, occasions, or other variables. To give a sense of why this might not be the best assumption, let’s take a focused look at the model implied trajectories for two participants. Here we will take cues from some Figure 4.8 from way back in <a href="04.html#sec-Prediction-intervals" class="quarto-xref"><span>Section 4.4.3.5</span></a>. We will plot the original data atop both the fitted lines and their 95% intervals, which expresses the mean structure, along with the 95% posterior predictive interval, which expresses the uncertainty of the <span class="math inline">\(\sigma_\epsilon\)</span> parameter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a>nd <span class="ot">&lt;-</span> dat <span class="sc">|&gt;</span> </span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(record_id <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">115</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(record_id, N_A.std, day01)</span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a><span class="fu">bind_cols</span>(</span>
<span id="cb170-6"><a href="#cb170-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># `fitted`</span></span>
<span id="cb170-7"><a href="#cb170-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fitted</span>(b14<span class="fl">.12</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb170-8"><a href="#cb170-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(),</span>
<span id="cb170-9"><a href="#cb170-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># `predict`</span></span>
<span id="cb170-10"><a href="#cb170-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>(b14<span class="fl">.12</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb170-11"><a href="#cb170-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb170-12"><a href="#cb170-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(Q2<span class="fl">.5</span><span class="sc">:</span>Q97<span class="fl">.5</span>) <span class="sc">|&gt;</span> </span>
<span id="cb170-13"><a href="#cb170-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_names</span>(<span class="st">"p_lower"</span>, <span class="st">"p_upper"</span>)</span>
<span id="cb170-14"><a href="#cb170-14" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb170-15"><a href="#cb170-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(nd) <span class="sc">|&gt;</span> </span>
<span id="cb170-16"><a href="#cb170-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb170-17"><a href="#cb170-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> day01)) <span class="sc">+</span></span>
<span id="cb170-18"><a href="#cb170-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">y =</span> Estimate, <span class="at">ymin =</span> Q2<span class="fl">.5</span>, <span class="at">ymax =</span> Q97<span class="fl">.5</span>),</span>
<span id="cb170-19"><a href="#cb170-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">stat =</span> <span class="st">"identity"</span>,</span>
<span id="cb170-20"><a href="#cb170-20" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">fill =</span> <span class="st">"#8B9DAF"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb170-21"><a href="#cb170-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> p_lower, <span class="at">ymax =</span> p_upper),</span>
<span id="cb170-22"><a href="#cb170-22" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">fill =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb170-23"><a href="#cb170-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> N_A.std),</span>
<span id="cb170-24"><a href="#cb170-24" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb170-25"><a href="#cb170-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb170-26"><a href="#cb170-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"negative affect (standardized)"</span>) <span class="sc">+</span></span>
<span id="cb170-27"><a href="#cb170-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> record_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-109-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Because of our fixed <span class="math inline">\(\sigma_\epsilon\)</span> parameter, the 95% posterior predictive interval is the same width for both participants. Yet look how closely participant the data points for participant 30 cluster not only within the center region of the posterior prediction intervals, but also almost completely within the 95% interval for the fitted line. In contrast, notice how much more spread out the data points for participant 115 are, and how many of them extend well beyond the posterior predictive interval. This difference in variability is ignored by conventional growth models. However, there’s no reason we can’t adjust our model to capture this kind of person-level variability, too. Enter the MELSM.</p>
</section>
<section id="learn-more-about-your-data-with-the-melsm" class="level3" data-number="14.6.3">
<h3 data-number="14.6.3" class="anchored" data-anchor-id="learn-more-about-your-data-with-the-melsm"><span class="header-section-number">14.6.3</span> Learn more about your data with the MELSM</h3>
<p>Mixed-effects location scale models (MELSMs) have their origins in the work of <a href="https://health.uchicago.edu/faculty/donald-hedeker-phd">Donald Hedeker</a> and colleagues <span class="citation" data-cites="hedekerApplicationMixedeffectsLocation2008 hedekerModelingWithinsubjectVariance2012">(<a href="references.html#ref-hedekerApplicationMixedeffectsLocation2008" role="doc-biblioref">Hedeker et al., 2008</a>, <a href="references.html#ref-hedekerModelingWithinsubjectVariance2012" role="doc-biblioref">2012</a>)</span>. <span class="citation" data-cites="rastModelingIndividualDifferences2012">Rast et al. (<a href="references.html#ref-rastModelingIndividualDifferences2012" role="doc-biblioref">2012</a>)</span> showcased an early application of the framework to the BUGS/JAGS software. More recently <a href="https://rastlab.ucdavis.edu/">Philippe Rast</a> and colleagues (particularly <a href="https://donaldrwilliams.github.io/">Donald Williams</a>) have adapted this approach for use within the Stan/<strong>brms</strong> ecosystem <span class="citation" data-cites="williamsBayesianNonlinearMixedeffects2019a williams2021beneath williams2020BayesianMulitivariate williams2022PuttingTheIndividual">(<a href="references.html#ref-williamsBayesianNonlinearMixedeffects2019a" role="doc-biblioref">Donald R. Williams et al., 2019</a>; <a href="references.html#ref-williams2020BayesianMulitivariate" role="doc-biblioref">Donald R. Williams et al., 2020</a>, <a href="references.html#ref-williams2022PuttingTheIndividual" role="doc-biblioref">2022</a>; <a href="references.html#ref-williams2021beneath" role="doc-biblioref">Donald R. Williams, Mulder, et al., 2021</a>)</span>.</p>
<p>Within the <strong>brms</strong> framework, MELSMs apply a distributional modeling approach <span class="citation" data-cites="Bürkner2022Distributional">(see <a href="references.html#ref-Bürkner2022Distributional" role="doc-biblioref">Bürkner, 2022b</a>)</span> to the multilevel growth model. Not only are parameters from the mean structure allowed to vary across groups, but parameters applied to <span class="math inline">\(\sigma\)</span> are allowed to vary across groups, too. Do you remember the little practice model <code>b10.1</code> from <a href="10.html#sec-Linking-linear-models-to-distributions" class="quarto-xref"><span>Section 10.2.2</span></a>? We simulated Gaussian data for two groups with the same mean parameter but different parameters for <span class="math inline">\(\sigma\)</span>. If you check back in that section, you’ll see that the <strong>brms</strong> default was to model <span class="math inline">\(\log \sigma\)</span> in that case. This is smart because when you define a model for <span class="math inline">\(\sigma\)</span>, you want to use a link function that ensures the predictions will be stay at zero and above. The MELSM approach of Hedeker, Rast, Williams and friends applies this logic to <span class="math inline">\(\sigma_\epsilon\)</span> in multilevel growth models. However, not only can <span class="math inline">\(\sigma_\epsilon\)</span> vary across groups in a fixed-effects sort of way, we can use multilevel partial pooling, too.</p>
<p>To get a sense of what this looks like, we’ll augment our previous model, but this time allowing <span class="math inline">\(\sigma_\epsilon\)</span> to vary across participants. You might express the updated statistical model as</p>
<p><span class="math display">\[\begin{align*}
\text{NA}_{ij} &amp; \sim \operatorname{Normal}(\mu_{ij}, \color{#A65141}{\sigma_{i}})\\
\mu_{ij} &amp; = \beta_0 + \beta_1 \text{time}_{ij} + u_{0i} + u_{1i} \text{time}_{ij} \\
\color{#A65141}{\log (\sigma_i )} &amp; \color{#A65141}= \color{#A65141}{\eta_0 + u_{2i}} \\
\begin{bmatrix} u_{0i} \\ u_{1i} \\ \color{#A65141}{u_{2i}} \end{bmatrix} &amp; \sim \operatorname{MVNormal}\begin{pmatrix} \begin{bmatrix} 0 \\ 0 \\ 0 \end{bmatrix}, \mathbf S \mathbf R \mathbf S \end{pmatrix} \\
\mathbf S &amp; = \begin{bmatrix} \sigma_0 &amp; 0 &amp; 0 \\ 0 &amp; \sigma_1 &amp; 0 \\ 0 &amp; 0 &amp; \sigma_2 \end{bmatrix} \\
\mathbf R &amp; = \begin{bmatrix} 1 &amp; \rho_{12} &amp; \rho_{13} \\ \rho_{21} &amp; 1 &amp; \rho_{23} \\ \rho_{31} &amp; \rho_{32} &amp; 1 \end{bmatrix} \\
\beta_0   &amp; \sim \operatorname{Normal}(0, 0.2) \\
\beta_1 \text{and } \eta_0 &amp; \sim \operatorname{Normal}(0, 1) \\
\sigma_0,\dots, \sigma_2     &amp; \sim \operatorname{Exponential}(1) \\
\mathbf R &amp; \sim \operatorname{LKJ}(2).
\end{align*}\]</span></p>
<p>In the opening likelihood statement from the prior model, we simply set <span class="math inline">\(\text{NA}_{ij} \sim \operatorname{Normal}\begin{pmatrix} \mu_{ij}, \sigma \end{pmatrix}\)</span>. For our first MELSM, we now refer to <span class="math inline">\(\sigma_i\)</span>, meaning the levels of variation not accounted for by the mean structure can vary across participants (hence the <span class="math inline">\(i\)</span> subscript). Two lines down, we see the formula for <span class="math inline">\(\log \begin{pmatrix} \sigma_i \end{pmatrix}\)</span> contains population-level intercept, <span class="math inline">\(\eta_0\)</span>, and participant-specific deviations around that parameter, <span class="math inline">\(u_{2i}\)</span>. In the next three lines, the plot deepens. We see that all three participant-level deviations, <span class="math inline">\(u_{0i},\dots,u_{2i}\)</span> are multivariate normal with means set to zero and variation expressed in the parameters <span class="math inline">\(\sigma_0,\dots,\sigma_2\)</span> of the <span class="math inline">\(\mathbf S\)</span> matrix. In the <span class="math inline">\(\mathbf R\)</span> matrix, we now have three correlation parameters, with <span class="math inline">\(\rho_{31}\)</span> and <span class="math inline">\(\rho_{32}\)</span> allowing us to assess the correlations among individual differences in variability and individual differences in starting points and change over time, respectively. Let’s fit the model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.13</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat,</span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb171-4"><a href="#cb171-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bf</span>(N_A.std <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">|</span>i<span class="sc">|</span> record_id),</span>
<span id="cb171-5"><a href="#cb171-5" aria-hidden="true" tabindex="-1"></a>     sigma <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span>i<span class="sc">|</span> record_id)),</span>
<span id="cb171-6"><a href="#cb171-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb171-7"><a href="#cb171-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b),</span>
<span id="cb171-8"><a href="#cb171-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd),</span>
<span id="cb171-9"><a href="#cb171-9" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb171-10"><a href="#cb171-10" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> Intercept, <span class="at">dpar =</span> sigma),</span>
<span id="cb171-11"><a href="#cb171-11" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">dpar =</span> sigma),</span>
<span id="cb171-12"><a href="#cb171-12" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb171-13"><a href="#cb171-13" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor)),</span>
<span id="cb171-14"><a href="#cb171-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">3000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb171-15"><a href="#cb171-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb171-16"><a href="#cb171-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.13"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We should note a few things about the <code>brm()</code> syntax. First, because we modeled both <span class="math inline">\(\mu_{ij}\)</span> and <span class="math inline">\(\sigma_i\)</span>, we nested both model formulas within the <code>bf()</code> function. Second, because the <strong>brms</strong> default is to use the log link when modeling <span class="math inline">\(\sigma_i\)</span>, there was no need to explicitly set it that way in the <code>family</code> line. However, we could have if we wanted to. Third, notice our use of the <code>|i|</code> syntax within the parentheses in the <code>formula</code> lines. If we had used the conventional <code>|</code> syntax, that would have not allowed our <span class="math inline">\(u_{2i}\)</span> parameters to correlate with <span class="math inline">\(u_{0i}\)</span> and <span class="math inline">\(u_{1i}\)</span> from the mean structure. It would have effectively set <span class="math inline">\(\rho_{31} = \rho_{32} = 0\)</span>. Finally, notice how within the <code>prior()</code> functions, we explicitly referred to those for the new <span class="math inline">\(\sigma\)</span> structure with the <code>dpar = sigma</code> operator.</p>
<p>Okay, time to check the model summary.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity; sigma = log 
Formula: N_A.std ~ 1 + day01 + (1 + day01 | i | record_id) 
         sigma ~ 1 + (1 | i | record_id)
   Data: dat (Number of observations: 13033) 
  Draws: 4 chains, each with iter = 3000; warmup = 1000; thin = 1;
         total post-warmup draws = 8000

Multilevel Hyperparameters:
~record_id (Number of levels: 193) 
                               Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sd(Intercept)                      0.76      0.04     0.69     0.84 1.00      697     1505
sd(day01)                          0.60      0.04     0.52     0.69 1.01     1215     2668
sd(sigma_Intercept)                0.69      0.03     0.63     0.77 1.01      704     1320
cor(Intercept,day01)              -0.33      0.08    -0.48    -0.17 1.01      725     1717
cor(Intercept,sigma_Intercept)     0.61      0.05     0.52     0.69 1.00      896     1870
cor(day01,sigma_Intercept)        -0.09      0.08    -0.25     0.06 1.01      672     1514

Regression Coefficients:
                Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept           0.03      0.05    -0.08     0.13 1.02      258      655
sigma_Intercept    -0.79      0.05    -0.88    -0.69 1.01      429     1096
day01              -0.16      0.05    -0.26    -0.06 1.00      749     1809

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>The ‘sigma_Intercept’ lines in the ‘Population-Level Effects’ section is the summary for our <span class="math inline">\(\eta_0\)</span> parameter. To get a sense of what this means out of the log space, just exponentiate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fixef</span>(b14<span class="fl">.13</span>)[<span class="st">"sigma_Intercept"</span>, <span class="sc">-</span><span class="dv">2</span>] <span class="sc">|&gt;</span> <span class="fu">exp</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Estimate      Q2.5     Q97.5 
0.4558183 0.4133110 0.5021022 </code></pre>
</div>
</div>
<p>To get a sense of the variation in that parameter across participants [i.e., <span class="math inline">\(\exp(\eta_0 + u_{2i})\)</span>], it’s best to plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coef</span>(b14<span class="fl">.13</span>)<span class="sc">$</span>record_id[, , <span class="st">"sigma_Intercept"</span>] <span class="sc">|&gt;</span> </span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">exp</span>() <span class="sc">|&gt;</span> </span>
<span id="cb176-3"><a href="#cb176-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb176-4"><a href="#cb176-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(Estimate) <span class="sc">|&gt;</span> </span>
<span id="cb176-5"><a href="#cb176-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">rank =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">n</span>()) <span class="sc">|&gt;</span> </span>
<span id="cb176-6"><a href="#cb176-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb176-7"><a href="#cb176-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> rank, <span class="at">y =</span> Estimate, <span class="at">ymin =</span> Q2<span class="fl">.5</span>, <span class="at">ymax =</span> Q97<span class="fl">.5</span>)) <span class="sc">+</span></span>
<span id="cb176-8"><a href="#cb176-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="at">color =</span> <span class="st">"#EEDA9D"</span>, <span class="at">linewidth =</span> <span class="fl">0.4</span>, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">20</span>) <span class="sc">+</span></span>
<span id="cb176-9"><a href="#cb176-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">"participants ranked by posterior mean"</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb176-10"><a href="#cb176-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">exp</span>(eta[<span class="dv">0</span>]<span class="sc">+</span><span class="fu">italic</span>(u)[<span class="dv">2</span>][<span class="fu">italic</span>(i)])))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-113-1.png" class="img-fluid figure-img" width="432"></p>
</figure>
</div>
</div>
</div>
<p>Looks like there’s a lot of variation in a parameter that was formerly fixed across participants as a single value <span class="math inline">\(\sigma_\epsilon\)</span>. Here’s what this looks like in terms of our posterior predictive distributions for participants 30 and 115, from before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>nd <span class="ot">&lt;-</span> dat <span class="sc">|&gt;</span> </span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(record_id <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">115</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(record_id, N_A.std, day01)</span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a><span class="fu">bind_cols</span>(</span>
<span id="cb177-6"><a href="#cb177-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fitted</span>(b14<span class="fl">.13</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb177-7"><a href="#cb177-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(),</span>
<span id="cb177-8"><a href="#cb177-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>(b14<span class="fl">.13</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb177-9"><a href="#cb177-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb177-10"><a href="#cb177-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(Q2<span class="fl">.5</span><span class="sc">:</span>Q97<span class="fl">.5</span>) <span class="sc">|&gt;</span> </span>
<span id="cb177-11"><a href="#cb177-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_names</span>(<span class="st">"p_lower"</span>, <span class="st">"p_upper"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb177-12"><a href="#cb177-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(nd) <span class="sc">|&gt;</span> </span>
<span id="cb177-13"><a href="#cb177-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb177-14"><a href="#cb177-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> day01)) <span class="sc">+</span></span>
<span id="cb177-15"><a href="#cb177-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">y =</span> Estimate, <span class="at">ymin =</span> Q2<span class="fl">.5</span>, <span class="at">ymax =</span> Q97<span class="fl">.5</span>),</span>
<span id="cb177-16"><a href="#cb177-16" aria-hidden="true" tabindex="-1"></a>              <span class="at">stat =</span> <span class="st">"identity"</span>,</span>
<span id="cb177-17"><a href="#cb177-17" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">fill =</span> <span class="st">"#8B9DAF"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb177-18"><a href="#cb177-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> p_lower, <span class="at">ymax =</span> p_upper),</span>
<span id="cb177-19"><a href="#cb177-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">fill =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb177-20"><a href="#cb177-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> N_A.std),</span>
<span id="cb177-21"><a href="#cb177-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb177-22"><a href="#cb177-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb177-23"><a href="#cb177-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"negative affect (standardized)"</span>) <span class="sc">+</span></span>
<span id="cb177-24"><a href="#cb177-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> record_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-114-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>That’s a big improvement. Let’s expand our skill set. Within the MELSM paradigm, one can use multiple variables to model participant-specific variation. Here we’ll add in our time variable.</p>
<p><span class="math display">\[\begin{align*}
\text{NA}_{ij} &amp; \sim \operatorname{Normal}(\mu_{ij}, \color{#A65141}{\sigma_{ij}}) \\
\mu_{ij} &amp; = \beta_0 + \beta_1 \text{time}_{ij} + u_{0i} + u_{1i} \text{time}_{ij} \\
\log(\sigma_{i\color{#A65141}j}) &amp; = \eta_0 + \color{#A65141}{\eta_1 \text{time}_{ij}} + u_{2i} + \color{#A65141}{u_{3i} \text{time}_{ij}} \\
\begin{bmatrix} u_{0i} \\ u_{1i} \\ u_{2i} \\ \color{#A65141}{u_{3i}} \end{bmatrix} &amp; \sim \operatorname{MVNormal}\begin{pmatrix} \begin{bmatrix} 0 \\ 0 \\ 0 \\ 0 \end{bmatrix}, \mathbf S \mathbf R \mathbf S \end{pmatrix} \\
\mathbf S &amp; = \begin{bmatrix} \sigma_0 &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; \sigma_1 &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; \sigma_2 &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; \sigma_3 \end{bmatrix} \\
\mathbf R &amp; = \begin{bmatrix} 1 &amp; \rho_{12} &amp; \rho_{13} &amp; \rho_{14} \\ \rho_{21} &amp; 1 &amp; \rho_{23} &amp; \rho_{24} \\ \rho_{31} &amp; \rho_{32} &amp; 1 &amp; \rho_{34} \\ \rho_{41} &amp; \rho_{42} &amp; \rho_{43} &amp; 1 \end{bmatrix} \\
\beta_0   &amp; \sim \operatorname{Normal}(0, 0.2) \\
\beta_1, \eta_0, \text{and } \eta_1 &amp; \sim \operatorname{Normal}(0, 1) \\
\sigma_0,\dots, \sigma_3 &amp; \sim \operatorname{Exponential}(1) \\
\mathbf R &amp; \sim \operatorname{LKJ}(2).
\end{align*}\]</span></p>
<p>Note how in the very first line we are now speaking in terms of <span class="math inline">\(\sigma_{ij}\)</span>. Variation in the criterion <span class="math inline">\(\text{NA}_{ij}\)</span> not accounted for by the mean structure can now vary across participants, <span class="math inline">\(i\)</span>, and time, <span class="math inline">\(j\)</span>. This results in four <span class="math inline">\(u_{xi}\)</span> terms, a <span class="math inline">\(4 \times 4\)</span> <span class="math inline">\(\mathbf S\)</span> matrix and a <span class="math inline">\(4 \times 4\)</span> <span class="math inline">\(\mathbf R\)</span> matrix. Here’s how you might fit the model with <code>brms::brm()</code>. <strong>Warning</strong>: it’ll probably take a couple hours.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.14</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat,</span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb178-4"><a href="#cb178-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bf</span>(N_A.std <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">|</span>i<span class="sc">|</span> record_id),</span>
<span id="cb178-5"><a href="#cb178-5" aria-hidden="true" tabindex="-1"></a>     sigma <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">|</span>i<span class="sc">|</span> record_id)),</span>
<span id="cb178-6"><a href="#cb178-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb178-7"><a href="#cb178-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b),</span>
<span id="cb178-8"><a href="#cb178-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd),</span>
<span id="cb178-9"><a href="#cb178-9" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb178-10"><a href="#cb178-10" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> Intercept, <span class="at">dpar =</span> sigma),</span>
<span id="cb178-11"><a href="#cb178-11" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b, <span class="at">dpar =</span> sigma),</span>
<span id="cb178-12"><a href="#cb178-12" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">dpar =</span> sigma),</span>
<span id="cb178-13"><a href="#cb178-13" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb178-14"><a href="#cb178-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor)),</span>
<span id="cb178-15"><a href="#cb178-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">3000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb178-16"><a href="#cb178-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb178-17"><a href="#cb178-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.14"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>At this point, <code>print()</code> is starting to return a lot of output.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(b14<span class="fl">.14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Family: gaussian 
  Links: mu = identity; sigma = log 
Formula: N_A.std ~ 1 + day01 + (1 + day01 | i | record_id) 
         sigma ~ 1 + day01 + (1 + day01 | i | record_id)
   Data: dat (Number of observations: 13033) 
  Draws: 4 chains, each with iter = 3000; warmup = 1000; thin = 1;
         total post-warmup draws = 8000

Multilevel Hyperparameters:
~record_id (Number of levels: 193) 
                                 Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sd(Intercept)                        0.76      0.04     0.68     0.84 1.00     1333     2618
sd(day01)                            0.60      0.04     0.52     0.69 1.00     2719     4834
sd(sigma_Intercept)                  0.70      0.04     0.63     0.78 1.00     2062     4164
sd(sigma_day01)                      0.36      0.04     0.29     0.44 1.00     5250     5865
cor(Intercept,day01)                -0.30      0.08    -0.46    -0.14 1.00     1506     3527
cor(Intercept,sigma_Intercept)       0.64      0.05     0.54     0.73 1.00     2000     3336
cor(day01,sigma_Intercept)          -0.20      0.08    -0.35    -0.04 1.00     1778     3586
cor(Intercept,sigma_day01)          -0.16      0.11    -0.36     0.05 1.00     5267     6138
cor(day01,sigma_day01)               0.61      0.09     0.42     0.77 1.00     4446     5372
cor(sigma_Intercept,sigma_day01)    -0.15      0.10    -0.34     0.05 1.00     4772     5813

Regression Coefficients:
                Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept           0.03      0.06    -0.09     0.13 1.00      573     1085
sigma_Intercept    -0.75      0.05    -0.85    -0.65 1.00     1154     2338
day01              -0.15      0.05    -0.26    -0.05 1.00     1404     2793
sigma_day01        -0.11      0.04    -0.18    -0.03 1.00     4379     5267

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
</div>
<p>Our new line for ‘sigma_day01’ suggests there is a general trend for less variation in negative affect ratings over time. However, the ‘sd(sigma_day01)’ line in the ‘Group-Level Effects’ section indicates even this varies a bit across participants. At this point, a lot of the action is now in the estimates for the <span class="math inline">\(\mathbf R\)</span> matrix. Here that is in a coefficient plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior_summary</span>(b14<span class="fl">.14</span>) <span class="sc">|&gt;</span> </span>
<span id="cb181-2"><a href="#cb181-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb181-3"><a href="#cb181-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"par"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb181-4"><a href="#cb181-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(par, <span class="st">"cor_"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb181-5"><a href="#cb181-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">rho =</span> <span class="fu">str_c</span>(<span class="st">"(rho["</span>, <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">31</span>, <span class="dv">32</span>, <span class="dv">41</span>, <span class="dv">42</span>, <span class="dv">43</span>), <span class="st">"])"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb181-6"><a href="#cb181-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">par =</span> <span class="fu">str_c</span>(<span class="st">"'"</span>, par, <span class="st">"'~"</span>, rho)) <span class="sc">|&gt;</span> </span>
<span id="cb181-7"><a href="#cb181-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb181-8"><a href="#cb181-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Estimate, <span class="at">xmin =</span> Q2<span class="fl">.5</span>, <span class="at">xmax =</span> Q97<span class="fl">.5</span>, <span class="at">y =</span> par)) <span class="sc">+</span></span>
<span id="cb181-9"><a href="#cb181-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="at">color =</span> <span class="st">"#FCF9F0"</span>, </span>
<span id="cb181-10"><a href="#cb181-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>), <span class="at">linewidth =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>)) <span class="sc">+</span></span>
<span id="cb181-11"><a href="#cb181-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="at">color =</span> <span class="st">"#B1934A"</span>) <span class="sc">+</span></span>
<span id="cb181-12"><a href="#cb181-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe) <span class="sc">+</span></span>
<span id="cb181-13"><a href="#cb181-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb181-14"><a href="#cb181-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"marginal posterior"</span>,</span>
<span id="cb181-15"><a href="#cb181-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb181-16"><a href="#cb181-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">size =</span> <span class="dv">9</span>),</span>
<span id="cb181-17"><a href="#cb181-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-117-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>Note how we attached the statistical terms from the lower triangle of the <span class="math inline">\(\mathbf R\)</span> matrix to the names from the <strong>brms</strong> output. Coefficient plots like this are somewhat helpful with MELSM parameter summaries, like this. But they leave something to be desired and they won’t scale well. One alternative is to present the posterior means in a correlation matrix plot. Our first step to prepare for the plot is to extract and wrangle the posterior summaries.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a>levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"beta[0]"</span>, <span class="st">"beta[1]"</span>, <span class="st">"eta[0]"</span>, <span class="st">"eta[1]"</span>)</span>
<span id="cb182-2"><a href="#cb182-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-3"><a href="#cb182-3" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">&lt;-</span> <span class="fu">posterior_summary</span>(b14<span class="fl">.14</span>) <span class="sc">|&gt;</span> </span>
<span id="cb182-4"><a href="#cb182-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb182-5"><a href="#cb182-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"param"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb182-6"><a href="#cb182-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(param, <span class="st">"cor_"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb182-7"><a href="#cb182-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">param =</span> <span class="fu">str_remove</span>(param, <span class="st">"cor_record_id__"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb182-8"><a href="#cb182-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(param, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"left"</span>, <span class="st">"right"</span>), <span class="at">sep =</span> <span class="st">"__"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb182-9"><a href="#cb182-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb182-10"><a href="#cb182-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">left =</span> <span class="fu">case_when</span>(</span>
<span id="cb182-11"><a href="#cb182-11" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]"</span>,</span>
<span id="cb182-12"><a href="#cb182-12" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]"</span>,</span>
<span id="cb182-13"><a href="#cb182-13" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"sigma_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]"</span>),</span>
<span id="cb182-14"><a href="#cb182-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">right =</span> <span class="fu">case_when</span>(</span>
<span id="cb182-15"><a href="#cb182-15" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]"</span>,</span>
<span id="cb182-16"><a href="#cb182-16" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"sigma_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]"</span>,</span>
<span id="cb182-17"><a href="#cb182-17" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"sigma_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]"</span></span>
<span id="cb182-18"><a href="#cb182-18" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb182-19"><a href="#cb182-19" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb182-20"><a href="#cb182-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">formatC</span>(Estimate, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">format =</span> <span class="st">"f"</span>) <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">"0."</span>, <span class="st">"."</span>)) <span class="sc">|&gt;</span></span>
<span id="cb182-21"><a href="#cb182-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">left  =</span> <span class="fu">factor</span>(left, <span class="at">levels =</span> levels),</span>
<span id="cb182-22"><a href="#cb182-22" aria-hidden="true" tabindex="-1"></a>         <span class="at">right =</span> <span class="fu">factor</span>(right, <span class="at">levels =</span> levels)) <span class="sc">|&gt;</span></span>
<span id="cb182-23"><a href="#cb182-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">right =</span> <span class="fu">fct_rev</span>(right))</span>
<span id="cb182-24"><a href="#cb182-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-25"><a href="#cb182-25" aria-hidden="true" tabindex="-1"></a>rho</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     left   right   Estimate  Est.Error       Q2.5       Q97.5 label
1 beta[0] beta[1] -0.3046514 0.08256235 -0.4604108 -0.13857319  -.30
2 beta[0]  eta[0]  0.6415259 0.04699774  0.5447721  0.72880139   .64
3 beta[1]  eta[0] -0.1990983 0.07869625 -0.3502391 -0.04456139  -.20
4 beta[0]  eta[1] -0.1613560 0.10549884 -0.3639229  0.04753939  -.16
5 beta[1]  eta[1]  0.6093562 0.08862745  0.4223299  0.76793831   .61
6  eta[0]  eta[1] -0.1501403 0.09681035 -0.3351708  0.04597338  -.15</code></pre>
</div>
</div>
<p>Note how instead of naming the correlations in terms of <span class="math inline">\(\rho_{xx}\)</span>, we are now referring to them as the correlations of the deviations among the population parameters, <span class="math inline">\(\beta_0\)</span> through <span class="math inline">\(\eta_1\)</span>. I’m hoping this will make sense in the plot. Here it is.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span> </span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> left, <span class="at">y =</span> right)) <span class="sc">+</span></span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Estimate)) <span class="sc">+</span></span>
<span id="cb184-4"><a href="#cb184-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> label),</span>
<span id="cb184-5"><a href="#cb184-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb184-6"><a href="#cb184-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">drop =</span> F, <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe, <span class="at">position =</span> <span class="st">"top"</span>) <span class="sc">+</span></span>
<span id="cb184-7"><a href="#cb184-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">drop =</span> F, <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe) <span class="sc">+</span></span>
<span id="cb184-8"><a href="#cb184-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="fu">expression</span>(rho),</span>
<span id="cb184-9"><a href="#cb184-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">low =</span> <span class="st">"#59708b"</span>, <span class="at">mid =</span> <span class="st">"#FCF9F0"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>, <span class="at">midpoint =</span> <span class="dv">0</span>, </span>
<span id="cb184-10"><a href="#cb184-10" aria-hidden="true" tabindex="-1"></a>                       <span class="at">labels =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="st">""</span>, <span class="dv">0</span>, <span class="st">""</span>, <span class="dv">1</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb184-11"><a href="#cb184-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">expression</span>(<span class="st">"The lower triangle for "</span><span class="sc">*</span><span class="fu">bold</span>(R)),</span>
<span id="cb184-12"><a href="#cb184-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">subtitle =</span> <span class="st">"Note, each cell is summarized by</span><span class="sc">\n</span><span class="st">its posterior mean."</span>) <span class="sc">+</span></span>
<span id="cb184-13"><a href="#cb184-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb184-14"><a href="#cb184-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb184-15"><a href="#cb184-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-119-1.png" class="img-fluid figure-img" width="336"></p>
</figure>
</div>
</div>
</div>
<p>Interestingly, the strongest two associations involve variation around our <span class="math inline">\(\eta\)</span> parameters. The posterior mean for <span class="math inline">\(\rho_{31}\)</span> indicates participants with higher baseline levels of <span class="math inline">\(\text{NA}_{ij}\)</span> tend to vary more in their responses, particularly in the beginning. The posterior mean for <span class="math inline">\(\rho_{42}\)</span> indicates participants who show greater increases in their <span class="math inline">\(\text{NA}_{ij}\)</span> responses over time also tend to show greater relative increases in variation in those responses. You can get a little bit of a sense for this by returning once again to our participants 30 and 115.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>nd <span class="ot">&lt;-</span> dat <span class="sc">|&gt;</span> </span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(record_id <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">115</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb185-3"><a href="#cb185-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(record_id, N_A.std, day01)</span>
<span id="cb185-4"><a href="#cb185-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-5"><a href="#cb185-5" aria-hidden="true" tabindex="-1"></a><span class="fu">bind_cols</span>(</span>
<span id="cb185-6"><a href="#cb185-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fitted</span>(b14<span class="fl">.14</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb185-7"><a href="#cb185-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(),</span>
<span id="cb185-8"><a href="#cb185-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>(b14<span class="fl">.14</span>, <span class="at">newdata =</span> nd) <span class="sc">|&gt;</span> </span>
<span id="cb185-9"><a href="#cb185-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb185-10"><a href="#cb185-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(Q2<span class="fl">.5</span><span class="sc">:</span>Q97<span class="fl">.5</span>) <span class="sc">|&gt;</span> </span>
<span id="cb185-11"><a href="#cb185-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_names</span>(<span class="st">"p_lower"</span>, <span class="st">"p_upper"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb185-12"><a href="#cb185-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(nd) <span class="sc">|&gt;</span> </span>
<span id="cb185-13"><a href="#cb185-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb185-14"><a href="#cb185-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> day01)) <span class="sc">+</span></span>
<span id="cb185-15"><a href="#cb185-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">y =</span> Estimate, <span class="at">ymin =</span> Q2<span class="fl">.5</span>, <span class="at">ymax =</span> Q97<span class="fl">.5</span>),</span>
<span id="cb185-16"><a href="#cb185-16" aria-hidden="true" tabindex="-1"></a>              <span class="at">stat =</span> <span class="st">"identity"</span>,</span>
<span id="cb185-17"><a href="#cb185-17" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#8B9DAF"</span>, <span class="at">fill =</span> <span class="st">"#8B9DAF"</span>, <span class="at">linewidth =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb185-18"><a href="#cb185-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> p_lower, <span class="at">ymax =</span> p_upper),</span>
<span id="cb185-19"><a href="#cb185-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="at">fill =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb185-20"><a href="#cb185-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> N_A.std),</span>
<span id="cb185-21"><a href="#cb185-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"#8B9DAF"</span>) <span class="sc">+</span></span>
<span id="cb185-22"><a href="#cb185-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb185-23"><a href="#cb185-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"negative affect (standardized)"</span>) <span class="sc">+</span></span>
<span id="cb185-24"><a href="#cb185-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> record_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-120-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>With respect to <span class="math inline">\(\rho_{31}\)</span>, participant 115 showed both a higher intercept and higher level of variability toward the beginning of the study than participant 30 did. The meaning of <span class="math inline">\(\rho_{42}\)</span> is less clear, with these two. But at least you can get a sense of why you might want to include a <span class="math inline">\(\eta_1\)</span> parameter to allow response variability to change over time, and why you might want to allow that parameter to vary across participants. Whereas response variability increased quite a bit for participant 115 over time, it stayed about the same for participant 30, perhaps even decreasing a bit.</p>
</section>
<section id="time-to-go-multivariate" class="level3" data-number="14.6.4">
<h3 data-number="14.6.4" class="anchored" data-anchor-id="time-to-go-multivariate"><span class="header-section-number">14.6.4</span> Time to go multivariate</h3>
<p>For our final stage in this progression, we will fit what <span class="citation" data-cites="williams2020BayesianMulitivariate">Donald R. Williams et al. (<a href="references.html#ref-williams2020BayesianMulitivariate" role="doc-biblioref">2020</a>)</span> called a M-MELSM, a multivariate mixed-effects location scale model. Recall these data have measures of both negative and positive affect. The standardized values for PA are waiting for us in the <code>P_A.std</code> column. Within the <strong>brms</strong> framework, this is a combination of sensibilities from Bürkner’s vignettes on distributional models <span class="citation" data-cites="Bürkner2022Distributional">(<a href="references.html#ref-Bürkner2022Distributional" role="doc-biblioref">Bürkner, 2022b</a>)</span> and multivariate models <span class="citation" data-cites="Bürkner2022Multivariate">(<a href="references.html#ref-Bürkner2022Multivariate" role="doc-biblioref">Bürkner, 2022c</a>)</span>. We might express the statistical model as</p>
<p><span class="math display">\[\begin{align*}
\color{#A65141}{\begin{bmatrix} \text{NA}_{ij} \\ \text{PA}_{ij} \end{bmatrix}} &amp; \color{#A65141}\sim \color{#A65141}{\operatorname{MVNormal}\begin{pmatrix} \begin{bmatrix}
\mu_{ij}^\text{NA} \\ \mu_{ij}^\text{PA} \end{bmatrix}, \mathbf \Sigma \end{pmatrix}} \\

\mu_{ij}^{\color{#A65141}{\text{NA}}} &amp; = \beta_0^\text{NA} + \beta_1^\text{NA} \text{time}_{ij} + u_{0i}^\text{NA} + u_{1i}^\text{NA} \\
\mu_{ij}^{\color{#A65141}{\text{PA}}} &amp; = \beta_0^\text{PA} + \beta_1^\text{PA} \text{time}_{ij} + u_{0i}^\text{PA} + u_{1i}^\text{PA} \\
\log \left ( \sigma_{ij}^{\color{#A65141}{\text{NA}}} \right ) &amp; = \eta_0^\text{NA} + \eta_1^\text{NA} \text{time}_{ij} + u_{2i}^\text{NA} + u_{3i}^\text{NA} \\
\log \left ( \sigma_{ij}^{\color{#A65141}{\text{PA}}} \right ) &amp; = \eta_0^\text{PA} + \eta_1^\text{PA} \text{time}_{ij} + u_{2i}^\text{PA} + u_{3i}^\text{PA}  \\

\begin{bmatrix} u_{0i}^\text{NA}, u_{1i}^\text{NA}, u_{2i}^\text{NA}, u_{3i}^\text{NA}, u_{0i}^\text{PA}, u_{1i}^\text{PA}, u_{2i}^\text{PA}, u_{3i}^\text{PA} \end{bmatrix}' &amp; \sim \operatorname{MVNormal}(\mathbf 0, \mathbf S \mathbf R \mathbf S) \\

\mathbf S &amp; = \begin{bmatrix} \sigma_0^\text{NA} &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; \sigma_1^\text{NA} &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; \sigma_2^\text{NA} &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; \sigma_3^\text{NA} &amp; 0 &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; 0 &amp; \sigma_0^\text{PA} &amp; 0 &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; \sigma_1^\text{PA} &amp; 0 &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; \sigma_2^\text{PA} &amp; 0 \\ 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; 0 &amp; \sigma_3^\text{PA}  \end{bmatrix} \\

\mathbf R &amp; = \begin{bmatrix}
1 &amp; \rho_{12} &amp; \rho_{13} &amp; \rho_{14} &amp; \rho_{15} &amp; \rho_{16} &amp; \rho_{17} &amp; \rho_{18} \\
\rho_{21} &amp; 1 &amp; \rho_{23} &amp; \rho_{24} &amp; \rho_{25} &amp; \rho_{26} &amp; \rho_{27} &amp; \rho_{28} \\
\rho_{31} &amp; \rho_{32} &amp; 1 &amp; \rho_{34} &amp; \rho_{35} &amp; \rho_{36} &amp; \rho_{37} &amp; \rho_{38} \\
\rho_{41} &amp; \rho_{42} &amp; \rho_{43} &amp; 1 &amp; \rho_{45} &amp; \rho_{46} &amp; \rho_{47} &amp; \rho_{48} \\
\rho_{51} &amp; \rho_{52} &amp; \rho_{53} &amp; \rho_{54} &amp; 1 &amp; \rho_{56} &amp; \rho_{57} &amp; \rho_{58} \\
\rho_{61} &amp; \rho_{62} &amp; \rho_{63} &amp; \rho_{64} &amp; \rho_{65} &amp; 1 &amp; \rho_{67} &amp; \rho_{68} \\
\rho_{71} &amp; \rho_{72} &amp; \rho_{73} &amp; \rho_{74} &amp; \rho_{75} &amp; \rho_{76} &amp; 1 &amp; \rho_{78} \\
\rho_{81} &amp; \rho_{82} &amp; \rho_{83} &amp; \rho_{84} &amp; \rho_{85} &amp; \rho_{86} &amp; \rho_{87} &amp; 1
\end{bmatrix} \\

\beta_0^\text{NA} \text{ and } \beta_0^\text{PA} &amp; \sim \operatorname{Normal}(0, 0.2) \\
\beta_1^\text{NA} \text{ and } \beta_1^\text{PA} &amp; \sim \operatorname{Normal}(0, 1) \\
\eta_0^\text{NA},\dots, \eta_1^\text{PA} &amp; \sim \operatorname{Normal}(0, 1) \\
\sigma_0^\text{NA},\dots, \sigma_1^\text{PA} &amp; \sim \operatorname{Exponential}(1) \\
\mathbf R &amp; \sim \operatorname{LKJ}(2),
\end{align*}\]</span></p>
<p>where the <span class="math inline">\(\text{NA}\)</span> and <span class="math inline">\(\text{PA}\)</span> superscripts indicate which variable is connected with which parameter. This is a straight multivariate generalization from the previous model, <code>b14.14</code>. Now we have eight parameters varying across participants, resulting in an <span class="math inline">\(8 \times 8\)</span> <span class="math inline">\(\mathbf S\)</span> matrix and an <span class="math inline">\(8 \times 8\)</span> <span class="math inline">\(\mathbf R\)</span> matrix. Here’s the <code>brm()</code> code.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a>b14<span class="fl">.15</span> <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb186-2"><a href="#cb186-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat,</span>
<span id="cb186-3"><a href="#cb186-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> gaussian,</span>
<span id="cb186-4"><a href="#cb186-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bf</span>(<span class="fu">mvbind</span>(N_A.std, P_A.std) <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">|</span>i<span class="sc">|</span> record_id),</span>
<span id="cb186-5"><a href="#cb186-5" aria-hidden="true" tabindex="-1"></a>     sigma <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> day01 <span class="sc">|</span>i<span class="sc">|</span> record_id)) <span class="sc">+</span> <span class="fu">set_rescor</span>(<span class="at">rescor =</span> <span class="cn">FALSE</span>),</span>
<span id="cb186-6"><a href="#cb186-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(<span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept, <span class="at">resp =</span> NAstd),</span>
<span id="cb186-7"><a href="#cb186-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b, <span class="at">resp =</span> NAstd),</span>
<span id="cb186-8"><a href="#cb186-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">resp =</span> NAstd),</span>
<span id="cb186-9"><a href="#cb186-9" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb186-10"><a href="#cb186-10" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> Intercept, <span class="at">dpar =</span> sigma, <span class="at">resp =</span> NAstd),</span>
<span id="cb186-11"><a href="#cb186-11" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b, <span class="at">dpar =</span> sigma, <span class="at">resp =</span> NAstd),</span>
<span id="cb186-12"><a href="#cb186-12" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">dpar =</span> sigma, <span class="at">resp =</span> NAstd),</span>
<span id="cb186-13"><a href="#cb186-13" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb186-14"><a href="#cb186-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="fl">0.2</span>), <span class="at">class =</span> Intercept, <span class="at">resp =</span> PAstd),</span>
<span id="cb186-15"><a href="#cb186-15" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b, <span class="at">resp =</span> PAstd),</span>
<span id="cb186-16"><a href="#cb186-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">resp =</span> PAstd),</span>
<span id="cb186-17"><a href="#cb186-17" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb186-18"><a href="#cb186-18" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> Intercept, <span class="at">dpar =</span> sigma, <span class="at">resp =</span> PAstd),</span>
<span id="cb186-19"><a href="#cb186-19" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">class =</span> b, <span class="at">dpar =</span> sigma, <span class="at">resp =</span> PAstd),</span>
<span id="cb186-20"><a href="#cb186-20" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sd, <span class="at">dpar =</span> sigma, <span class="at">resp =</span> PAstd),</span>
<span id="cb186-21"><a href="#cb186-21" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb186-22"><a href="#cb186-22" aria-hidden="true" tabindex="-1"></a>            <span class="fu">prior</span>(<span class="fu">lkj</span>(<span class="dv">2</span>), <span class="at">class =</span> cor)),</span>
<span id="cb186-23"><a href="#cb186-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">iter =</span> <span class="dv">3000</span>, <span class="at">warmup =</span> <span class="dv">1000</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>,</span>
<span id="cb186-24"><a href="#cb186-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">14</span>,</span>
<span id="cb186-25"><a href="#cb186-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"fits/b14.15"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note how we used the <code>resp</code> argument to indicate which priors went with which criterion variables. For the sake of space, I’ll skip showing the <code>print()</code> output. By all means, check that summary out if you fit this model on your own. Though there may be some substantive insights to glean from looking at the population-level parameters and the hierarchical <span class="math inline">\(\sigma\)</span>’s, I’d argue the main action is in the <span class="math inline">\(\mathbf R\)</span> matrix. This time we’ll jump straight to showcasing their posterior means in a correlation matrix plot.</p>
<p>First we wrangle.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a>levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"beta[0]^'NA'"</span>, <span class="st">"beta[1]^'NA'"</span>, <span class="st">"eta[0]^'NA'"</span>, <span class="st">"eta[1]^'NA'"</span>,</span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"beta[0]^'PA'"</span>, <span class="st">"beta[1]^'PA'"</span>, <span class="st">"eta[0]^'PA'"</span>, <span class="st">"eta[1]^'PA'"</span>)</span>
<span id="cb187-3"><a href="#cb187-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-4"><a href="#cb187-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Two different options for ordering the parameters</span></span>
<span id="cb187-5"><a href="#cb187-5" aria-hidden="true" tabindex="-1"></a><span class="co"># levels &lt;- c("beta[0]^'NA'", "beta[1]^'NA'", "beta[0]^'PA'", "beta[1]^'PA'", "eta[0]^'NA'", "eta[1]^'NA'", "eta[0]^'PA'", "eta[1]^'PA'")</span></span>
<span id="cb187-6"><a href="#cb187-6" aria-hidden="true" tabindex="-1"></a><span class="co"># levels &lt;- c("beta[0]^'NA'", "beta[0]^'PA'", "beta[1]^'NA'", "beta[1]^'PA'","eta[0]^'NA'", "eta[0]^'PA'", "eta[1]^'NA'", "eta[1]^'PA'")</span></span>
<span id="cb187-7"><a href="#cb187-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-8"><a href="#cb187-8" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">&lt;-</span> <span class="fu">posterior_summary</span>(b14<span class="fl">.15</span>) <span class="sc">|&gt;</span> </span>
<span id="cb187-9"><a href="#cb187-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb187-10"><a href="#cb187-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"param"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb187-11"><a href="#cb187-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">str_detect</span>(param, <span class="st">"cor_"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb187-12"><a href="#cb187-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">param =</span> <span class="fu">str_remove</span>(param, <span class="st">"cor_record_id__"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb187-13"><a href="#cb187-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(param, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"left"</span>, <span class="st">"right"</span>), <span class="at">sep =</span> <span class="st">"__"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb187-14"><a href="#cb187-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb187-15"><a href="#cb187-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">left =</span> <span class="fu">case_when</span>(</span>
<span id="cb187-16"><a href="#cb187-16" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"NAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'NA'"</span>,</span>
<span id="cb187-17"><a href="#cb187-17" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"NAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'NA'"</span>,</span>
<span id="cb187-18"><a href="#cb187-18" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"sigma_NAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'NA'"</span>,</span>
<span id="cb187-19"><a href="#cb187-19" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"sigma_NAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'NA'"</span>,</span>
<span id="cb187-20"><a href="#cb187-20" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"PAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'PA'"</span>,</span>
<span id="cb187-21"><a href="#cb187-21" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"PAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'PA'"</span>,</span>
<span id="cb187-22"><a href="#cb187-22" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"sigma_PAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'PA'"</span>,</span>
<span id="cb187-23"><a href="#cb187-23" aria-hidden="true" tabindex="-1"></a>      left <span class="sc">==</span> <span class="st">"sigma_PAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'PA'"</span></span>
<span id="cb187-24"><a href="#cb187-24" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb187-25"><a href="#cb187-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">right =</span> <span class="fu">case_when</span>(</span>
<span id="cb187-26"><a href="#cb187-26" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"NAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'NA'"</span>,</span>
<span id="cb187-27"><a href="#cb187-27" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"NAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'NA'"</span>,</span>
<span id="cb187-28"><a href="#cb187-28" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"sigma_NAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'NA'"</span>,</span>
<span id="cb187-29"><a href="#cb187-29" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"sigma_NAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'NA'"</span>,</span>
<span id="cb187-30"><a href="#cb187-30" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"PAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'PA'"</span>,</span>
<span id="cb187-31"><a href="#cb187-31" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"PAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'PA'"</span>,</span>
<span id="cb187-32"><a href="#cb187-32" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"sigma_PAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'PA'"</span>,</span>
<span id="cb187-33"><a href="#cb187-33" aria-hidden="true" tabindex="-1"></a>      right <span class="sc">==</span> <span class="st">"sigma_PAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'PA'"</span></span>
<span id="cb187-34"><a href="#cb187-34" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb187-35"><a href="#cb187-35" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb187-36"><a href="#cb187-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">formatC</span>(Estimate, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">format =</span> <span class="st">"f"</span>) <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">"0."</span>, <span class="st">"."</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb187-37"><a href="#cb187-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">left  =</span> <span class="fu">factor</span>(left, <span class="at">levels =</span> levels),</span>
<span id="cb187-38"><a href="#cb187-38" aria-hidden="true" tabindex="-1"></a>         <span class="at">right =</span> <span class="fu">factor</span>(right, <span class="at">levels =</span> levels)) <span class="sc">|&gt;</span> </span>
<span id="cb187-39"><a href="#cb187-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">right =</span> <span class="fu">fct_rev</span>(right))</span>
<span id="cb187-40"><a href="#cb187-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-41"><a href="#cb187-41" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          left        right   Estimate  Est.Error       Q2.5       Q97.5 label
1 beta[0]^'NA' beta[1]^'NA' -0.2966930 0.08151278 -0.4492407 -0.13220155  -.30
2 beta[0]^'NA'  eta[0]^'NA'  0.6329190 0.04584605  0.5367935  0.71611940   .63
3 beta[1]^'NA'  eta[0]^'NA' -0.1869214 0.07811868 -0.3345458 -0.02844261  -.19
4 beta[0]^'NA'  eta[1]^'NA' -0.1494339 0.10282028 -0.3467451  0.05961868  -.15
5 beta[1]^'NA'  eta[1]^'NA'  0.5752405 0.09059977  0.3848376  0.73694797   .58
6  eta[0]^'NA'  eta[1]^'NA' -0.1433994 0.09469969 -0.3213243  0.04859973  -.14</code></pre>
</div>
</div>
<p>Now we plot!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span> </span>
<span id="cb189-2"><a href="#cb189-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(<span class="fu">rename</span>(rho, <span class="at">right =</span> left, <span class="at">left =</span> right),</span>
<span id="cb189-3"><a href="#cb189-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"left"</span>, <span class="st">"right"</span>, <span class="st">"Estimate"</span>, <span class="st">"Est.Error"</span>, <span class="st">"Q2.5"</span>, <span class="st">"Q97.5"</span>, <span class="st">"label"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb189-4"><a href="#cb189-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb189-5"><a href="#cb189-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> left, <span class="at">y =</span> right)) <span class="sc">+</span></span>
<span id="cb189-6"><a href="#cb189-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Estimate)) <span class="sc">+</span></span>
<span id="cb189-7"><a href="#cb189-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">4.5</span>, <span class="at">color =</span> <span class="st">"#100F14"</span>) <span class="sc">+</span></span>
<span id="cb189-8"><a href="#cb189-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">4.5</span>, <span class="at">color =</span> <span class="st">"#100F14"</span>) <span class="sc">+</span></span>
<span id="cb189-9"><a href="#cb189-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> label),</span>
<span id="cb189-10"><a href="#cb189-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">family =</span> <span class="st">"Courier"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb189-11"><a href="#cb189-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe, <span class="at">position =</span> <span class="st">"top"</span>) <span class="sc">+</span></span>
<span id="cb189-12"><a href="#cb189-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">labels =</span> ggplot2<span class="sc">:::</span>parse_safe) <span class="sc">+</span></span>
<span id="cb189-13"><a href="#cb189-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="fu">expression</span>(rho),</span>
<span id="cb189-14"><a href="#cb189-14" aria-hidden="true" tabindex="-1"></a>                       <span class="at">low =</span> <span class="st">"#59708b"</span>, <span class="at">mid =</span> <span class="st">"#FCF9F0"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>, <span class="at">midpoint =</span> <span class="dv">0</span>,</span>
<span id="cb189-15"><a href="#cb189-15" aria-hidden="true" tabindex="-1"></a>                       <span class="at">labels =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="st">""</span>, <span class="dv">0</span>, <span class="st">""</span>, <span class="dv">1</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb189-16"><a href="#cb189-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb189-17"><a href="#cb189-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb189-18"><a href="#cb189-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-125-1.png" class="img-fluid figure-img" width="600"></p>
</figure>
</div>
</div>
</div>
<p>The <code>full_join()</code> business just before the <strong>ggplot2</strong> code is how we got the full <span class="math inline">\(8 \times 8\)</span> matrix. If you’re curious, see what happens if you run the code without that part.</p>
<p>To help orient you to the plot, I’ve divided it into four quadrants. The upper left and lower right quadrants are the correlations among the varying parameters for the <code>N_A.std</code> and <code>P_A.std</code> ratings, respectively. The other two quadrants are the correlations for those parameters between <code>N_A.std</code> and <code>P_A.std</code>. As a reminder, this matrix, as with any other correlation matrix, is symmetrical across the diagonal.</p>
<p>To my eye, a few things pop out. First, the correlations within <code>N_A.std</code> are generally higher than those within <code>P_A.std</code>. Second, the correlations among the parameters between <code>N_A.std</code> and <code>P_A.std</code> are generally higher than those within them. Finally, all three of the largest correlations have to do with variation in the <span class="math inline">\(\eta\)</span> parameters. Two of them are basically the same as those we focused on for <code>b14.14</code>. The new one, <span class="math inline">\(\rho_{73}\)</span>, indicates that participants’ baseline ratings for <code>N_A.std</code> tended to vary in a similar way as their baseline ratings for <code>P_A.std</code>.</p>
<section id="plot-with-uncertainty" class="level4" data-number="14.6.4.1">
<h4 data-number="14.6.4.1" class="anchored" data-anchor-id="plot-with-uncertainty"><span class="header-section-number">14.6.4.1</span> Plot with uncertainty</h4>
<p>As fond as I am with that last correlation plot, it has a glaring defect: there is no expression of uncertainty. Sometimes we express uncertainty with percentile-based intervals. Other times we do so with marginal densities. But the correlation plots only describe the marginal posteriors for all the <span class="math inline">\(\rho\)</span> parameters with their means–no uncertainty. If you have one or a small few correlations to plot, coefficient or density plots might do. However, they don’t scale well. If you don’t believe me, just try. I’m not sure there are any good solutions to this, but it can be helpful to at least try grappling with the issue.</p>
<p>Fortunately for us, <a href="https://twitter.com/mjskay">Matthew Kay</a> (creator of the <a href="https://mjskay.github.io/tidybayes/"><strong>tidybayes</strong> package</a>) has already tried his hand at a few approaches. For all the deets, check out the <a href="https://github.com/mjskay/uncertainty-examples/blob/master/multivariate-regression.md">multivariate-regression.md</a> file in his <a href="https://github.com/mjskay/uncertainty-examples">uncertainty-examples</a> GitHub repo. One of his more imaginative approaches is to use what he calls dithering. Imagine breaking each of the cells in our correlation plot, above, into a <span class="math inline">\(50 \times 50 = 2{,}500\)</span>-cell grid. Now assign each of the cells within that grid one of the values from the HMC draws of that correlation’s posterior distribution. Then color code each of those cells by that value in the same basic way we color coded our previous correlation plots. Simultaneously do that for all of the correlations within the <span class="math inline">\(\mathbf R\)</span> matrix and plot them in a faceted plot. That’s the essence of the dithering approach. This is all probably hard to make sense of in words. Hopefully it will all come together with a little code and the resulting plot. Hold on to your hat.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a>levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"beta[0]^'NA'"</span>, <span class="st">"beta[1]^'NA'"</span>, <span class="st">"eta[0]^'NA'"</span>, <span class="st">"eta[1]^'NA'"</span>,</span>
<span id="cb190-2"><a href="#cb190-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"beta[0]^'PA'"</span>, <span class="st">"beta[1]^'PA'"</span>, <span class="st">"eta[0]^'PA'"</span>, <span class="st">"eta[1]^'PA'"</span>)</span>
<span id="cb190-3"><a href="#cb190-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-4"><a href="#cb190-4" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">&lt;-</span> <span class="fu">as_draws_df</span>(b14<span class="fl">.15</span>) <span class="sc">|&gt;</span> </span>
<span id="cb190-5"><a href="#cb190-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">"cor_"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb190-6"><a href="#cb190-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">50</span> <span class="sc">*</span> <span class="dv">50</span>) <span class="sc">|&gt;</span> </span>
<span id="cb190-7"><a href="#cb190-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(<span class="fu">crossing</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, <span class="at">y =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb190-8"><a href="#cb190-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(x<span class="sc">:</span>y)) <span class="sc">|&gt;</span> </span>
<span id="cb190-9"><a href="#cb190-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">name =</span> <span class="fu">str_remove</span>(name, <span class="st">"cor_record_id__"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb190-10"><a href="#cb190-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(name, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"col"</span>, <span class="st">"row"</span>), <span class="at">sep =</span> <span class="st">"__"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb190-11"><a href="#cb190-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb190-12"><a href="#cb190-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">case_when</span>(</span>
<span id="cb190-13"><a href="#cb190-13" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"NAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'NA'"</span>,</span>
<span id="cb190-14"><a href="#cb190-14" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"NAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'NA'"</span>,</span>
<span id="cb190-15"><a href="#cb190-15" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"sigma_NAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'NA'"</span>,</span>
<span id="cb190-16"><a href="#cb190-16" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"sigma_NAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'NA'"</span>,</span>
<span id="cb190-17"><a href="#cb190-17" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"PAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'PA'"</span>,</span>
<span id="cb190-18"><a href="#cb190-18" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"PAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'PA'"</span>,</span>
<span id="cb190-19"><a href="#cb190-19" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"sigma_PAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'PA'"</span>,</span>
<span id="cb190-20"><a href="#cb190-20" aria-hidden="true" tabindex="-1"></a>      col <span class="sc">==</span> <span class="st">"sigma_PAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'PA'"</span></span>
<span id="cb190-21"><a href="#cb190-21" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb190-22"><a href="#cb190-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">row =</span> <span class="fu">case_when</span>(</span>
<span id="cb190-23"><a href="#cb190-23" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"NAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'NA'"</span>,</span>
<span id="cb190-24"><a href="#cb190-24" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"NAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'NA'"</span>,</span>
<span id="cb190-25"><a href="#cb190-25" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"sigma_NAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'NA'"</span>,</span>
<span id="cb190-26"><a href="#cb190-26" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"sigma_NAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'NA'"</span>,</span>
<span id="cb190-27"><a href="#cb190-27" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"PAstd_Intercept"</span>       <span class="sc">~</span> <span class="st">"beta[0]^'PA'"</span>,</span>
<span id="cb190-28"><a href="#cb190-28" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"PAstd_day01"</span>           <span class="sc">~</span> <span class="st">"beta[1]^'PA'"</span>,</span>
<span id="cb190-29"><a href="#cb190-29" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"sigma_PAstd_Intercept"</span> <span class="sc">~</span> <span class="st">"eta[0]^'PA'"</span>,</span>
<span id="cb190-30"><a href="#cb190-30" aria-hidden="true" tabindex="-1"></a>      row <span class="sc">==</span> <span class="st">"sigma_PAstd_day01"</span>     <span class="sc">~</span> <span class="st">"eta[1]^'PA'"</span></span>
<span id="cb190-31"><a href="#cb190-31" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb190-32"><a href="#cb190-32" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb190-33"><a href="#cb190-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">col =</span> <span class="fu">factor</span>(col, <span class="at">levels =</span> levels),</span>
<span id="cb190-34"><a href="#cb190-34" aria-hidden="true" tabindex="-1"></a>         <span class="at">row =</span> <span class="fu">factor</span>(row, <span class="at">levels =</span> levels))</span>
<span id="cb190-35"><a href="#cb190-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-36"><a href="#cb190-36" aria-hidden="true" tabindex="-1"></a>rho <span class="sc">|&gt;</span> </span>
<span id="cb190-37"><a href="#cb190-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(<span class="fu">rename</span>(rho, <span class="at">col =</span> row, <span class="at">row =</span> col),</span>
<span id="cb190-38"><a href="#cb190-38" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"col"</span>, <span class="st">"row"</span>, <span class="st">"value"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb190-39"><a href="#cb190-39" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb190-40"><a href="#cb190-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">fill =</span> value)) <span class="sc">+</span></span>
<span id="cb190-41"><a href="#cb190-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>() <span class="sc">+</span></span>
<span id="cb190-42"><a href="#cb190-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb190-43"><a href="#cb190-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb190-44"><a href="#cb190-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="fu">expression</span>(rho),</span>
<span id="cb190-45"><a href="#cb190-45" aria-hidden="true" tabindex="-1"></a>                       <span class="at">low =</span> <span class="st">"#59708b"</span>, <span class="at">mid =</span> <span class="st">"#FCF9F0"</span>, <span class="at">high =</span> <span class="st">"#A65141"</span>, <span class="at">midpoint =</span> <span class="dv">0</span>,</span>
<span id="cb190-46"><a href="#cb190-46" aria-hidden="true" tabindex="-1"></a>                       <span class="at">labels =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="st">""</span>, <span class="dv">0</span>, <span class="st">""</span>, <span class="dv">1</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb190-47"><a href="#cb190-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(row <span class="sc">~</span> col, <span class="at">labeller =</span> label_parsed, <span class="at">switch =</span> <span class="st">"y"</span>) <span class="sc">+</span></span>
<span id="cb190-48"><a href="#cb190-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="14_files/figure-html/unnamed-chunk-126-1.png" class="img-fluid figure-img" width="600"></p>
</figure>
</div>
</div>
</div>
<p>From Kay’s GitHub repo, we read: “This is akin to something like an icon array. You should still be able to see the average color (thanks to the human visual system’s ensembling processing), but also get a sense of the uncertainty by how ‘dithered’ a square looks.” Hopefully this will give you a little inspiration to find new and better ways to express the posterior uncertainty in your Bayesian correlation plots. If you come up with any great solutions, <a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed/issues">let the rest of us know</a>!</p>
</section>
</section>
<section id="growth-modelmelsm-wrap-up" class="level3" data-number="14.6.5">
<h3 data-number="14.6.5" class="anchored" data-anchor-id="growth-modelmelsm-wrap-up"><span class="header-section-number">14.6.5</span> Growth model/MELSM wrap-up</h3>
<p>This bonus section introduced a lot of material. To learn more about the conventional multilevel growth model and its extensions, check out</p>
<ul>
<li>Singer and Willett’s <span class="citation" data-cites="singerAppliedLongitudinalData2003">(<a href="references.html#ref-singerAppliedLongitudinalData2003" role="doc-biblioref">2003</a>)</span> text, <a href="https://oxford.universitypressscholarship.com/view/10.1093/acprof:oso/9780195152968.001.0001/acprof-9780195152968"><em>Applied longitudinal data analysis: Modeling change and event occurrence</em></a>;</li>
<li>My <strong>brms</strong>/<strong>tidyverse</strong> translation of that text, <a href="https://solomon.quarto.pub/alda/"><em>Applied Longitudinal Data Analysis in brms and the tidyverse </em></a>; or</li>
<li>Hoffman’s <span class="citation" data-cites="hoffmanLongitudinalAnalysisModeling2015">(<a href="references.html#ref-hoffmanLongitudinalAnalysisModeling2015" role="doc-biblioref">2015</a>)</span> text, <a href="https://www.routledge.com/Longitudinal-Analysis-Modeling-Within-Person-Fluctuation-and-Change/Hoffman/p/book/9780415876025"><em>Longitudinal analysis: Modeling within-person fluctuation and change</em></a>.</li>
</ul>
<p>To learn more about the MELSM approach and its extensions, check out</p>
<ul>
<li><span class="citation" data-cites="hedekerApplicationMixedeffectsLocation2008">Hedeker et al. (<a href="references.html#ref-hedekerApplicationMixedeffectsLocation2008" role="doc-biblioref">2008</a>)</span>, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2424261/"><em>An application of a mixed-effects location scale model for analysis of ecological momentary assessment (EMA) data</em></a>;</li>
<li><span class="citation" data-cites="hedekerModelingWithinsubjectVariance2012">Hedeker et al. (<a href="references.html#ref-hedekerModelingWithinsubjectVariance2012" role="doc-biblioref">2012</a>)</span>, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3655706/"><em>Modeling between- and within-subject variance in ecological momentary assessment (EMA) data using mixed-effects location scale models</em></a>;</li>
<li>Williams’ and colleagues’ <span class="citation" data-cites="williams2020BayesianMulitivariate">(<a href="references.html#ref-williams2020BayesianMulitivariate" role="doc-biblioref">2020</a>)</span> paper, <a href="https://psyarxiv.com/4kfjp/"><em>Bayesian multivariate mixed-effects location scale modeling of longitudinal relations among affective traits, states, and physical activity</em></a>;</li>
<li>Williams’ and colleagues’ <span class="citation" data-cites="williams2021beneath">(<a href="references.html#ref-williams2021beneath" role="doc-biblioref">2021</a>)</span> paper, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8572133/"><em>Beneath the surface: Unearthing within-person variability and mean relations with Bayesian mixed models</em></a>;</li>
<li>Williams’ and colleagues’ <span class="citation" data-cites="williamsBayesianNonlinearMixedeffects2019a">(<a href="references.html#ref-williamsBayesianNonlinearMixedeffects2019a" role="doc-biblioref">2019</a>)</span> paper, <a href="https://doi.org/10.3758/s13428-019-01255-9"><em>A Bayesian nonlinear mixed-effects location scale model for learning</em></a>;</li>
<li>Williams’ tutorial blog post, <a href="https://rstudio-pubs-static.s3.amazonaws.com/593838_838639e42b1643df9886a26cb922ab05.html"><em>A defining feature of cognitive interference tasks: Heterogeneous within-person variance</em></a>.</li>
</ul>
<p>From a <strong>brms</strong> standpoint, it might also be helpful to brush up on</p>
<ul>
<li>Bürkner’s <span class="citation" data-cites="Bürkner2022Distributional">(<a href="references.html#ref-Bürkner2022Distributional" role="doc-biblioref">2022b</a>)</span> vignette, <a href="https://CRAN.R-project.org/package=brms/vignettes/brms_distreg.html"><em>Estimating distributional models with brms</em></a> and</li>
<li>Bürkner’s <span class="citation" data-cites="Bürkner2022Multivariate">(<a href="references.html#ref-Bürkner2022Multivariate" role="doc-biblioref">2022c</a>)</span> vignette, <a href="https://CRAN.R-project.org/package=brms/vignettes/brms_multivariate.html"><em>Estimating multivariate models with brms</em></a>.</li>
</ul>
</section>
</section>
<section id="session-info" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="session-info">Session info</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: aarch64-apple-darwin20
Running under: macOS Ventura 13.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] ape_5.8-1             ggtree_4.1.1.002      ggrepel_0.9.6         rethinking_2.42       dagitty_0.3-4        
 [6] ggdag_0.2.13          bayesplot_1.15.0.9000 patchwork_1.3.2       brms_2.23.0           Rcpp_1.1.0           
[11] tidybayes_3.0.7       posterior_1.6.1.9000  cmdstanr_0.9.0        dutchmasters_0.1.0    lubridate_1.9.4      
[16] forcats_1.0.1         stringr_1.6.0         dplyr_1.1.4           purrr_1.2.1           readr_2.1.5          
[21] tidyr_1.3.2           tibble_3.3.1          ggplot2_4.0.1         tidyverse_2.0.0      

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      tensorA_0.36.2.1        rstudioapi_0.17.1       jsonlite_2.0.0          shape_1.4.6.1          
  [6] magrittr_2.0.4          TH.data_1.1-4           estimability_1.5.1      farver_2.1.2            rmarkdown_2.30         
 [11] fs_1.6.6                vctrs_0.7.0             memoise_2.0.1           htmltools_0.5.9         distributional_0.5.0   
 [16] curl_7.0.0              gridGraphics_0.5-1      StanHeaders_2.36.0.9000 htmlwidgets_1.6.4       plyr_1.8.9             
 [21] sandwich_3.1-1          emmeans_1.11.2-8        zoo_1.8-14              cachem_1.1.0            igraph_2.2.0           
 [26] lifecycle_1.0.5         pkgconfig_2.0.3         Matrix_1.7-3            R6_2.6.1                fastmap_1.2.0          
 [31] digest_0.6.39           aplot_0.2.9             ps_1.9.1                invgamma_1.2            labeling_0.4.3         
 [36] timechange_0.3.0        polyclip_1.10-7         abind_1.4-8             compiler_4.5.1          fontquiver_0.2.1       
 [41] withr_3.0.2             S7_0.2.1                backports_1.5.0         inline_0.3.21           viridis_0.6.5          
 [46] QuickJSR_1.8.1          hexbin_1.28.5           pkgbuild_1.4.8          ggforce_0.5.0           MASS_7.3-65            
 [51] rappdirs_0.3.3          loo_2.9.0.9000          tools_4.5.1             glue_1.8.0              nlme_3.1-168           
 [56] grid_4.5.1              checkmate_2.3.3         reshape2_1.4.5          generics_0.1.4          gtable_0.3.6           
 [61] tzdb_0.5.0              hms_1.1.4               tidygraph_1.3.1         utf8_1.2.6              pillar_1.11.1          
 [66] ggdist_3.3.3            yulab.utils_0.2.3       splines_4.5.1           tweenr_2.0.3            treeio_1.34.0          
 [71] lattice_0.22-7          survival_3.8-3          tidyselect_1.2.1        fontLiberation_0.1.0    knitr_1.51             
 [76] fontBitstreamVera_0.1.1 arrayhelpers_1.1-0      gridExtra_2.3           V8_8.0.1                stats4_4.5.1           
 [81] xfun_0.55               graphlayouts_1.2.2      bridgesampling_1.2-1    matrixStats_1.5.0       rstan_2.36.0.9000      
 [86] stringi_1.8.7           lazyeval_0.2.2          ggfun_0.2.0             yaml_2.3.12             boot_1.3-31            
 [91] evaluate_1.0.5          codetools_0.2-20        ggraph_2.2.2            gdtools_0.4.4           emo_0.0.0.9000         
 [96] ggplotify_0.1.3         cli_3.6.5               RcppParallel_5.1.11-1   systemfonts_1.3.1       xtable_1.8-4           
[101] processx_3.8.6          coda_0.19-4.1           svUnit_1.0.8            rstantools_2.5.0.9000   assertthat_0.2.1       
[106] Brobdingnag_1.2-9       viridisLite_0.4.2       mvtnorm_1.3-3           tidytree_0.4.7          ggiraph_0.9.2          
[111] scales_1.4.0            nleqslv_3.3.5           crayon_1.5.3            rlang_1.1.7             multcomp_1.4-29        </code></pre>
</div>
</div>
</section>
<section id="comments" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="comments">Comments</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list" style="display: none">
<div id="ref-angristDoesCompulsorySchool1991" class="csl-entry" role="listitem">
Angrist, J. D., &amp; Keueger, A. B. (1991). Does compulsory school attendance affect schooling and earnings? <em>The Quarterly Journal of Economics</em>, <em>106</em>(4), 979–1014. <a href="https://doi.org/10.2307/2937954">https://doi.org/10.2307/2937954</a>
</div>
<div id="ref-betancourtRobustGaussianProcesses2017" class="csl-entry" role="listitem">
Betancourt, M. (2017). <em>Robust <span>Gaussian</span> processes in <span>Stan</span></em>. <a href="https://betanalpha.github.io/assets/case_studies/gp_part3/part3.html">https://betanalpha.github.io/assets/case_studies/gp_part3/part3.html</a>
</div>
<div id="ref-brms2022RM" class="csl-entry" role="listitem">
Bürkner, P.-C. (2022a). <em><span class="nocase">brms</span> reference manual, <span>Version</span> 2.18.0</em>. <a href="https://CRAN.R-project.org/package=brms/brms.pdf">https://CRAN.R-project.org/package=brms/brms.pdf</a>
</div>
<div id="ref-Bürkner2022Distributional" class="csl-entry" role="listitem">
Bürkner, P.-C. (2022b). <em>Estimating distributional models with brms</em>. <a href="https://CRAN.R-project.org/package=brms/vignettes/brms_distreg.html">https://CRAN.R-project.org/package=brms/vignettes/brms_distreg.html</a>
</div>
<div id="ref-Bürkner2022Multivariate" class="csl-entry" role="listitem">
Bürkner, P.-C. (2022c). <em>Estimating multivariate models with brms</em>. <a href="https://CRAN.R-project.org/package=brms/vignettes/brms_multivariate.html">https://CRAN.R-project.org/package=brms/vignettes/brms_multivariate.html</a>
</div>
<div id="ref-Bürkner2022Phylogenetic" class="csl-entry" role="listitem">
Bürkner, P.-C. (2022d). <em>Estimating phylogenetic multilevel models with brms</em>. <a href="https://CRAN.R-project.org/package=brms/vignettes/brms_phylogenetics.html">https://CRAN.R-project.org/package=brms/vignettes/brms_phylogenetics.html</a>
</div>
<div id="ref-gelmanWhyHighorderPolynomials2019" class="csl-entry" role="listitem">
Gelman, A., &amp; Imbens, G. (2019). Why high-order polynomials should not be used in regression discontinuity designs. <em>Journal of Business &amp; Economic Statistics</em>, <em>37</em>(3), 447–456. <a href="https://doi.org/10.1080/07350015.2017.1366909">https://doi.org/10.1080/07350015.2017.1366909</a>
</div>
<div id="ref-hedekerApplicationMixedeffectsLocation2008" class="csl-entry" role="listitem">
Hedeker, D., Mermelstein, R. J., &amp; Demirtas, H. (2008). An application of a mixed-effects location scale model for analysis of ecological momentary assessment (<span>EMA</span>) data. <em>Biometrics</em>, <em>64</em>(2), 627–634. <a href="https://doi.org/10.1111/j.1541-0420.2007.00924.x">https://doi.org/10.1111/j.1541-0420.2007.00924.x</a>
</div>
<div id="ref-hedekerModelingWithinsubjectVariance2012" class="csl-entry" role="listitem">
Hedeker, D., Mermelstein, R. J., &amp; Demirtas, H. (2012). Modeling between- and within-<span>Subject</span> variance in ecological momentary assessment (<span>EMA</span>) data using mixed-effects location scale models. <em>Statistics in Medicine</em>, <em>31</em>(27). <a href="https://doi.org/10.1002/sim.5338">https://doi.org/10.1002/sim.5338</a>
</div>
<div id="ref-hoffmanLongitudinalAnalysisModeling2015" class="csl-entry" role="listitem">
Hoffman, L. (2015). <em>Longitudinal analysis: <span>Modeling</span> within-<span>Person</span> fluctuation and change</em> (1 edition). Routledge. <a href="https://www.routledge.com/Longitudinal-Analysis-Modeling-Within-Person-Fluctuation-and-Change/Hoffman/p/book/9780415876025">https://www.routledge.com/Longitudinal-Analysis-Modeling-Within-Person-Fluctuation-and-Change/Hoffman/p/book/9780415876025</a>
</div>
<div id="ref-R-invgamma" class="csl-entry" role="listitem">
Kahle, D., &amp; Stamey, J. (2017). <em><span class="nocase">invgamma</span>: <span>The</span> inverse gamma distribution</em> [Manual]. <a href="https://CRAN.R-project.org/package=invgamma">https://CRAN.R-project.org/package=invgamma</a>
</div>
<div id="ref-kayMarginalDistributionSingle2020" class="csl-entry" role="listitem">
Kay, M. (2020). <em>Marginal distribution of a single correlation from an <span>LKJ</span> distribution</em>. <a href="https://mjskay.github.io/ggdist/reference/lkjcorr_marginal.html">https://mjskay.github.io/ggdist/reference/lkjcorr_marginal.html</a>
</div>
<div id="ref-kosterFoodSharingNetworks2014" class="csl-entry" role="listitem">
Koster, J. M., &amp; Leckie, G. (2014). Food sharing networks in lowland <span>Nicaragua</span>: <span>An</span> application of the social relations model to count data. <em>Social Networks</em>, <em>38</em>, 100–110. <a href="https://doi.org/10.1016/j.socnet.2014.02.002">https://doi.org/10.1016/j.socnet.2014.02.002</a>
</div>
<div id="ref-kruschkeDoingBayesianData2015" class="csl-entry" role="listitem">
Kruschke, J. K. (2015). <em>Doing <span>Bayesian</span> data analysis: <span>A</span> tutorial with <span>R</span>, <span>JAGS</span>, and <span>Stan</span></em>. Academic Press. <a href="https://sites.google.com/site/doingbayesiandataanalysis/">https://sites.google.com/site/doingbayesiandataanalysis/</a>
</div>
<div id="ref-kurzAppliedLongitudinalDataAnalysis2026" class="csl-entry" role="listitem">
Kurz, A. S. (2026). <em>Applied longitudinal data analysis in brms and the tidyverse</em> (version 0.0.4). <a href="https://solomon.quarto.pub/alda/">https://solomon.quarto.pub/alda/</a>
</div>
<div id="ref-mcelreathStatisticalRethinkingBayesian2015" class="csl-entry" role="listitem">
McElreath, R. (2015). <em>Statistical rethinking: <span>A Bayesian</span> course with examples in <span>R</span> and <span>Stan</span></em>. CRC press. <a href="https://xcelab.net/rm/statistical-rethinking/">https://xcelab.net/rm/statistical-rethinking/</a>
</div>
<div id="ref-mcelreathStatisticalRethinkingBayesian2020" class="csl-entry" role="listitem">
McElreath, R. (2020). <em>Statistical rethinking: <span>A Bayesian</span> course with examples in <span>R</span> and <span>Stan</span></em> (Second Edition). CRC Press. <a href="https://xcelab.net/rm/statistical-rethinking/">https://xcelab.net/rm/statistical-rethinking/</a>
</div>
<div id="ref-R-ape" class="csl-entry" role="listitem">
Paradis, Emmanuel, Blomberg, S., Bolker, B., Brown, J., Claramunt, S., Claude, J., Cuong, H. S., Desper, R., Didier, G., Durand, B., Dutheil, J., Ewing, R., Gascuel, O., Guillerme, T., Heibl, C., Ives, A., Jones, B., Krah, F., Lawson, D., … de Vienne, D. (2022). <em><span class="nocase">ape</span>: <span>Analyses</span> of phylogenetics and evolution</em> [Manual]. <a href="https://CRAN.R-project.org/package=ape">https://CRAN.R-project.org/package=ape</a>
</div>
<div id="ref-ape2019" class="csl-entry" role="listitem">
Paradis, E., &amp; Schliep, K. (2019). <span class="nocase">ape</span> 5.0: <span>An</span> environment for modern phylogenetics and evolutionary analyses in <span>R</span>. <em>Bioinformatics (Oxford, England)</em>, <em>35</em>, 526–528. <a href="https://doi.org/10.1093/bioinformatics/bty633">https://doi.org/10.1093/bioinformatics/bty633</a>
</div>
<div id="ref-pengMasteringSoftwareDevelopment2017" class="csl-entry" role="listitem">
Peng, R. D., Kross, S., &amp; Anderson, B. (2017). <em>Mastering software development in {<span>R</span>}</em>. <a href="https://github.com/rdpeng/RProgDA">https://github.com/rdpeng/RProgDA</a>
</div>
<div id="ref-rastModelingIndividualDifferences2012" class="csl-entry" role="listitem">
Rast, P., Hofer, S. M., &amp; Sparks, C. (2012). Modeling individual differences in within-person variation of negative and positive affect in a mixed effects location scale model using <span>BUGS</span>/<span>JAGS</span>. <em>Multivariate Behavioral Research</em>, <em>47</em>(2), 177–200. <a href="https://doi.org/10.1080/00273171.2012.658328">https://doi.org/10.1080/00273171.2012.658328</a>
</div>
<div id="ref-singerAppliedLongitudinalData2003" class="csl-entry" role="listitem">
Singer, J. D., &amp; Willett, J. B. (2003). <em>Applied longitudinal data analysis: <span>Modeling</span> change and event occurrence</em>. Oxford University Press, USA. <a href="https://oxford.universitypressscholarship.com/view/10.1093/acprof:oso/9780195152968.001.0001/acprof-9780195152968">https://oxford.universitypressscholarship.com/view/10.1093/acprof:oso/9780195152968.001.0001/acprof-9780195152968</a>
</div>
<div id="ref-streetCoevolutionCulturalIntelligence2017" class="csl-entry" role="listitem">
Street, S. E., Navarrete, A. F., Reader, S. M., &amp; Laland, K. N. (2017). Coevolution of cultural intelligence, extended life history, sociality, and brain size in primates. <em>Proceedings of the National Academy of Sciences</em>, <em>114</em>(30), 7908–7914. <a href="https://doi.org/10.1073/pnas.1620734114">https://doi.org/10.1073/pnas.1620734114</a>
</div>
<div id="ref-R-dutchmasters" class="csl-entry" role="listitem">
Thoen, E. (2022). <em><span class="nocase">dutchmasters</span></em> [Manual]. <a href="https://github.com/EdwinTh/dutchmasters">https://github.com/EdwinTh/dutchmasters</a>
</div>
<div id="ref-vermeerGirlPearlEarring1665" class="csl-entry" role="listitem">
Vermeer, J. (1665). <em>Girl with a pearl earring</em>.
</div>
<div id="ref-watsonPANASDevelopment1988" class="csl-entry" role="listitem">
Watson, D., Clark, L. A., &amp; Tellegen, A. (1988). Development and validation of brief measures of positive and negative affect: <span>The PANAS</span> scales. <em>Journal of Personality and Social Psychology</em>, <em>54</em>(6), 1063–1070. <a href="https://doi.org/10.1037/0022-3514.54.6.1063">https://doi.org/10.1037/0022-3514.54.6.1063</a>
</div>
<div id="ref-wickhamGgplot2ElegantGraphics2016" class="csl-entry" role="listitem">
Wickham, H. (2016). <em><span class="nocase">ggplot2</span>: <span>Elegant</span> graphics for data analysis</em>. Springer-Verlag New York. <a href="https://ggplot2-book.org/">https://ggplot2-book.org/</a>
</div>
<div id="ref-williams2020BayesianMulitivariate" class="csl-entry" role="listitem">
Williams, Donald R., Martin, S. R., Liu, S., &amp; Rast, P. (2020). Bayesian multivariate mixed-effects location scale modeling of longitudinal relations among affective traits, states, and physical activity. <em>European Journal of Psychological Assessment</em>, <em>36</em>(6), 981–997. <a href="https://doi.org/10.1027/1015-5759/a000624">https://doi.org/10.1027/1015-5759/a000624</a>
</div>
<div id="ref-williamsBayesianMultivariateMixedeffects2021" class="csl-entry" role="listitem">
Williams, Donald R., Martin, S. R., Liu, S., &amp; Rast, P. (2021). Bayesian multivariate mixed-effects location scale modeling of longitudinal relations among affective traits, states, and physical activity. <em>European Journal of Psychological Assessment</em>. <a href="https://doi.org/10.1027/1015-5759/a000624">https://doi.org/10.1027/1015-5759/a000624</a>
</div>
<div id="ref-williams2022PuttingTheIndividual" class="csl-entry" role="listitem">
Williams, Donald R., Martin, S. R., &amp; Rast, P. (2022). Putting the individual into reliability: <span>Bayesian</span> testing of homogeneous within-<span>Person</span> variance in hierarchical models. <em>Behavior Research Methods</em>, <em>54</em>(3), 1272–1290. <a href="https://doi.org/10.3758/s13428-021-01646-x">https://doi.org/10.3758/s13428-021-01646-x</a>
</div>
<div id="ref-williams2021beneath" class="csl-entry" role="listitem">
Williams, Donald R., Mulder, J., Rouder, J. N., &amp; Rast, P. (2021). Beneath the surface: <span>Unearthing</span> within-<span>Person</span> variability and mean relations with <span>Bayesian</span> mixed models. <em>Psychological Methods</em>, <em>26</em>(1), 74. <a href="https://doi.org/10.1037/met0000270">https://doi.org/10.1037/met0000270</a>
</div>
<div id="ref-williamsBayesianNonlinearMixedeffects2019a" class="csl-entry" role="listitem">
Williams, Donald R., Zimprich, D. R., &amp; Rast, P. (2019). A <span>Bayesian</span> nonlinear mixed-effects location scale model for learning. <em>Behavior Research Methods</em>, <em>51</em>(5), 1968–1986. <a href="https://doi.org/10.3758/s13428-019-01255-9">https://doi.org/10.3758/s13428-019-01255-9</a>
</div>
<div id="ref-yuUsingGgtreeVisualize2020" class="csl-entry" role="listitem">
Yu, G. (2020a). Using ggtree to visualize data on tree-like structures. <em>Current Protocols in Bioinformatics</em>, <em>69</em>(1), e96. <a href="https://doi.org/10.1002/cpbi.96">https://doi.org/10.1002/cpbi.96</a>
</div>
<div id="ref-yuDataIntegrationManipulation2020" class="csl-entry" role="listitem">
Yu, G. (2020b). <em>Data integration, manipulation and visualization of phylogenetic trees</em>. <a href="https://yulab-smu.github.io/treedata-book/">https://yulab-smu.github.io/treedata-book/</a>
</div>
<div id="ref-yuTwoMethodsMapping2018" class="csl-entry" role="listitem">
Yu, G., Lam, T. T.-Y., Zhu, H., &amp; Guan, Y. (2018). Two methods for mapping and visualizing associated data on phylogeny using ggtree. <em>Molecular Biology and Evolution</em>, <em>35</em>(12), 3041–3043. <a href="https://doi.org/10.1093/molbev/msy194">https://doi.org/10.1093/molbev/msy194</a>
</div>
<div id="ref-yuGgtreePackageVisualization2017" class="csl-entry" role="listitem">
Yu, G., Smith, D. K., Zhu, H., Guan, Y., &amp; Lam, T. T.-Y. (2017). <span class="nocase">ggtree</span>: <span>An R</span> package for visualization and annotation of phylogenetic trees with their covariates and other associated data. <em>Methods in Ecology and Evolution</em>, <em>8</em>(1), 28–36. <a href="https://doi.org/10.1111/2041-210X.12628">https://doi.org/10.1111/2041-210X.12628</a>
</div>
</div>
</section>


</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/solomon\.quarto\.pub\/sr2\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<input type="hidden" id="giscus-base-theme" value="light">
<input type="hidden" id="giscus-alt-theme" value="dark">
<script>
  function loadGiscus() {
    // Function to get the theme based on body class
    const getTheme = () => {
      let baseTheme = document.getElementById('giscus-base-theme').value;
      let altTheme = document.getElementById('giscus-alt-theme').value;
      return document.body.classList.contains('quarto-dark') ? altTheme : baseTheme;
    };
    const script = document.createElement("script");
    script.src = "https://giscus.app/client.js";
    script.async = true;
    script.dataset.repo = "ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed";
    script.dataset.repoId = "MDEwOlJlcG9zaXRvcnkxNjE5MTg0MTE=";
    script.dataset.category = "General";
    script.dataset.categoryId = "DIC_kwDOCaaty84C1AJZ";
    script.dataset.mapping = "title";
    script.dataset.reactionsEnabled = "1";
    script.dataset.emitMetadata = "0";
    script.dataset.inputPosition = "top";
    script.dataset.theme = getTheme();
    script.dataset.lang = "en";
    script.crossOrigin = "anonymous";
    // Append the script to the desired div instead of at the end of the body
    document.getElementById("quarto-content").appendChild(script);
  }
  loadGiscus();
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./13.html" class="pagination-link" aria-label="Models With Memory">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Models With Memory</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./15.html" class="pagination-link" aria-label="Missing Data and Other Opportunities">
        <span class="nav-page-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Missing Data and Other Opportunities</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/ASKurz/Statistical_Rethinking_with_brms_ggplot2_and_the_tidyverse_2_ed/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>